{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Palaeoprot/pFind/blob/main/pFind_Analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lNSR-u5cC5f_"
      },
      "source": [
        "#**pFind Analysis** Scripts\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "taooEB7d0GxA"
      },
      "source": [
        "# Global Varibles"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "yc9xr8HoVdwJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#!pip install Biopython"
      ],
      "metadata": {
        "id": "9oFIm2X9jRKS"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R1Ec3Q5d0nzW"
      },
      "source": [
        "## Import Packages"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "GiAz-Tf_OGox",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8552b604-f550-421f-85bb-952b37f348a6"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "# Data manipulation:\n",
        "import pandas as pd\n",
        "pd.set_option('display.max_rows', 500)  # Adjust the value as needed\n",
        "from itertools import islice\n",
        "#import sketch\n",
        "from collections import  defaultdict, Counter\n",
        "from prettytable import PrettyTable\n",
        "\n",
        "# Debugging:\n",
        "import traceback\n",
        "\n",
        "# File I/O and path handling:\n",
        "import os\n",
        "import copy, os\n",
        "import re\n",
        "import requests\n",
        "import json\n",
        "from datetime import datetime\n",
        "from typing import Dict, Any\n",
        "\n",
        "# Numerical analysis and statistics:\n",
        "import numpy as np\n",
        "from scipy import stats\n",
        "from statistics import mode, multimode  # Consider removing if unused\n",
        "\n",
        "# Third-party modules (Commented out modules can be imported as needed):\n",
        "#from Bio import SeqIO  # Only import if used\n",
        "# from icecream import ic  # Import on demand if needed\n",
        "\n",
        "# Google Colab specific for mounting Google Drive:\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wYc8zjTK0v1N"
      },
      "source": [
        "## Sequence dictionaries"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "file_path = '/content/drive/MyDrive/Colab_Notebooks/FASTA_dicts/sequences_dict.json'\n",
        "\n",
        "# Reading the JSON file and converting it back into a dictionary\n",
        "with open(file_path, 'r') as json_file:\n",
        "    sequences_dict = json.load(json_file)\n",
        "\n",
        "\n",
        "# Print only the first few items of the dictionary\n",
        "for key in list(sequences_dict.keys())[:10]:  # Adjust the number as needed\n",
        "    print(f\"{key}: {sequences_dict[key]}\")\n",
        "\n",
        "# # Or, if you want to check a specific entry\n",
        "# key_to_check = 'COL1A1-Bos_tau'\n",
        "# if key_to_check in sequences_dict:\n",
        "#     print(f\"{key_to_check}: {sequences_dict[key_to_check]}\")\n",
        "# else:\n",
        "#     print(f\"Key '{key_to_check}' not found in the dictionary.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EBa6tYVUnXY3",
        "outputId": "8329feca-b677-4819-acc0-d3b485d34fe5"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "COL1A1-Bos_tau: {'GNC': 'COL1A1', 'species': 'Bos taurus', 'offset': -7, 'sequence': 'STGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEGGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEK'}\n",
            "COL1A1-Bub_bub: {'GNC': 'COL1A1', 'species': 'Bubalus bubalis', 'offset': -7, 'sequence': 'STGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEGGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDSGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEVGPPGPPGPSGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEKAHDGGRYYRA'}\n",
            "COL1A1-Equ_asi: {'GNC': 'COL1A1', 'species': 'Equus asinus', 'offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGARGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGESGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGKEGGKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPVGPPGLAGPPGESGREGSPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGEAGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSAGFDFSFLPQPPQEKSHDGGR'}\n",
            "COL1A1-Equ_cab: {'GNC': 'COL1A1', 'species': 'Equus caballus', 'offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGARGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGESGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGKEGGKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPVGPPGLAGPPGESGREGSPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGEAGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKSHDGGR'}\n",
            "COL1A1-Hip_amp: {'GNC': 'COL1A1', 'species': 'Hippopotamus amphibius', 'offset': 0, 'sequence': 'GPMGPSGPRGIPGPPGAPGPQGF------PGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPSGPQGARGIPGTAGIPGMKGHRGFSGIDGAKGDAGPAGPK---------GAPGQMGPRGIPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAA---------------GANGAPGIAGAPGFPGARGPSGPQGPSGPSGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGIPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGIPGAKGITGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGIPGPAGPPGEAGKPGEQGVPGDIGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGIQGMPGERGAAGIPGPKGDRGDAGPKGADGSPGKDGVRGITGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAK----------GTGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGK------RGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGIPGQRGERGFPGIPGPSGEPGKQGPSGTSGERGPPGPMGPPGIAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGAAGPQGPRGDKGETGEQGDRGIKGHRGFSGIQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGINGIPGPIGPPGPR-----------PGPPGPPGPPGPP'}\n",
            "COL1A1-Ovi_Cap: {'GNC': 'COL1A1', 'species': 'Ovis Capra', 'offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGSDGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRAGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGAPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGTPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEKAHDGGR'}\n",
            "COL1A1-Pog_vit: {'GNC': 'COL1A1', 'species': 'Pogona vitticeps', 'offset': 0, 'sequence': 'GPMGPSGPRGPPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDPGPAGPKGEPGSPGENGAPGQVGPRGLPGERGRPGAPGPAGARGNDGSPGAAGPPGPTGPAGPPGFPGAVGPKGETGAQGSRGSEGPQGARGEPGPPGPAGAAGPSGNPGTDGQPGAKGSPGAPGIAGAPGFPGARGPAGPQGPAGAPGPKGNSGEPGAPGNKGDTGAKGETGPAGVQGPPGPPGEEGKRGSRGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGPAGPKGSTGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPAGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGAVGAVGAPGKDGEVGAQGPPGPTGPAGERGEQGPSGAPGFQGLPGPAGAPGEAGKPGEQGVPGDVGAPGPAGARGERGFPGERGVQGAAGPPGPRGANGSPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDTGPKGLDGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPQGPAGPTGARGAPGDRGEPGPAGPAGFAGPPGTDGQPGAKGEPGDAGAKGDAGPPGPAGATGPAGPAGPIGAPGPKGARGSPGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPVGKEGAKGPRGETGPAGRPGEAGPAGPPGPPGEKGSPGSDGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGEAGREGSPGAEGAPGRDGPAGPKGDRGETGPAGPPGAPGAPGAPGPMGPAGKNGDRGETGPAGPAGPAGPAGARGAAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDVGPAGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGR'}\n",
            "COL1A1-Alg_sin: {'GNC': 'COL1A1', 'species': 'Alligator sinensis', 'offset': 0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGAVGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGQPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGDVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQ'}\n",
            "COL1A1-Lox_afr: {'GNC': 'COL1A1', 'species': 'Loxodonta africana', 'offset': -8, 'sequence': 'SAGGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPVSPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPAGPQGPSGAPGPKGNSGEPGAPGSKGDAGAKGEPGPVGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGAAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGPRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGAPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGGKGPRGETGPAGRPGEVGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIGGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSSGERGPPGPAGPPGLAGPPGESGREGAPGAEGSPGRDGSPGPKGDRGETGPSGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPAGPAGVRGPAGPQGPRGDKGETGEQGDRGLKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGAFDFSFLPQPPQEKAHDGGRYYRA'}\n",
            "COL1A1-Sus_Scr: {'GNC': 'COL1A1', 'species': 'Sus scrofa', 'offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGGKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGVQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGPSGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGAPGPKGDRGESGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPVGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGGR'}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 50,
      "metadata": {
        "id": "0ZfuPxSWGpdG"
      },
      "outputs": [],
      "source": [
        "# # non_sequences_dict = {\n",
        "# #     'TRYUP-Sus_sco': {'species': 'Bos taurus','offset': 0, 'sequence': 'IVGGYTCAANSIPYQVSLNSGSHFCGGSLINSQWVVSAAHCYKSRIQVRLGEHNIDVLEGNEQFINAAKIITHPNFNGNTLDNDIMLIKLSSPATLNSRVATVSLPRSCAAAGTECLISGWGNTKSSGSSYPSLLQCLKAPVLSDSSCKSSYPGQITGNMICVGFLEGGKDSCQGDSGGPVVCNGQLQGIVSWGYGCAQKNKPGVYTKVCNYVNWIQQTIAAN'},\n",
        "# #     'AHSG-Bos_tau' : {'species': 'Bos taurus','offset': 0, 'sequence':'IPLDPVAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTQHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNDSRVVHAVEVALATFNAESNGSYLQLVEISRAQFVPLPVSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVRVTCTLFQTQPVIPQPQPDGAEAEAPSAVPDAAGPTPSAAGPPVASVVVGPSVVAVPLPLHRAHYDLRHTFSGVASVESSSGEAFHVGKTPIVGQPSIPGGPVRLCPGRIRYFKI'},\n",
        "# #     'AHSG-Bub_bub' : {'species': 'Bos taurus','offset': 0, 'sequence':'IPLDPVAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNDSRVVHAVEVALATFNAQSNGSYLQLVEISRAQFVPLPASVSVEFAVAATDCIAKDVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVAVTCTLFQTQPVILQPQPDGAEAGAPSAVPDAAGPAPSAAGPPVASVVVGPSVVAVPLPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVRLCPGRIRYFKI'},\n",
        "# #     'AHSG-Equ_cab' : {'species': 'Bos taurus','offset': 1, 'sequence':' LPNGLSPAYRQLNCDDPETEQAALLAVDYINSHIHQGYKHVLNQIDKVQVWAQPTGESFKLEIDTLETTCHALDPTPLANCSVRQLTQHAVEGDCDVRLLKQNGQFSVSFVKCKSSPDSAEDVRKVCLDCPLLAPLNDTRVVHAVEAALAAFNAQNNGSYFQLVEISRAQLVPLPVSVHVEFAVAATDCVAKEVIDPAKCNLLAEKQYGFCKATLTEKVGGEDVAVTCTVFQTQPVVLLPQPDGPDVGVPGPVADAVTPAPSPADLPVASLVVGPVVVAASQLPPPVHRAHYDLRHAFAGVGSGESASGEAFHVEKPPKVAHPNTAAAAGPVVRPCPGRIRYFKII'},\n",
        "# #     'AHSG-Ovi_are' : {'species': 'Bos taurus','offset': 1, 'sequence':'IPLDPIAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLVNCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNNSQVVHAAEVALATFNAQNNGSYFQLVEISRAQFVPLPGSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVTVTCTLFQTQPVIPQPQPEGAEAGAPSAVPDAAVPDAAVPAPSAAGLPVGSVVAGPSVVAVPLPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVHLCPGRIRYFKI'},\n",
        "# #     'AHSG-Cap_hir' : {'species': 'Bos taurus','offset': 1, 'sequence':'IPLDPIAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNNSQVVHAAEVALATFNAQNNGSYFQLVEISRAQFVPLPVSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVAVTCTLFQTQPVIPQPQPEGAEAGAPSAVPDAAVPAPSAAGLPVGSVVAGPSVVAVALPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVHLCPGRIRYFKI'},\n",
        "# #     'AHSG-Equ_asn' : {'species': 'Bos taurus','offset': 1, 'sequence':' LPNGLSPAYRQLNCDDPETEQAALLAVDYINSHIHQGYKHVLNQIDKVQVWAQPTGESFKLEIDTLETTCHALDPTPLANCSVRQLTQHAVEGDCDVRLLKQNGQFSVSFVKCKSSPDSAEDVRKVCLDCPLLAPLNDTRVVHAVEAALAAFNAQNNGSYFQLVEISRAQLVPLPVSVHVEFAVAATDCVAKEVIDPAKCNLLAEKQYGFCKATLTEKVGGEDVAVTCTVFQTQPVVPLPQPDGPDVGVPGPVADAATPEPSPADLPVASLVVGPVVVAAPQLPPPVHRAHYDLRHAFAGVGSGESASGEAFHVEKPPKVAHPNTAAAAGPVVRPCPGRIRYFKI'},\n",
        "# #     'TPM1-Cam_dro' : {'species': 'Bos taurus','offset': 0, 'sequence':'MDAIKKKMQMLKLDKENAIDRAEQAEADKKQAEDRCKQLEEEQQALQKKLKGTEDEVEKYSESVKDAQEKLEQAEKKATDAEADVASLNRRIQLVEEELDRAQERLATALQKLEEAEKAADESERGMKVIENRAMKDEEKMELQEMQLKEAKHIAEDSDRKYEEVARKLVILEGELERSEERAEVAESRARQLEEELRTMDQALKSLMASEEEEGLGLPNSICCHSHNFALLFSPPPLPHCAISTVSPTVSSHPPPPHTPCSKCGDLEEELKIVTNNLKSLEAQADKYSTKEDKYEEEIKLLEEKLKEAETRAEFAERSVAKLEKTIDDLEDEVYAQKMKYKAISEELDNALNDITSL'},\n",
        "# #     'KS1-Hom_sap' : {'species': 'Bos taurus','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTR'},\n",
        "# #     'KS9-Hom_sap' : {'species': 'Bos taurus','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTR'},\n",
        "# #     'KS2-Hom_sap' : {'species': 'Bos taurus','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTRR'},\n",
        "# #     'ALB-Bos_tau' : {'species': 'Bos taurus','offset': 0, 'sequence':'MKWVTFISLLLLFSSAYSRGVFRRDTHKSEIAHRFKDLGEEHFKGLVLIAFSQYLQQCPFDEHVKLVNELTEFAKTCVADESHAGCEKSLHTLFGDELCKVASLRETYGDMADCCEKQEPERNECFLSHKDDSPDLPKLKPDPNTLCDEFKADEKKFWGKYLYEIARRHPYFYAPELLYYANKYNGVFQECCQAEDKGACLLPKIETMREKVLASSARQRLRCASIQKFGERALKAWSVARLSQKFPKAEFVEVTKLVTDLTKVHKECCHGDLLECADDRADLAKYICDNQDTISSKLKECCDKPLLEKSHCIAEVEKDAIPENLPPLTADFAEDKDVCKNYQEAKDAFLGSFLYEYSRRHPEYAVSVLLRLAKEYEATLEECCAKDDPHACYSTVFDKLKHLVDEPQNLIKQNCDQFEKLGEYGFQNALIVRYTRKVPQVSTPTLVEVSRSLGKVGTRCCTKPESERMPCTEDYLSLILNRLCVLHEKTPVSEKVTKCCTESLVNRRPCFSALTPDETYVPKAFDEKLFTFHADICTLPDTEKQIKKQTALVELLKHKPKATEEQLKTVMENFVAFVDKCCAADDKEACFAVEGPKLVVSTQTALA'},\n",
        "# #     'ALB-Equ_cab' : {'species': 'Bos taurus','offset': 0, 'sequence':'MKWVTFVSLLFLFSSAYSRGVLRRDTHKSEIAHRFNDLGEKHFKGLVLVAFSQYLQQCPFEDHVKLVNEVTEFAKKCAADESAENCDKSLHTLFGDKLCTVATLRATYGELADCCEKQEPERNECFLTHKDDHPNLPKLKPEPDAQCAAFQEDPDKFLGKYLYEVARRHPYFYGPELLFHAEEYKADFTECCPADDKLACLIPKLDALKERILLSSAKERLKCSSFQNFGERAVKAWSVARLSQKFPKADFAEVSKIVTDLTKVHKECCHGDLLECADDRADLAKYICEHQDSISGKLKACCDKPLLQKSHCIAEVKEDDLPSDLPALAADFAEDKEICKHYKDAKDVFLGTFLYEYSRRHPDYSVSLLLRIAKTYEATLEKCCAEADPPACYRTVFDQFTPLVEEPKSLVKKNCDLFEEVGEYDFQNALIVRYTKKAPQVSTPTLVEIGRTLGKVGSRCCKLPESERLPCSENHLALALNRLCVLHEKTPVSEKITKCCTDSLAERRPCFSALELDEGYVPKEFKAETFTFHADICTLPEDEKQIKKQSALAELVKHKPKATKEQLKTVLGNFSAFVAKCCGREDKEACFAEEGPKLVASSQLALA'},\n",
        "# #     'ALB-Equ_asn' : {'species': 'Bos taurus','offset': 0, 'sequence':'MKWVTFVSLLFLFSSAYFRGVLRRDTHKSEIAHRFNDLGEKHFKGLVLVAFSQYLQQCPFEDHVKLVNEVTEFAKKCAADESAENCDKSLHTLFGDKLCTVATLRATYGELADCCEKQEPERNECFLTHKDDHPNLPKLKPEPDAQCAAFQEDPDKFLGKYLYEVARRHPYFYGPELLFHAEEYKADFTECCPADDKAGCLIPKLDALKERILLSSAKERLKCSSFQKFGERAFKAWSVARLSQKFPKADFAEVSKIVTDLTKVHKECCHGDLLECADDRADLTKYICEHQDSISGKLKACCDKPLLQKSHCIAEVKEDDLPSDLPALAADFAEDKEICKHYKDAKDVFLGTFLYEYSRRHPDYSVSLLLRIAKTYEATLEKCCAEADPPACYATVFDQFTPLVEEPKSLVKKNCDLFEEVGEYDFQNALIVRYTKKAPQVSTPTLVEIGRTLGKVGSRCCKLPESERLPCSENHLALALNRLCVLHEKTPVSEKITKCCTDSLAERRPCFSALELDEGYIPKEFKAETFTFHADICTLPEDEKQIKKQSALAELVKHKPKATKEQLKTVLGNFSAFVAKCCGAEDKEACFAEEGPKLVASSQLALA'},\n",
        "# #     'ALB-Sus_scr' : {'species': 'Bos taurus','offset': 0, 'sequence':'MKWVTFISLLFLFSSAYSRGVFRRDTYKSEIAHRFKDLGEQYFKGLVLIAFSQHLQQCPYEEHVKLVREVTEFAKTCVADESAENCDKSIHTLFGDKLCAIPSLREHYGDLADCCEKEEPERNECFLQHKNDNPDIPKLKPDPVALCADFQEDEQKFWGKYLYEIARRHPYFYAPELLYYAIIYKDVFSECCQAADKAACLLPKIEHLREKVLTSAAKQRLKCASIQKFGERAFKAWSLARLSQRFPKADFTEISKIVTDLAKVHKECCHGDLLECADDRADLAKYICENQDTISTKLKECCDKPLLEKSHCIAEAKRDELPADLNPLEHDFVEDKEVCKNYKEAKHVFLGTFLYEYSRRHPDYSVSLLLRIAKIYEATLEDCCAKEDPPACYATVFDKFQPLVDEPKNLIKQNCELFEKLGEYGFQNALIVRYTKKVPQVSTPTLVEVARKLGLVGSRCCKRPEEERLSCAEDYLSLVLNRLCVLHEKTPVSEKVTKCCTESLVNRRPCFSALTPDETYKPKEFVEGTFTFHADLCTLPEDEKQIKKQTALVELLKHKPHATEEQLRTVLGNFAAFVQKCCAAPDHEACFAVEGPKFVIEIRGILA'},\n",
        "\n",
        "# #     }\n",
        "\n",
        "# sequences_dict = {\n",
        "#     'COL1A1-Bos_tau': {'species': 'Bos taurus','offset': -7, 'sequence': 'STGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEGGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEK'},\n",
        "#     'COL1A1-Bub_bub': {'species': 'Bubalus bubalis','offset': -7, 'sequence': 'STGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEGGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDSGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEVGPPGPPGPSGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEKAHDGGRYYRA'},\n",
        "#     'COL1A1-Equ_asi': {'species': 'Equus asinus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGARGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGESGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGKEGGKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPVGPPGLAGPPGESGREGSPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGEAGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSAGFDFSFLPQPPQEKSHDGGR'},\n",
        "#     'COL1A1-Equ_cab': {'species': 'Equus caballus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGARGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGESGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGKEGGKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPVGPPGLAGPPGESGREGSPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGEAGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKSHDGGR'},\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "#     'COL1A1-Hip_amp': {'species': 'Hippopotamus amphibius','offset': 0, 'sequence': 'GPMGPSGPRGIPGPPGAPGPQGF------PGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPSGPQGARGIPGTAGIPGMKGHRGFSGIDGAKGDAGPAGPK---------GAPGQMGPRGIPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAA---------------GANGAPGIAGAPGFPGARGPSGPQGPSGPSGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGIPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGIPGAKGITGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGIPGPAGPPGEAGKPGEQGVPGDIGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGIQGMPGERGAAGIPGPKGDRGDAGPKGADGSPGKDGVRGITGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAK----------GTGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGK------RGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGIPGQRGERGFPGIPGPSGEPGKQGPSGTSGERGPPGPMGPPGIAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGAAGPQGPRGDKGETGEQGDRGIKGHRGFSGIQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGINGIPGPIGPPGPR-----------PGPPGPPGPPGPP'},\n",
        "#     'COL1A1-Ovi_Cap': {'species': 'Ovis Capra','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGSDGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRAGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGAPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGTPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEKAHDGGR'},\n",
        "#     'COL1A1-Pog_vit': {'species': 'Pogona vitticeps','offset': -0, 'sequence': 'GPMGPSGPRGPPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDPGPAGPKGEPGSPGENGAPGQVGPRGLPGERGRPGAPGPAGARGNDGSPGAAGPPGPTGPAGPPGFPGAVGPKGETGAQGSRGSEGPQGARGEPGPPGPAGAAGPSGNPGTDGQPGAKGSPGAPGIAGAPGFPGARGPAGPQGPAGAPGPKGNSGEPGAPGNKGDTGAKGETGPAGVQGPPGPPGEEGKRGSRGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGPAGPKGSTGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPAGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGAVGAVGAPGKDGEVGAQGPPGPTGPAGERGEQGPSGAPGFQGLPGPAGAPGEAGKPGEQGVPGDVGAPGPAGARGERGFPGERGVQGAAGPPGPRGANGSPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDTGPKGLDGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPQGPAGPTGARGAPGDRGEPGPAGPAGFAGPPGTDGQPGAKGEPGDAGAKGDAGPPGPAGATGPAGPAGPIGAPGPKGARGSPGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPVGKEGAKGPRGETGPAGRPGEAGPAGPPGPPGEKGSPGSDGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGEAGREGSPGAEGAPGRDGPAGPKGDRGETGPAGPPGAPGAPGAPGPMGPAGKNGDRGETGPAGPAGPAGPAGARGAAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDVGPAGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGR'},\n",
        "#     'COL1A1-Alg_sin': {'species': 'Alligator sinensis','offset': -0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGAVGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGQPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGDVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQ'},\n",
        "#     'COL1A1-Lox_afr': {'species': 'Loxodonta africana','offset': -8, 'sequence': 'SAGGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPVSPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPAGPQGPSGAPGPKGNSGEPGAPGSKGDAGAKGEPGPVGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGAAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGPRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGAPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGGKGPRGETGPAGRPGEVGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIGGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSSGERGPPGPAGPPGLAGPPGESGREGAPGAEGSPGRDGSPGPKGDRGETGPSGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPAGPAGVRGPAGPQGPRGDKGETGEQGDRGLKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGAFDFSFLPQPPQEKAHDGGRYYRA'},\n",
        "#     'COL1A1-Sus_Scr': {'species': 'Sus scrofa','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGGKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGVQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEAGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGPSGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGAPGPKGDRGESGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPVGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGGR'},\n",
        "#     'COL1A1-Mam_mam': {'species': 'Mammuthus primigenius','offset': -8, 'sequence': 'SAGGISVPGPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPAGPQGPSGAPGPKGNSGEPGAPGSKGDAGAKGEPGPIGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGAAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGSNGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGPRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGAPGPIGNVGAPGAKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGGKGPRGETGPAGRPGEVGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIGGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSSGERGPPGPAGPPGLAGPPGESGREGAPGAEGSPGRDGSPGPKGDRGETGPSGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPAGPAGVRGPAGPQGPRGDKGETGEQGDRGLKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPPGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGAFDFSFLPQPPQEK'},\n",
        "#     'COL1A1-Tur_tur': {'species': 'Tursiops truncatus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGTPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGSTGPAGPPGFPGAVGAKGEAGPQGSRGSEGPQGVRGEPGPPGPAGATGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGLAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGSRGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAAGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPSGPPGEAGKPGEQGAPGDIGAPGPSGARGERGFPGERGVQGPSGPAGPRGSNGPPGNDGAKGDAGAPGAPGNQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGAPGSAGPTGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPSGPAGKEGSKGPRGETGPAGRAGEVGPPGPPGPAGEKGAPGADGPAGSPGSPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGSPGAPGPVGPAGKSGDRGETGPAGPAGPIGPAGARGPTGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGTPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDFSFLPQPPQEKAQDGGR'},\n",
        "\n",
        "\n",
        "#     'COL1A2-Lox_afr': {'species': 'Loxodonta africana','offset': -5, 'sequence': 'KGIGLGPGPMGLMGPRGPPGATGPPGSPGFQGPPGEPGEPGQTGPAGSRGPAGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQIGARGLPGERGRVGGPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGEIGPVGNPGPSGPAGPRGEAGLPGVSGPVGPPGNPGANGLAGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGSKGEPGSAGPQGPPGPSGEEGKRGSSGEAGSAGPAGPPGLRGGPGSRGLPGADGRAGVMGPPGSRGASGPAGVRGPSGDSGRPGEPGVMGPRGLPGSPGNVGPAGKEGPAGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPAGDPGKNGDKGHAGLAGPRGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPSGTAGEAGKPGERGIPGEFGLPGPAGPRGERGPPGQSGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPGGLPGERGAAGIPGGKGEKGETGLRGDTGNTGRDGARGAPGAVGAPGPAGATGDRGEAGPAGSAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQAGAKGERGTKGPKGENGPVGPTGPVGAAGPAGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPAGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGASGPPGFAGEKGSSGEPGTAGPPGTPGPQGILGPPGILGLPGSRGERGLPGVAGAVGEPGPLGIAGPPGARGPPGAVGSPGVNGAPGEAGRDGNPGSDGPPGRDGLPGHKGERGYPGNAGPVGTAGAPGPQGPLGPAGKHGNRGEPGPAGSVGPVGAVGPRGPSGPQGARGDKGEAGDKGPRGLPGFKGHNGLQGLPGLAGQHGDQGSPGSVGPAGPRGPAGPSGPVGKDGRPGHAGAVGPAGVRGSQGSQGPSGPPGPPGPPGPPGPSGGGYDFGYDGDFYRA'},\n",
        "#     'COL1A2-Bos_mut': {'species': 'Bubalus bubalis','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGASGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A2-Bos_tau': {'species': 'Bos taurus','offset': -4, 'sequence': 'GGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYR'},\n",
        "#     'COL1A2-Equ_asi': {'species': 'Equus asinus','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHKGLDGLKGQPGAPGVKGEPGAPGENGTPGQAGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPAGAAGATGARGLVGEPGPAGSKGESGNKGEPGAAGPQGPPGPSGEEGKRGPNGEPGSTGPAGPPGLRGSPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGEPGKPGDKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGARGERGPPGESGAAGPAGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGEIGNPGRDGARGAPGAVGAPGPAGANGDRGEAGAAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGVTGFPGAAGRTGPPGPSGISGPPGPPGAAGKEGLRGPRGDQGPVGRAGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVAGSLGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGSDGPPGRDGQPGHKGERGYPGNAGPVGAVGAPGPHGPVGPTGKHGNRGEPGPVGSVGPVGAVGPRGPSGPQGVRGDKGEPGDKGPRGLPGIKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPTGPVGKDGRSGQPGTVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYDGDFYRA'},\n",
        "#     'COL1A2-EquxPrz': {'species': 'EquusxPrezwal','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQAGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPAGAAGATGARGLVGEPGPAGSKGESGNKGEPGAAGPQGPPGPSGEEGKRGPNGEPGSTGPAGPPGLRGSPGSRGLPGADGRAGVMGPAGSRGASGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGEPGKPGDKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGARGERGPPGESGAAGPAGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGEIGNPGRDGARGAPGAVGAPGPAGANGDRGEAGAAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGVTGFPGAAGRTGPPGPSGISGPPGPPGAAGKEGLRGPRGDQGPVGRAGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVAGSLGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGSDGPPGRDGQPGHKGERGYPGNAGPVGAVGAPGPHGPVGPTGKHGHRGEPGPVGSVGPVGAVGPRGPSGPQGVRGDKGEPGDKGPRGLPGIKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPTGPVGKDGRSGQPGTVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYDGDFYRA'},\n",
        "#     'COL1A2-Equ_cab': {'species': 'Equus caballus','offset': 0, 'sequence': '---------GPMGIMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGAR--------AGEDGHPGKPGRPGERGVVGPQGARGFPGTPGIPGFK --GHNGIDGIKGQPGAPGVKGEPGAPGENGTPGQAGARGIPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGEIGPVGNPGPAGPAGPRGEVGIPGISGPVGPPGNPGANGITGAKGAAGIPGVAGAPGIPGPRGIPGPAGAAGATGARGIVGEPGPAGSKGESGNKGEPGAAGPQGPPGPSGEEGKRGPNGEPGSTGPAGPPGIR------GIPGADGRAGVMGPAGSRGASGPAGVRGPNGDSGRPGEPGIMGPRGFPGSPGNIGPAGKEGPVGIPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGEPGKPGDKGDAGIAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGIPGPAGTAGEVGKPGERGIPGEFGIPGPAGAR---GPPGESGAAGPAGPIGSRGPSGPPGPDGNKGEPGVIGAPGTAGPSGPSGIPGERGAAGIPGGK---------GEIGNPGRDGARGAPGAVGAPGPAGANGDRGEAGAAGPAGPAGPR------GEVGPAGPNGFAGPAGAAGQPGAKGER---GPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGVTGFPGAAGRTGPPGPSGISGPPGPPGAAGK-------GDQGPVGRAGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGIIGAPGIIGIPGSR---GIPGVAGSIGEPGPIGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGSDGPPGR----------GYPGNAGPVGAVGAPGPHGPVGPTGK---RGEPGPVGSVGPVGAVGPRGPSGPQGVR------------------GHNGIQGIPGIAGQHGDQGAPGSVGPAGPRGPAGPTGPVGKD--SGQPGTVGPAGVRGSQ'},\n",
        "#     'COL1A2-Cap_hir': {'species': 'Capra hircus','offset': -7, 'sequence': 'QFDGKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPAGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKAGEKGHAGLAGPRGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDVGSPGRDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPAGISGPPGPPGPAGKEGLRGPRGDQGPVGRTGEPGAAGPPGFVGEKGPSGEPGTAGPPGTPGPQGFLGPPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPTGKHGSRGEPGPVGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPTGPAGKDGRTGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGFDGDFYRA'},\n",
        "#     'COL1A2-Ovi_are': {'species': 'Ovis aries','offset': -7, 'sequence': 'QFDGKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPAGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKAGEKGHAGLAGPRGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDVGSPGRDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPAGISGPPGPPGPAGKEGLRGPRGDQGPVGRTGEPGAAGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPTGKHGSRGEPGPVGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPTGPAGKDGRTGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGFDGDFYRA'},\n",
        "#     'COL1A2-Bub_bub': {'species': 'Bubalus bubalis','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGPAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPSGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGVRGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPAGISGPPGPPGPAGKEGLRGPRGDQGPVGRTGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGLPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGFDGDFYRA'},\n",
        "\n",
        "#     'COL1A2-Pog_vit': {'species': 'Pogona vitticeps','offset': -12, 'sequence':'QYDISKSADYGPGPMGIMGPRGPPGPNGPPGSPGYQGPPGEPGEPGQTGPQGARGPMGPPGKAGEDGHPGKPGRPGERGPSGPQGARGFPGTPGLPGFKGIRGHNGLDGRKGEPGTAGNKGEPGAAGENGTPGQPGARGAPGERGRVGANGPVGARGADGSNGPTGPAGPIGAAGPPGFPGGPGAKGEIGAVGNPGPSGPAGPRGEAGLPGASGPVGPAGNPGTNGINGAKGAAGLPGAAGPPGLPGPRGILGAPGPSGPAGARGLTGEPGPAGSKGETGSKGEPGSAGPAGPQGPSGEEGKRGSTGEPGSSGLPGPAGLRGVPGSRGLPGSDGRAGVIGPAGNRGATGLPGAKGPNGDTGRPGEPGLMGARGLPGQSGSPGPAGKEGAVGLPGADGRAGPIGPAGPRGEPGNIGFPGPKGPNGEPGKPGDRGNVGVTGSRGAPGPDGNNGAQGAPGVIGNPGPKGDAGPAGPPGFQGLPGPSGPVGEAGKPGERGIHGEFGLPGPAGSRGERGPPGESGATGPVGPIGSRGPSGAPGPDGNKGEPGNVGAAGAPGPAGSNGIPGERGAAGMPGGKGEKGPTGFRGEPGATGRDGSRGIPGAVGAPGPAGGAGERGEAGPAGQAGPPGSRGIPGERGEVGPAGPQGFAGPPGAAGQPGAKGERGQRGPKGDVGPQGAIGAVGNAGPAGPLGAAGPAGTRGDSGPPGMTGAPGSAGRPGPPGPAGISGPPGPPGPAGKEGIRGPRGDTGPTGRSGETGISGPPGFQGEKGPSGESGAAGPPGTPGPQGILGAPGILGLPGSRGERGLPGIAGAVGEPGPLGIAGPPGARGAPGPIGSPGVNGAPGEAGRDGNPGNDGAPGRNGAPGYKGERGAPGSPGPVGAAGAPGAHGIVGPVGKTGNRGEPGPVGPAGSTGPAGPRGLPGPQGPRGEKGVVGDKGPRGLPGLKGHNGLQGLPGIAGHHGEQGPSGIPGPAGPRGPAGPSGPAGKDGRNGLPGPVGPAGVRGSQGSQGPSGPPGPPGLPGPAGPNGGGYEIGYDMEYYRA'},\n",
        "#     'COL1A2-Cro_por': {'species': 'Crocodylus porosus','offset': -12, 'sequence':'QYDGAKAGDYGQGPMGLMGPRGPPGTSGPPGPPGFQGPHGEPGEPGQTGAQGPRGPPGPPGKAGEDGHPGKSGRSGERGVSGPQGARGFPGTPGLPGFKGIRGHNGLDGQKGQAGTPGIKGESGAPGENGTPGQSGARGLPGERGRIGAPGPAGARGSDGSGGPTGPAGPIGAAGPPGFPGAPGAKGEIGAAGNVGPSGPAGPRGEAGLPGSSGPAGPPGNPGANGLAGAKGATGLPGVAGAPGLPGPRGIPGPPGPAGAAGTRGLVGEPGPAGAKGESGNKGEPGAAGPPGPAGPSGEEGKKGTTGEPGSSGPPGPAGLRGVPGSRGLPGADGRAGVMGPAGSRGATGPAGAKGPTGDNGRPGEPGLMGPRGLPGQPGSSGPAGKEGPVGFPGADGRVGPIGPAGARGEPGNIGFPGPKGPTGDPGKPGDRGHAGLAGARGAPGPEGNNGAQGPPGVAGNPGAKGEQGPAGPPGFQGLPGPSGPAGEAGKPGERGMAGEFGAPGPAGSRGERGPPGESGAVGPVGSIGSRGPSGPPGTDGNKGEPGNVGNAGTAGPSGASGAPGERGIAGIPGPKGEKGATGLRGDTGATGRDGARGAPGAIGAPGPAGGAGERGEGGPAGPAGPSGARGIPGERGEPGPAGPTGFAGPAGAAGQPGAKGERGPKGPKGESGPQGSVGPVGASGPSGPVGAAGPAGPRGDGGPPGVTGFPGAAGRTGPPGPSGITGPPGPSGPAGKDGVRGPRGDTGPVGRTGEQGIVGPPGFSGEKGPSGESGAAGPPGSPGPQGILGAPGVLGLPGSRGERGLPGISGATGEPGPLGIAGPSGARGPPGPVGSAGLNGAPGEAGRDGNPGHDGAPGRDGAPGFKGERGAPGNNGPAGAVGAPGAHGQVGPVGKPGNRGDPGPVGPSGPAGAFGARGLAGPQGARGEKGETGDKGHRGMPGFKGHNGLQGLPGLAGQHGDQGPPGSTGPAGPRGPSGPSGPAGKDGRNGLPGPIGPAGVRGSQGSQGPAGPPGPPGLPGPPGANGGGYEVGYDLEYYRA'},\n",
        "#     'COL1A2-Hip_amp': {'species': 'Hippopotamus amphibius','offset': 0, 'sequence': 'GPMGIMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRSGERGVVGPQGARGFPGTPGIPGFKGIRGHNGIDGIKGQPGAPGVKGEPGAP---------GARGIPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGEIGPVGSPGASGPAGPRGEVGIPGVSGPVGPPGN----------GAAGIPGVAGAPGIPGPRGIPGPVGAAGATGARGIVGEPGPAGSK------------GPQGIPGPSGEEGKRGSTGEIGPAGPPGPPGIRGSPGSRGIPGADGRAGVMGIPGSRGATGPAGVRG------------------FPGSPGNIGPAGKEGPVGIPGIDGRPGPTGPAGAR----NIGFPGPKGPTGD---NGDKGHAGIAGARGAPGPDGNNGAQGPP--Q---GGKGEQGPAGPPGFQGIPGPAGTAGEAGKPGERGIPGEFGIPGPAGPRGERGPPGESGAAGPTGPIGNRGPSGPAGPDGNKGEPGVVGAPGTAGPSGPSGIPGERGAAGIPGPKGEKGEPGIR-------RDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGER---GTRGDGGPPGATGFPGAAGR------------------------------TGPPGPSGISGPPGPPGPAGKEGIRGPRGDQGPVGRSGETGASGPPGFAGEK-------------TPGPQGIIGAPGFIGIPGSRGERGIPGVAGSVGEPGPIGIAGPPGARGPPGAVGNPGVNGAPGEAGRDGNPGSDGPPGR----GHKGERGYP-------GAGAPGPQGPVGPTGKHGNRGEPGPAGVVGPTGAVGPRGPSGPQGIRGDKGEPGDKGPRGIPGIKGHNGIQGIPGIAGHHGDQGAPGSVGPAGPRGPAGPSGPVGKDGRTGHPGAVGPAGIRGSQGPAGPPGPPGPPGPPGPS'},\n",
        "#     'COL1A2-Mel_mel': {'species': 'Meles meles','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSAGPQGPPGPSGEEGKRGPNGEAGSTGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNIGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKPGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPGGLPGERGAAGVPGGKGEKGETGLRGEVGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGAHGPPGFAGEKGPSGEAGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGPVGAVGAPGPHGPVGPTGKHGNRGEPGPAGSVGPVGAAGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRTGHPGTVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPVKNWYRNSKVKKHTWLGETINDGTQFEYNTEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLADGCSKKTNEWRKTIIEYKTNKPSRLPILDIAPLDIGGADQEFRVDVGPVCFK'},\n",
        "#     'COL1A2_Alu_mel': {'species': 'Ailuropoda melanoleuca','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSVGPQGPPGPSGEEGKRGPNGEAGSAGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGSTGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNVGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGEPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPGGLPGERGAAGVPGGKGEKGETGLRGEVGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGAHGPPGFAGEKGPSGEPGTAGPPGTAGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGPVGTVGAPGPHGPVGPTGKHGNRGEPGPAGAVGPVGAVGPRGPSGPQGVRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRTGHPGTVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPAKNWYRNSKVKKHIWLGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVMLQGSNDVELVAEGNSRFTYTVLVDGCSKKTNEWGKTIIEYKTNKPSRLPILDIAPLDIGGADQEFRVDVGPVCFK'},\n",
        "#     'COL1A2-Enh_lut': {'species': 'Enhydra lutris kenyoni','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSAGPQGPPGPSGEEGKRGPNGEAGSTGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNIGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKPGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPGGLPGERGSAGVPGGKGEKGETGLRGEVGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPSGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGAHGPPGFAGEKGPSGEPGTAGPPGTSGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGSVGAVGAPGPHGPVGPTGKHGNRGEPGPAGSVGPVGAAGPRGPSGPQGVRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSLGPAGPRGPAGPSGPAGKDGRTGHPGTVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPVKNWYRNSKVKKHTWLGETINGGTQFEYNTEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLADGCSKKTNEWRKTIIEYKTNKPSRLPILDIAPLDIGGADQEFRVDVGPVCFK'},\n",
        "#     'COL1A2-Cro_cro': {'species': 'Crocuta crocuta','offset': -7, 'sequence': '---------------------------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------GELGPVGNPGPAGPAGPRGEMGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSAGPQGPPGPSGEEGKRGPNGEAGSTGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGATGPAGVRGPNGDAGRPGEPGLMGPRGFPGAPGNVGPAGKEGPMGVPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGEPGKGGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGEPGLRGEIGSPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGVRGPRGDQGPVGRTGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGLAGPPGARGGPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGPVGAAGAPGPHGPVGPAGKHGN----GPAGVVGPVGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGSPGSVGPAGPRGPAGPSGPVGKDGRTGHPGSVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENILAKNWYRNSKVKKHVWLGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYNVLADGCSKKTNEWRKTIIEYKTNKPSRLPILDIAPSDIGGADQEIRVDVGPVCFK'},\n",
        "#     'COL1A2-Fel_sil': {'species': 'Felis silvestris catus','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEMGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSAGPQGPPGPSGEEGKRGPNGEAGSAGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGATGPAGVRGPNGDAGRPGEPGLMGPRGFPGAPGNVGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKNGDKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGEIGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGISGPPGARGPSGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGPVGAVGAPGPHGPVGPTGKHGNRGEPGPAGVVGPVGAVGPRGPTGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPSGPMGKDGRTGHPGSVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENILAKNWYRNSKVKKHVWLGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLVDGCSKKTNEWKKTIIEYKTNKPSRLPILDIAPLDIGGADQEIRVDVGPVCFK'},\n",
        "#     'COL1A2-Enh_lut': {'species': 'Enhydra lutris kenyoni','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSAGPQGPPGPSGEEGKRGPNGEAGSTGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNIGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKPGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPGGLPGERGSAGVPGGKGEKGETGLRGEVGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPSGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGAHGPPGFAGEKGPSGEPGTAGPPGTSGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGSVGAVGAPGPHGPVGPTGKHGNRGEPGPAGSVGPVGAAGPRGPSGPQGVRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSLGPAGPRGPAGPSGPAGKDGRTGHPGTVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPVKNWYRNSKVKKHTWLGETINGGTQFEYNTEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLADGCSKKTNEWRKTIIEYKTNKPSRLPILDIAPLDIGGADQEFRVDVGPVCFK'},\n",
        "#     'COL1A2-Ail_mel': {'species': 'Ailuropoda melanoleuca','offset': -7, 'sequence': 'QYDPGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGSVGPQGPPGPSGEEGKRGPNGEAGSAGPSGPPGLRGSPGSRGLPGADGRAGVMGPPGPRGSTGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNVGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGEPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPGGLPGERGAAGVPGGKGEKGETGLRGEVGNPGRDGARGAPGAVGAPGPAGATGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGAHGPPGFAGEKGPSGEPGTAGPPGTAGPQGLLGAPGILGLPGSRGERGLPGVSGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNIGPVGTVGAPGPHGPVGPTGKHGNRGEPGPAGAVGPVGAVGPRGPSGPQGVRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRTGHPGTVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPAKNWYRNSKVKKHIWLGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVMLQGSNDVELVAEGNSRFTYTVLVDGCSKKTNEWGKTIIEYKTNKPSRLPILDIAPLDIGGADQEFRVDVGPVCFK'},\n",
        "#     'COL1A2-Del_leu': {'species': 'Delphinapterus leucas','offset': -6, 'sequence': 'QYDGKGVGIGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGTPGVKGEPGAPGENGIPGQIGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPLGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGSRGEVGLPGVSGPVGPPGNPGANGLHGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGSKGEPGAAGPTGPPGPSGEEGKRGTTGEIGSAGPPGPPGLRGNPGSRGLPGADGRAGVMGPHGSRGGTGPAGVRGPSGDSGRPGEPGLMGPRGFPGSPGNVGPAGKEGPMGLPGIDGRPGPIGPAGTRGEPGNIGFPGPKGPTGDPGKNGEKGHAGLAGPRGAPGPDGNNGAQGPPGPQGVSGGKGEQGPAGPPGFQGLPGPAGTAGEAGKAGERGIPGEFGLPGPAGSRGERGPPGESGAAGPAGPIGSRGPSGPAGPDGNKGEPGVVGAPGTAGPSGPNGLPGERGAAGIPGGKGEKGETGLRGDVGSHGRDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPVGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPTGPTGPVGAVGPAGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGPAGKEGLRGPRGDQGPVGRTGETGASGPPGFVGEKGPSGEPGTAGSPGTPGPQGLLGVPGFLGLPGSRGERGLPGVAGSMGEPGPLGIAGPAGARGPPGAVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQAGHKGDRGYPGNAGPTGTAGAPGPQGPVGPTGKHGNRGEPGPSGPAGLAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGTVGPAGPRGPSGPSGPSGKDGRTGHPGAVGPAGIRGSQGSQGPSGPPGPPGPPGPPGPSGGGYDFGFEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWVDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPVKNWYRSSKAKKHVWVGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLVDGCSKKINEWRKTIIEYKTNKPSRLPILDIAPLDIGGADQEIRLNIGPVCFK'},\n",
        "#     'COL1A2-Can_lup': {'species': 'Canus Lupu','offset': -6, 'sequence': 'QYDGKGVGLGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGEIGPVGNPGPAGPAGPRGEVGLPGVSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGIVGEPGPAGSKGESGNKGEPGSAGAQGPPGPSGEEGKRGPNGEAGSAGPSGPPGLRGSPGSRGLPGADGPAGVMGPPGPRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGAPGNVGPAGKEGPMGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKNGDKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGPRGERGPPGESGAAGPSGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGASGPGGLPGERGAAGIPGGKGEKGETGLRGEIGNPGRDGARGAPGAMGAPGPAGATGDRGEAGPAGPAGPAGPRGTPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPIGSAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGAAGKEGLRGPRGDQGPVGRTGETGASGPPGFTGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGNDGPPGRDGQAGHKGERGYPGNIGPVGAVGAPGPHGPVGPTGKHGNRGEPGPAGSVGPVGAVGPRGPSGPQGIRGDKGEPGEKGPRGLPGLKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRTGQPGTVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYEGDFYRADQPRSPPSLRPKDYEVDATLKSLNNQIETLLTPEGSRKNPARTCRDLRLSHPEWSSGYYWIDPNQGCTMDAIKVYCDFSTGETCIRAQPENIPAKNWYRNSKVKKHIWLGETINGGTQFEYNVEGVTTKEMATQLAFMRLLANHASQNITYHCKNSIAYMDEETGNLKKAVILQGSNDVELVAEGNSRFTYTVLVDGCSKKTNEWRKTIIEYKTNKPSRLPILDIAPLDIGDADQEFRVDVGPVCFK'},\n",
        "\n",
        "#     'COL3A1-Equ_cab': {'species': 'Equus caballus','offset': -17, 'sequence': 'QYESYDVKAGVAGGGLAGYPGPVGPPGPPGPPGASGHPGSPGSPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGLPGPPGIKGPAGIPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGLPGENGAPGPMGPRGAPGERGRPGLPGAAGSRGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGNAGAPGPPGPPGASGSPGGKGEMGPAGIPGAPGLIGARGPPGPSGANGAPGQRGPAGEPGKNGAKGEPGARGERGEAGSPGIPGAKGEDGKDGSPGEPGANGLPGAAGERGAPGFRGPAGPNGLPGEKGAPGDRGSPGPAGPRGAAGEPGRDGVPGGPGIRGVPGSPGGPGSDGKPGPPGSQGESGRPGPPGPSGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPPGPAGKNGETGPQGPPGPTGPSGDKGDAGLPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGSPGIPGGKGDSGAPGERGPPGAAGPPGLRGGAGPPGPEGGKGPAGAPGPPGAAGTPGLQGMPGERGGPGGPGVKGDKGEPGSAGADGAPGKDGPRGPTGPIGPPGPAGQPGDKGEGGAPGLPGIAGPRGGPGERGEVGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGAPGIAGPPGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGARGLPGPPGSNGNPGPPGPSGAPGKDGPAGPPGSSGSPGSPGVSGPKGDAGQPGEKGSPGPQGPPGVPGPSGLIGITGARGLAGPPGMPGARGSPGPQGVKGESGKPGAPGHNGERGPPGPQGLPGLAGTAGEPGRDGNPGSDGLPGRDGAPGGKGDRGENGSPGAPGAPGHPGPPGPVGPAGKNGDRGESGPAGPAGAPGPAGARGAPGPQGPRGDKGETGERGSTGIKGHRGFPGSPGSPGSPGPAGHQGAVGSPGPAGPRGPAGPSGPPGKDGTLGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGVSAISGVIGPEKSGGYAPYYG'},\n",
        "#     'COL3A1-Equ_asi': {'species': 'Equus asinus','offset': -17, 'sequence': 'QYESYDVKAGVAGGGLAGYPGPVGPPGPPGPPGASGHPGSPGSPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGLPGPPGIKGPAGIPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGLPGENGAPGPMGPRGAPGERGRPGLPGAAGSRGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGNAGAPGPPGPPGASGSPGGKGEMGPAGIPGAPGLIGARGPPGPSGANGAPGQRGPAGEPGKNGAKGEPGARGERGEAGSPGIPGAKGEDGKDGSPGEPGANGLPGAAGERGAPGFRGPAGPNGLPGEKGAPGDRGSPGPAGPRGAAGEPGRDGVPGGPGIRGVPGSPGGPGSDGKPGPPGSQGESGRPGPPGPSGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPPGPAGKNGETGPQGPPGPTGPSGDKGDAGLPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGSPGIPGGKGDSGAPGERGPPGAAGPPGLRGGAGPPGPEGGKGPAGAPGPPGAAGTPGLQGMPGERGGPGGPGVKGDKGEPGSAGADGAPGKDGPRGPTGPIGPPGPAGQPGDKGEGGAPGLPGIAGPRGGPGERGEVGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGAPGIAGPPGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGARGLPGPPGSNGNPGPPGPSGAPGKDGPAGPPGSSGSPGSPGVSGPKGDAGQPGEKGSPGPQGPPGVPGPSGLIGITGARGLAGPPGMPGARGSPGPQGVKGESGKPGAPGHNGERGPPGPQGLPGLAGTAGEPGRDGNPGSDGLPGRDGAPGGKGDRGENGSPGAPGAPGHPGPPGPVGPAGKNGDRGESGPAGPAGAPGPAGARGAPGPQGPRGDKGETGERGSTGIKGHRGFPGSPGSPGSPGPAGHQGAVGSPGPAGPRGPAGPSGPPGKDGTLGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGVSAISGVVGPEKSGGYAPYYG'},\n",
        "#     'COL3A1-Bos_tau': {'species': 'Bos taurus','offset': -9, 'sequence': ' SGVAGGGIAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGFPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGVPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLIGARGPPGPPGTNGVPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIAGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGDRGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAGGPPGPRGGAGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGVPGIAGPRGGPGERGEQGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPAGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGGRGPPGPPGSNGNPGPPGSSGAPGKDGPPGPPGSNGAPGSPGISGPKGDSGPPGERGAPGPQGPPGAPGPLGIAGLTGARGLAGPPGMPGARGSPGPQGIKGENGKPGPSGQNGERGPPGPQGLPGLAGTAGEPGRDGNPGSDGLPGRDGAPGAKGDRGENGSPGAPGAPGHPGPPGPVGPAGKSGDRGETGPAGPSGAPGPAGSRGPPGPQGPRGDKGETGERGAMGIKGHRGFPGNPGAPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGASGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGVAAIAGVGAEK'},\n",
        "#     'COL3A1-Bos_ind': {'species': 'Bos indicus','offset': -17, 'sequence': 'QYEAYDVKSGXAGGGIAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGFPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGVPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLIGARGPPGPPGTNGVPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIAGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGDRGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAGGPPGPRGGAGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGVPGIAGPRGGPGERGEQGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPAGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGGRGPPGPPGSNGNPGPPGSSGAPGKDGPPGPPGSNGAPGSPGISGPKGDSGPPGERGAPGPQGPPGAPGPLGIAGLTGARGLAGPPGMPGARGSPGPQGIKGENGKPGPSGQNGERGPPGPQGLPGLAGTAGEPGRDGNPGSDGLPGRDGAPGAKGDRGENGSPGAPGAPGHPGPPGPVGPAGKSGDRGETGPAGPSGAPGPAGSRGPPGPQGPRGDKGETGERGAMGIKGHRGFPGNPGAPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGASGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGVAAIAGVGAEKAGGFAPYYG'},\n",
        "#     'COL3A1-Bos_mut' : {'species': 'Bos mutus','offset': -17, 'sequence': 'QYEAYDVKSGIAGGGIAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGFPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGVPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLIGARGPPGPPGTNGVPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIAGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGDRGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAGGPPGPRGGAGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGVPGIAGPRGGPGERGEQGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEG------------------------------------------------------------------------------------------------------GPPGAAG---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------PPGKDGASGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGVAAIAGVGAEKAGGFAPYYG'},\n",
        "#     'COL3A1-Bub_bub' : {'species': 'Bubalus bubalis','offset': -17, 'sequence': 'QYEAYDVKSGIAGGGIAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGFPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGIPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLIGARGPPGPSGTNGVPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIPGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGERGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGESGAPGTPGGKGDSGAPGERGPPGAGGPPGPRGGPGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGVPGIAGPRGGPGERGEQGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPAGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGGRGPPGPPGSNGNPGPPGSSGAPGKDGPPGPPGSNGAPGSPGISGPKGDAGPPGERGAPGTQGPPGAPGPLGIAGVAGARGLAGPPGMPGARGSPGPQGIKGENGKPGPSGQNGERGPPGPQGLPGLAGAAGEPGRDGNPGSDGLPGRDGAPGTKGDRGENGSPGAPGAPGHPGPAGPVGPPGKSGDRGETGPAGPSGAPGPAGSRGPPGPQGPRGDKGETGERGATGIKGHRGFPGNPGAPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGTSGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGVAAIAGVGAEKAGGFAPYYG'},\n",
        "#     'COL3A1-Lox_afr' : {'species': 'Loxodonta africana','offset': -17, 'sequence': 'QYDSYDVKAGGAGLVGGYPGPAGPPGPPGPPGTSGHPGSPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPAGPAGKDGESGRPGRTGERGLTGPPGLKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGLPGENGSPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGLPGPPGPPGTSGFPGSPGAKGEVGPAGSPGSNGSPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLLGARGPPGPPGTNGIPGLRGSAGEPGKNGAKGDTGPRGERGEAGSPGIPGPKGEDGKDGSPGEPGANGSPGTAGERGAPGFRGPAGANGLPGEKGPSGERGGPGPAGPRGVPGEPGRDGLPGGPGMRGVPGSPGGPGSDGKPGPPGSQGESGRPGPPGPSGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGASGDKGESGPPGPPGIQGLPGTGGPPGENGKPGEPGPKGDAGAPGTPGGKGDSGAPGERGPPGQAGPPGVRGGAGPPGPEGGKGAPGPPGPPGTSGSPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPVGPIGPPGPPGQNGDKGESGAPGLPGPPGARGGPGERGEHGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPSGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGARGLPGPPGSSGNPGPPGSSGAPGKDGPPGPPGSNGSPGTPGVSGPKGDSGQPGEKGPPGPQGPPGAPGPLGLAGTTGSRGLAGLPGMPGARGSPGPQGPKGENGKPGPSGQNGERGPPGAQGLPGLAGAAGEPGRDGNPGSDGLPGRDGSPGTKGDRGENGSPGAPGAPGHPGPAGPVGPAGKSGDRGETGPAGPAGAPGPAGARGPAGAPGPRGDKGETGERGATGIKGHRGFPGNPGSPGSPGPAGHQGAIGSPGPAGPRGPVGPSGPPGKDGTSGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGGVGAVGLGAVGEKSGGYSPYYG'},\n",
        "#     'COL3A1-Sus_scr': {'species': 'Sus scrofa','offset': -17, 'sequence': 'QYESYDVKAGVAGGGIGGYPGPAGPPGPPGPPGVSGHPGAPGSPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGLPGPPGLKGPAGMPGFPGMKGHRGFDGRNGEKGDTGAPGLKGENGLPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGPSGSPGQRGEPGPQGHAGAAGPPGPPGSNGSPGGKGEMGPAGIPGAPGLMGARGPPGPPGTNGAPGQRGAAGEPGKNGAKGEPGPRGERGEAGSPGIPGPKGEDGKDGSPGEPGANGLPGAAGERGMPGFRGAPGANGLPGEKGPAGERGGPGPAGPRGVAGEPGRDGVPGGPGLRGMPGSPGGPGSDGKPGPPGSQGESGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGLPGPPGKNGETGPQGPPGPTGPGGDKGDTGPPGQQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAVGPSGPRGGAGPPGPEGGKGPAGPPGPPGAAGTPGLQGMPGERGGSGGPGPKGDKGDPGGSGADGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGLPGIAGPRGGPGERGEHGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGIAGQPGGTGPPGPPGPQGVKGERGSPGGPGAAGFPGGRGLPGPPGSNGNPGPPGSSGPPGKDGPPGPPGSSGAPGSPGVSGPKGDAGQPGEKGSPGPQGPPGAPGPGGISGITGARGLAGPPGMPGARGSPGPQGVKGENGKPGPSGLNGERGPPGPQGLPGLAGAAGEPGRDGNPGSDGLPGRDGAPGSKGDRGENGSPGAPGAPGHPGPPGPVGPAGKNGDRGETGPAGPAGAPGPAGSRGAPGPQGPRGDKGETGERGANGIKGHRGFPGNPGAPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGASGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGGGAAAIAGVGGEKAGGFAPYYG'},\n",
        "#     'COL3A1-Ovi_are': {'species': 'Ovis aries','offset': -17, 'sequence': 'QYEAYDVKSGVAGGVGLAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGLPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGIPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLLGARGPPGPPGTNGAPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIAGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGERGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAGGPPGPRGGAGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGTSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGEGGAPGVPGIAGPRGGPGERGEPGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPPGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGGRGPPGPPGSNGNPGPPGSSGAPGKDGPPGPPGSNGAPGSPGISGPKGDAGQPGERGAPGPQGPPGAPGPLGIAGLTGARGLAGPPGMPGARGSPGPQGIKGENGKPGPSGQNGERGPPGPQGLPGLAGAAGEPGRDGNPGSDGLPGRDGAPGTKGDRGENGSPGAPGAPGHPGPPGPVGPAGKNGDRGETGPAGPSGAPGPAGSRGAPGPQGPRGDKGETGERGANGIKGHRGFPGNPGPPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGTSGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGAAAIAGIGGEKAGGFAPYYG'},\n",
        "#     'COL5A1-Bos_tau': {'species': 'Bos taurus','offset': -9, 'sequence': ' AQPADLLKVLDFHNLPDGITKTTGFCAARRSSKGPDVAYRVTKDAQLSAPTKQLYPASAFPEDFSILTTVKAKKGSQAFLVSIYNEQGIQQIGLEMGRSPVFLYEDHTGKPGPEDYPLFRGINLSDGKWHRIALSVHKKNVTLILDCKKKTTKFLDRSEHPVIDVNGIVVFGTRILDEEVFEGDIQQLLLVSDHRAAYDYCEHYSPDCDTAVPDKPQSQDPNPDEYYPDGDNEGDTYYYEYPYYEDTDDTGKDPITTKTPVEAARETTEVAEELTQPPTEAPSAPSTTEGPGKEEDPGIGDYDYVPSEDYYTPPPYEDFGYGEGIENPDENPDQLPDQGARAEVPTSTVLTSNDSNPAPPPEEGIDDLEEEFTEETIKNLDESYYDPYYDPTISPSEIGPGMPANQDTIYEGMGGPRGEKGQKGEPAIIEPGMLLEGPPGPEGPAGLPGPPGTMGPTGQVGDPGERGPPGRPGLPGADGLPGPPGTMLMLPFRFGGGGDAGSKGPMVSAQESQAQAILQQARLALRGPAGPMGLTGRPGPMGPPGSGGLKGEPGDMGPQGPRGVQGPPGPAGKPGRRGRAGSDGARGMPGQTGPKGDRGFDGLAGLPGEKGHRGDPGPSGPPGPPGEDGERGDDGEVGPRGLPGEPGPRGLLGPKGPPGPPGPPGVTGMDGQPGPKGNVGPQGEPGPPGQQGNPGAQGLPGPQGAIGPPGEKGPLGKPGLPGMPGADGPPGHPGKEGPPGEKGGQGPPGPQGPIGYPGPRGVKGADGIRGLKGTKGEKGEDGFPGFKGDMGIKGDRGEIGPPGPRGEDGPEGPKGRGGPNGDPGPLGPPGEKGKLGVPGLPGYPGRQGPKGSIGFPGFPGANGEKGGRGTPGKPGPRGQRGPTGPRGERGPRGITGKPGPKGNSGGDGPAGPPGERGPNGPQGPTGFPGPKGPPGPPGKDGLPGHPGQRGETGFQGKTGPPGPPGVVGPQGPTGETGPMGERGHPGPPGPPGEQGLPGLAGKEGTKGDPGPAGLPGKDGPPGLRGFPGDRGLPGPVGALGLKGSEGPPGPPGPAGSPGERGPAGAAGPIGIPGRPGPQGPPGPAGEKGAPGEKGPQGPAGRDGLQGPVGLPGPAGPVGPPGEDGDKGEIGEPGQKGSKGDKGEQGPPGPTGPQGPIGQPGPSGADGEPGPRGQQGLFGQKGDEGPRGFPGPPGPVGLQGLPGPPGEKGETGDVGQMGPPGPPGPRGPSGAPGADGPQGPPGGIGNPGAVGEKGEPGEAGEPGLPGEGGPPGPKGERGEKGESGPSGAAGPPGPKGPPGDDGPKGSPGPVGFPGDPGPPGEPGPAGQDGPPGDKGDDGEPGQTGSPGPTGEPGPSGPPGKRGPPGPAGPEGRQGEKGAKGEAGLEGPPGKTGPIGPQGAPGKPGPDGLRGIPGPVGEQGLPGAPGPDGPPGPMGPPGLPGLKGDSGPKGEKGHPGLIGLIGPPGEQGEKGDRGVPGPQGSSGPKGEQGITGPSGPIGPPGPPGLPGPPGPKGAKGSSGPTGPKGEAGQPGPPGPPGPPGEVIQPLPIQASRTRRNIDASQLMDDAEGEQYMDYADGMEEIFGSLNSLKLEIEQMKRPLGTPQNPARTCKDLQLCHPDFPDGEYWVDPNQGCSRDSFKVYCNFTAGGSTCVFPDKKSEGARITSWPKENPGSWFSEFKRGKLLSYVDAEGNPVGVVQMTFLRLLSASAHQNITYNCYQSVAWQDAATGSYDKAMRFLGSNDEEMSYDNSPYIRALVDGCATKKGYQKTVLEIDTPKVEQVPIVDIMFNDFGEASQKFGFEVGPACFLG'},\n",
        "#     'COL6A1-Bos_tau': {'species': 'Bos taurus','offset': -9, 'sequence': ' QVCCSFECQPARGPPGPRGDPGYEGERGKPGLPGEKGEAGDPGRPGDLGPVGYQGMKGEKGSRGEKGSRGPKGYKGEKGKRGMDGVDGMKGETGFPGLPGCKGSPGFDGIQGPPGPKGDPGAFGLKGQKGEPGADGEPGRPGSTGPPGDEGEPGEPGPPGEKGEAGDEGNAGPDGAPGERGGPGERGPRGTPGARGPRGDPGEAGPQGDQGREGPVGVPGDPGEAGPIGPKGYRGDEGPPGTEGPKGAPGPAGPPGDPGLMGERGEDGPPGNGTEGFPGFPGYPGSRGPPGINGTKGYPGLKGDEGEAGDPGEDNNDIAARGAKGAKGYRGPEGPQGPPGHVGPPGPDECEILDIIMKMCSCCECKCGPIDILFVLD'},\n",
        "#     'COL6A1-Bub_bub': {'species': 'Bubalus bubalis','offset': -11, 'sequence': ' VMKKVSCLEIPGPPGPKGYRGQKGAKGNMGEPGEPGQKGRQGDPGIEGPIGFPGPKGVPGFKGEKGEFGADGRKGAPGLAGKNGTDGQKGKLGRIGPPGCKGDTGDRGPDGYVGEAGSPGERGDQGSKGDPGRPGRRGPPGENGAKGSKGYQGNNGAPGSPGLKGAKGGPGPRGPKGEPGRRGDPGTKGGPGSDGPKGEKGDPGPEGPRGLAGEVGNKGAKGDRGLPGPRGPQGTVGEPGKQGSRGDPGDAGPRGDSGQPGPKGDPGRPGFSYPGPRGAPGDKGEPGPPGPEGSRGDFGSKGEPGRKGQKGEPADPGPPGEPGPRGQRGAPGPEGEPGPPGDPGLTECDVMTYVRETCGCCDCEKRCGALDVVFVIDSSESIGYTNFTLEKNFVINVVNRLGAIAKDPKSETGTRVGVVQYS'},\n",
        "#     'COL6A2-Bos_tau': {'species': 'Bos taurus','offset': -7, 'sequence': 'VSCLEIPGPPGPKGYRGQKGAKGNMGEPGEPGQKGRQGDPGIEGPIGFPGPKGVPGFKGEKGEFGADGRKGAPGLAGKNGTDGQKGKLGRIGPPGCKGDTGDRGPDGYVGEAGSPGERGDQGSKGDPGRPGRRGPPGENGAKGSKGYQGNNGAPGSPGLKGAKGGPGPRGPKGEPGRRGDPGTKGGPGSDGPKGEKGDPGPEGPRGLAGEVGNKGAKGDRGLPGPRGPQGTVGEPGKQGSRGDPGDAGPRGDSGQPGPKGDPGRPGFSYPGPRGAPGDKGEPGPPGPEGGRGDFGSKGEPGRKGQKGEPADPGPPGEPGPRGQRGAPGPEGEPGPPGDPGLTECDVMTYVRETCGCCDCEKRCGALDVVFVIDSSESIGYTNFTLEKNFVINVVNRLGAIAKDPKSETGTRVGVVQYSHEGTFEAIQLDDERIDSLSSFKEAVKNLEWIAGGTWTPSALKFAYNKLIKESRRQKTRVFAVVITDGRHDPRDDDLNLRALCNHEVTVTAIGIGDMFHEKHESENLYSIACDKPQQVRNMTLFSDLVAEKFIDDMEDVLCPDPQIVCPDLPCQTELYVAQCTQRPVDIVFLLDGSERLGEQNFHKVRRFVEEVSRRLTLARRDDDPLNARVALLQFGGPREQQVAFPLTSNLTVIQEALASARYLNSFSHVGTGIVQAINQVVQGARAGARRHAELSFVFLTDGVTGNDSLDEAVHSMRKQNVVPTVVAVGSDVDTDVLSKISLGDPAAVFREKDYDSLAQPGFFDRFIRWIC'},\n",
        "\n",
        "#     'COL12A1-Bos_tau': {'species': 'Bos taurus','offset': -9, 'sequence': ' MRIRLPPALAALGAALLLSSIEAEVEPPSDLNFKIIDENTVHMSWAIPPDPIVGFKITVDPTTDGPTKEFTLSASTTETLLTELIPEIEYVVTITSYDEVEESVPVIGQLTIQTGGPTKPGEKKPGKTEIQKCSVSAWTDLVFLVDGSWSVGRNNFKYILDFIAALVSAFDIGEEKTRVGVVQYSSDTRTEFNLNQYYQQEELLAAIKKIPYKGGNTMTGEAIDYLIKNTFTESAGARVGFPKVAIIITDGKSQDEVEIPARELRNIGVEVFSLGIKAADAKELKQIASTPSLNHVFNVANFDAIVDIQNEIISQVCSGVDEQLGELVSGEEVVEPPSNLIATEVSSKYVKLSWSPSPSAVTGYKVLLIPMTTGSRQHALSVGPQTTTLSVRDLSADTEYQISVSAMKGLTSSEPVSIMEKTQPMKVQVECSRGVDIKADIVFLVDGSYSIGIANFVKVRAFLEVLVKSFEISPNRVQISLVQYSRDPHTEFTLKKFTKVEDIIEAINTFPYRGGSTNTGKAMTYVREKIFVPSKGSRSNVPKVMILITDGKSSDAFRDPAIKLRNSDVEIFAVGVKDAVRSELEAIASPPAETHVFTVEDFDAFQRISFELTQSICLRIEQELAAIKKKAYVPPKDINFSEVTSNSLKTSWSPAGENVFSYHITYKEATGDDQVTLVEPASSTSVVLSNLKPETLYLVNVTAEYEDGFSIPLAGEETTAEVKGAPRNLKVTDETTDSFKITWTQAPGRVLRYRITYRPVAGGESKEVTTPANQRRRTLENLTPDTKYEVSVIPEYVSGPGSPLTGNAATEEVRGNPRDLRVSDPTTSTMKLSWTRAPGKVKQYLITYTPVAGGDTQEVTVKGDTASTVLRGLNEGTQYALSVTALYASGAGDALFGEGTTLEERGSPGDLITKDITDTSIGAYWTSAPGMVRGYRVSWQSLYDDIETGEKSLPADVIHTVIENLQPETKYRVSIFATYSSGEGEPLSGEAMTELSEDSKTLKVDEETENTMRVTWKPAPGKVVNYRVMYRPRGGGRQMVAKVSPTVTSTVLKRLQPQTTYDITVLPVYKTGEGKLRQGSGTTASRFKSPRNLKTSDPTMSSFRVTWEPAPGEVKGYKVTFHPTGDDRRLGELVVGPYDNTVVLEELRAGTTYKVNVFGMFDGGESSPLVGQEMTTLSDTTVMPILSSGMECLTRAEADIVLLVDGSWSIGRANFRTVRSFISRIVEVFEIGPKRVQIALAQYSGDPRTEWQLNAHKDKKSLLQAVANLPYKGGNTLTGMALNFIRQQSFKTQAGMRPRARKIGVLITDGKSQDDVEAPSKKLKDEGVELFAIGIKNADEVELKMIATDPDDIHAYNVADFDSLSRIVDDLTNNLCNSVKGPGDLEAPSNLVLSERTHRSMRVSWTPPSDSVDRYKVEYYPVSGGKRQEFYVSRLERSTVLKDLKPETEYVVNVYSVVEDEYSEPLKGTEKTLPVPVVSLNIYDIGSTTMHVQWQPVGGATGYTVTYEPINATVPTKPKEMRLEPTVNDVQLSDLFPNTEYAVTVQAVLHDLTSEPVTAREVTLPLPSPRDLKLRDVTHSTMNVAWEPAPGKVRKYIVRYKTPEEEAKEVEVDRSRASTPLKDLFSQTLYTVSVSAVYEEGESPPVTAQETTHTVPAPTNLQITEVTPESFRGTWDHGASDVSLYRITWAPSGSADKMETILNGDENTLVFQNLNPNTLYEVSVTAIYPDESESDDLIGSERTLRLIPLTTQAPKSGPRNLQVYNATSNSLTVKWDPASGRVQKYRITYQPSTGEGNEQTTTIGGRQNSVLLQKLKPDTPYTITVSSLYPDGEGGRMTGRGKTKPLNTVRNLRVYDPSTSSLNVRWDHAEGSPRQYKLFYAPTAGGPEELVPIPGNTNYAILRNLQPDTQYTVTVVPVYSEGDGGRTSDTGRTLVRGLARNVQVYNPTPNSLDVRWDSAPGPVQQYRIVYSPQDGTRPSESIVVPGNTYTVHLEQLIPDTPYSVNIVALYSDAEGNPSPAQGRTLPRSGPRNLRVFGETTNSLSVAWDHADGPVQQYRIIYSPTVGDPIDEYTTVPGRRNNVILQPLQPDTPYKITVIPVYEDGDGTHLTGNGRTVGLLPPQNIHISDEWYTRFRVSWDPSPSPVLGYKIVYKPVGSNEPMEVFVGEVTSYTLHNLNPSTTYDVNVYAQYDSGLSVPLTDQGTTLYLNVTDLKTYQVGWDTFCVRWLAHRAATSYRLKLSPADGTRGQEITVRGSETSHCFTDLSPETDYGVTVFVQTPNLEGPGVSVKERTTLKPTEAPTEPPTPPPPPTIPPAREVCKGAKADIVFLTDASWSIGDDNFNKVVKFIFHTVGGFDEISPAGIQVSFVQYSDEVKSEFKLNTYNDKALALGALQNIRYRGGNTRTGKALTFIKEKVLTWESGMRKNVPKVLVVVTDGRSQDEVKKAALVIQQSGFSVFVVGVADVDYNELANIASKPSERHVFIVDDFESFEKIEDNLITFVCETATSSCPLIYLDGYTSPGFKMLEAYNLTEKNFASVQGVSLESGSFPSYSSYRLQKNAFVNQPTVELHPNGLPPSYTIILLFRLLPETPSDPFAIWQITDRDYKPQVGVIADPSSKTLSFFNKDTRGEVQTVTFDTDEVKALFYGSFHKVHIVVTSKSVKIYIDCYEIVEKDIKEAGNITTDGYEILGKLLKGERKSATFQIQNFDIVCSPVWTSRDRCCDIPSRRDEAKCPALPNACTCTQDSVGPPGPPGPAGGPGAKGPRGERGVSGAVGPPGPRGDTGPPGPQGPPGPQGPNGLSIPGEQGRQGMKGDAGEPGLPGRTGTPGLPGPPGPMGPPGDRGFTGKDGAMGPRGPPGPPGSPGSPGVTGPSGKPGKPGDHGRPGPSGLKGEKGDRGDIASQNMMRAVARQVCEQLISGQMNRFNQMLNQIPNDYHSSRNQPGPPGPPGPPGSAGARGEPGPGGRPGFPGTPGMQGPQGERGLPGEKGERGIGSPGPRGLPGPPGPQGESRTGPPGSTGSRGPPGPPGRPGNSGIRGPPGPPGYCDSSQCASIPYNGQGYPGSG'},\n",
        "#     'COLunk-Bos_tau': {'species': 'Bos taurus','offset': -9, 'sequence': ' MRKHRHLPLVAILCLFFSGFSFTRGQQQPAQDSADIIFLIDGSNNTGSVNFAVILDFLVNLLERLSIGTQQIRVGVVQYSDEPRTMFSLNSYSTKAQVLDAVKALGFIGGELANVGLALDFVVENHFTRAGGSRAEEGVPQVLVLISAGPSSDEIRDGVVALKQASVFSFGLGAQAASKAELQHIATNDNLVFTVPEFRSLGDVQEQLLPYIVGVAQRHIVLQPPTIVTQVIEVNKRDIVFLVDGSSKLGLTNFNAIRDFIAKVIQRLEIRQDLIQVAVAQYADTVRPEFYFNTYPSKREVINAVRKMKALDGSALYTGSALDFVRNNLFTEAAGYRAAEGVPKLLVLVTGGKSLDAVSQPAQELKRSGILAFAVGNKVADQAELEEIAFDSSLVFTATEFRPAPLQGVLPGLLGPLRTLTGTTEVRVNKRDIIFLLDGSSNVGETNFPYVRDFVMNLVNSLDVGSDHIRVGLVQFSDTPVTEFSLNTYPTKSELLAHLRQMQLQGGSVLNTGAALSYVHANHFTEAGGSRIQDHVPQLLLLLTAGQSEDSYLQAANALARAGILTFCVGTSQADRAELEEIAFNPGLVYLMDDFSSLPALPQQLIQPLTTYVSGGVEEVPLAQPESKRDILFLFDGSANLMGQFTAARDFLYKVIDELDVKPEGTRVAVAQFSDDVKVESRFDQHQNKPEILNLVKRMKLKTGKALNLGYALDYAQRYIFVKSAGSRIEDGVLQFLVLLVAGKSSDRVDTPALNLKQSGVVPFILQAKNADPAELELIVPSPAFILVAESLPKIGDLQPQIVNLLKSVQNGAPAPVSVEKDVVFLIDGSEGVRSGFPLLKEFVQRVVESLDVGPDRVRVAVVQYSDRTRPEFYLNSYMDQQSVVGAIRGLTLLGGPAPNTGAALEFVLRNILVGSAGSRIAEGVPQLLIVLTADRSGDDVRGPSVVLRRGGAVPIGIGIGNADITEMQTLSFVPDFAVVIPTFRQLGTIQQVISERVTQLSREELSRLQASVTPLTTPVVSSKRDVVFLIDGSQSAGPEFQYIRTLIERLVDYLDVGFDTTRVAVIQFSDDPRVEFLLNVHSSKDEVQNAVRRLRPKGGRQVNVGGALEYVARNIFKRPLGSRIEEGVPQFLVLISSGKSDDEVEDSAIELKQFGVAPLTIARNVDQEELVKISLSPEYVFSVNTFRELPSLEQKLLTPLTTLTAGQIQQLLASTRYPPPAVESDAADIVFLIDSSDGVKPDGIAHIRDFVIRIVRRLNVGPNKVRIGVLQFSNDVFPEFQLKTYKSQASVLDAIRRLRFKGGSPLNTGKALEFVARNYFVKSAGSRIEDGVPQHLVLFLGGKSQDDISRYSQVIKSAGIASLGVGDRNIDRTELQTITSDPRLVFTVREFRDLPSIEERMVNSFGSSGVTPAPPGVDTPSPSRPEKKKADIVFLLDGSINFRRDSFQEVLRFVSEIVDTVYEGGDSIQVGLVQYNSDPTDEFFLKDFPTKQQIIDAINKVVYKGGRHANTKVGLEHLRRNHFVPEAGSRLDQRVPQIAFVITGGKSVEDAQEASMALTQRGVKVFAVGVRNIDSEEVGKIASNSATAFRVGNVQELSELSEQVLETLHDAMHETLCPGVTDVSKACNLDVILGFDGSGDQNVFVAQKGLEPKVDTILRRISQMQKISCSGSQLPTVRVSVVALTPSGPVEAFDFAEYQPELFEKFQNMRAQHPYVLTADTLKLYQNKFRQASMDNVKVVIHFTDGVDGDLADLQRASEELRQEGVRALILVGLERVANLEQLMQLEFGRGFTYNRPLRLNLLDLDYELAEQLDNIAEKACCGIPCKCSGQRGDRGPIGSIGPKGIPGEDGYRGYPGDEGGPGERGPPGVNGTQGFQGCPGQRGVKGSRGFPGEKGELGEIGLDGLDGEDGDKGLPGTSGEKGSPGRRGDKGPKGDKGERGDVGIRGDPGNSGQDSQQRGAKGETGDIGPVGLPGRDGVSGGPGEPGKSGGFGRRGPAGAKGNQGGPGQQGPVGEQGTRGAQGPPGPTGPPGLIGEQGIPGPRGGGGTAGAPGERGRTGPLGRKGEPGDPGPKGSIGNRGPRGETGDDGRDGVGSEGRIGKKGERGFPGYPGPKGNPGEPGTDGPPGPKGIRGRRGNSGPPGIAGQKGDPGYPGPSGYKGSRGDSMDQCALVQSIKDKCPCCYGPLECPVFPTELAFALDTSEGVTQDRFSQMREVVLKILDDLTIAESNCPRGARVAVVTYNNEVTTEIRFADSKKKSVLLDKIKNLQVSLTSKQQSLETAMSFVARNTFKRVRSGFLMRKVAVFFSNKPTRATPQLREAVLKLSDAGITPLFLTSQEDRQLINALQINNTAVGHALVLPAGGDLTDFLKNVLTCHVCLDICNIDPSCGFGSWRPSFRDRRAAGSDVDIDMAFILDSSESTTPFQFNEMRKYIEYLVRQLDVSPDPKASQHFARVAVVQHAPYEAVDNASVPPVKVEFSLTDYGSKEKLLAFLGSRMTQLQGTRALGRAIDYTIENIFESAPNPRDLKIVVLMLTGEVQKQQLEEAQRAILQAKCKGYFFVILGIGRKVNVKELYSSASEPNDVFFKLVDKSTELNEEPLMRFGRLLPSFVSSKDAFYLSPDIRKQCDWFQADQPAKNLVQFGYKQINVPNNVTSSPTSKLVTTTKPVITTTKPITVVNLPASKPAATRPVDERPAAGRSVATKMDSARPVATKTEATKPVATKTEATKTEAAKPVVTKAEAAKPTATKTEATKTEAAKLMATKTEATKPVAARPEAAKTATARPAAAAKPAAAKPAPARPPTAARSMATRSEAPRPQAVKLAASRPGAAKPVVKAPREIHVSEVTENSAKLHWERPEPPSPYLYNLTVTSAHDQAVVLKQNLTVTDRVIGGLLPGQTYHVTVTCSLRSQVRAIYQGSFSTKKIQPPPPQTERSASSATINLVVSADRLAGSKADICKLPKDGGTCREFVLKWYYDSVTENCARFWYGGCGGNENRFNSQDECEKVCPPVPIKQPGVIAAMGT'},\n",
        "#     'FibMod-Bos_mut' : {'species': 'Bos mutus','offset': -1, 'sequence': 'MTPKESRVRRMHSPGRLEEEGCSCLPRCLVWKEKWTQRLTVPTSRQPPWACSSPEPSMEKCAAGTRANSTMNSICDYAQTKKGSKVSLPRGDTHPSGRCSLLKHKMQWASILLLAGLCSLSWAQYEEDSHWWFQFLRNQQSTYDDPYDPYPYEPYEPYPYGGEEGPAYAYGSPPQPEPRDCPQECDCPPNFPTAMYCDNRNLKYLPFVPSRMKYVYFQNNQISSIQEGVFDNATGLLWIALHGNQITSDKVGKKVFSKLRHLERLYLDHNNLTRIPSPLPRSLRELHLDHNQISRVPNNALEGLENLTALYLHHNEIQEVGSSMKGLRSLILLDLSYNHLRKVPDGLPSALEQLYLEHNNVFSVPDSYFRGSPKLLYVRLSHNSLTNNGLASNTFNSSSLLELDLSYNQLQKIPPVSTNLENLYLQGNRINEFSISSFCTVVDVMNFSKLQVLRLDGNEIKRSAMPADAPLCLRLASLIEI'},\n",
        "\n",
        "#     'K2C1-Hom_sap': {'species': 'Homo sapiens','offset': 0, 'sequence': 'MSRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTR'},\n",
        "#     'K22E-Hom_sap': {'species': 'Homo sapiens','offset': 0, 'sequence': 'MSCQISCKSRGRGGGGGGFRGFSSGSAVVSGGSRRSTSSFSCLSRHGGGGGGFGGGGFGSRSLVGLGGTKSISISVAGGGGGFGAAGGFGGRGGGFGGGSSFGGGSGFSGGGFGGGGFGGGRFGGFGGPGGVGGLGGPGGFGPGGYPGGIHEVSVNQSLLQPLNVKVDPEIQNVKAQEREQIKTLNNKFASFIDKVRFLEQQNQVLQTKWELLQQMNVGTRPINLEPIFQGYIDSLKRYLDGLTAERTSQNSELNNMQDLVEDYKKKYEDEINKRTAAENDFVTLKKDVDNAYMIKVELQSKVDLLNQEIEFLKVLYDAEISQIHQSVTDTNVILSMDNSRNLDLDSIIAEVKAQYEEIAQRSKEEAEALYHSKYEELQVTVGRHGDSLKEIKIEISELNRVIQRLQGEIAHVKKQCKNVQDAIADAEQRGEHALKDARNKLNDLEEALQQAKEDLARLLRDYQELMNVKLALDVEIATYRKLLEGEECRMSGDLSSNVTVSVTSSTISSNVASKAAFGGSGGRGSSSGGGYSSGSSSYGSGGRQSGSRGGSGGGGSISGGGYGSGGGSGGRYGSGGGSKGGSISGGGYGSGGGKHSSGGGSRGGSSSGGGYGSGGGGSSSVKGSSGEAFGSSVTFSFR'},\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "#     'TRYUP-Sus_sco': {'species': 'Bos taurus','offset': 0, 'sequence': 'IVGGYTCAANSIPYQVSLNSGSHFCGGSLINSQWVVSAAHCYKSRIQVRLGEHNIDVLEGNEQFINAAKIITHPNFNGNTLDNDIMLIKLSSPATLNSRVATVSLPRSCAAAGTECLISGWGNTKSSGSSYPSLLQCLKAPVLSDSSCKSSYPGQITGNMICVGFLEGGKDSCQGDSGGPVVCNGQLQGIVSWGYGCAQKNKPGVYTKVCNYVNWIQQTIAAN'},\n",
        "\n",
        "#     'TPM1-Cam_dro' : {'species': 'Camelus dromarius','offset': 0, 'sequence':'MDAIKKKMQMLKLDKENAIDRAEQAEADKKQAEDRCKQLEEEQQALQKKLKGTEDEVEKYSESVKDAQEKLEQAEKKATDAEADVASLNRRIQLVEEELDRAQERLATALQKLEEAEKAADESERGMKVIENRAMKDEEKMELQEMQLKEAKHIAEDSDRKYEEVARKLVILEGELERSEERAEVAESRARQLEEELRTMDQALKSLMASEEEEGLGLPNSICCHSHNFALLFSPPPLPHCAISTVSPTVSSHPPPPHTPCSKCGDLEEELKIVTNNLKSLEAQADKYSTKEDKYEEEIKLLEEKLKEAETRAEFAERSVAKLEKTIDDLEDEVYAQKMKYKAISEELDNALNDITSL'},\n",
        "#     'KS1-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTR'},\n",
        "#     'KS9-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'MSCRQFSSSYLSRSGGGGGGGLGSGGSIRSSYSRFSSSGGGGGGGRFSSSSGYGGGSSRVCGRGGGGSFGYSYGGGSGGGFSASSLGGGFGGGSRGFGGASGGGYSSSGGFGGGFGGGSGGGFGGGYGSGFGGFGGFGGGAGGGDGGILTANEKSTMQELNSRLASYLDKVQALEEANNDLENKIQDWYDKKGPAAIQKNYSPYYNTIDDLKDQIVDLTVGNNKTLLDIDNTRMTLDDFRIKFEMEQNLRQGVDADINGLRQVLDNLTMEKSDLEMQYETLQEELMALKKNHKEEMSQLTGQNSGDVNVEINVAPGKDLTKTLNDMRQEYEQLIAKNRKDIENQYETQITQIEHEVSSSGQEVQSSAKEVTQLRHGVQELEIELQSQLSKKAALEKSLEDTKNRYCGQLQMIQEQISNLEAQITDVRQEIECQNQEYSLLLSIKMRLEKEIETYHNLLEGGQEDFESSGAGKIGLGGRGGSGGSYGRGSRGGSGGSYGGGGSGGGYGGGSGSRGGSGGSYGGGSGSGGGSGGGYGGGSGGGHSGGSGGGHSGGSGGNYGGGSGSGGGSGGGYGGGSGSRGGSGGSHGGGSGFGGESGGSYGGGEEASGSGGGYGGGSGKSSHS'},\n",
        "#     'KS2-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTRR'},\n",
        "#     'KS10-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'MSVRYSSSKHYSSSRSGGGGGGGGCGGGGGVSSLRISSSKGSLGGGFSSGGFSGGSFSRGSSGGGCFGGSSGGYGGLGGFGGGSFRGSYGSSSFGGSYGGIFGGGSFGGGSFGGGSFGGGGFGGGGFGGGFGGGFGGDGGLLSGNEKVTMQNLNDRLASYLDKVRALEESNYELEGKIKEWYEKHGNSHQGEPRDYSKYYKTIDDLKNQILNLTTDNANILLQIDNARLAADDFRLKYENEVALRQSVEADINGLRRVLDELTLTKADLEMQIESLTEELAYLKKNHEEEMKDLRNVSTGDVNVEMNAAPGVDLTQLLNNMRSQYEQLAEQNRKDAEAWFNEKSKELTTEIDNNIEQISSYKSEITELRRNVQALEIELQSQLALKQSLEASLAETEGRYCVQLSQIQAQISALEEQLQQIRAETECQNTEYQQLLDIKIRLENEIQTYRSLLEGEGSSGGGGRGGGSFGGGYGGGSSGGGSSGGGHGGGHGGSSGGGYGGGSSGGGSSGGGYGGGSSSGGHGGSSSGGYGGGSSGGGGGGYGGGSSGGGSSSGGGYGGGSSSGGHKSSSSGSVGESSSKGPRY'},\n",
        "#     'KS1-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'SRQFSSRSGYRSGGGFSSGSAGIINYQRRTTSSSTRRSGGGGGRFSSCGGGGGSFGAGGGFGSRSLVNLGGSKSISISVARGGGRGSGFGGGYGGGGFGGGGFGGGGFGGGGIGGGGFGGFGSGGGGFGGGGFGGGGYGGGYGPVCPPGGIQEVTINQSLLQPLNVEIDPEIQKVKSREREQIKSLNNQFASFIDKVRFLEQQNQVLQTKWELLQQVDTSTRTHNLEPYFESFINNLRRRVDQLKSDQSRLDSELKNMQDMVEDYRNKYEDEINKRTNAENEFVTIKKDVDGAYMTKVDLQAKLDNLQQEIDFLTALYQAELSQMQTQISETNVILSMDNNRSLDLDSIIAEVKAQYEDIAQKSKAEAESLYQSKYEELQITAGRHGDSVRNSKIEISELNRVIQRLRSEIDNVKKQISNLQQSISDAEQRGENALKDAKNKLNDLEDALQQAKEDLARLLRDYQELMNTKLALDLEIATYRTLLEGEESRMSGECAPNVSVSVSTSHTTISGGGSRGGGGGGYGSGGSSYGSGGGSYGSGGGGGGGRGSYGSGGSSYGSGGGSYGSGGGGGGHGSYGSGSSSGGYRGGSGGGGGGSSGGRGSGGGSSGGSIGGRGSSSGGVKSSGGSSSVKFVSTTYSGVTR'},\n",
        "#     'KS1-Ovi_are' : {'species': 'Ovis aries','offset': 0, 'sequence':'SLHVNMSRHFSSRSGYRCGGGGGGFSSGSAGVVSYQRRSTSSSVRRSGGGGGRFSGGVCGGGAGGGFGVAGGGGRGGFGGGYGGGYGGGGFGGGGFGGGGFGGGGFGGGGIGGGFGGGGFGGGGFGGGGVWGGGGGIQESQEREQIKSLNNRFASFIDKVRFLEQQNQVLQTKWELLQQINTSTRTYSLEPFFEAYISNLRRNVDQLKSDRSRLDSELKNMQDLVEDYRSKYEDEINKRTNAENEFVGIKKDVDAAYITKVDLQAKFDNLLQEIDFFKTLFQAELSQMQTHISDTNVILSMDNNRNLDLNSIIDEVKNQYEEIAQRSKAEAEALYQTKYEELQITAGQHGDSLKNTKMEISELNRVIQRLRSEIDSVKKQISSLQQAISDAEQRGENAIKDAQNKLNELEDALQKAKEDMARLLRDYQELMNTKLALDVEIATYRTLLEGEESRMSGECVPNVSVSVSTSHTSISGGGGRGGGGFSSGGGGYSSGGGGGGGYGSGGGGGYGSGGGSSYGSRGGRADPPRWGRAGLYGRTSSSCLSRHGGGSGVFGGGGFGGGGFSSRSLINLGGSRSTSISVSGGGGSFGSYGGFGGRGSGFGGGLGVGGGSFGGGRFGGGGGFGGFGGPGGFGGPAGFGGYGSGGRGSSSGGGFSSGSSSYSSGGRRSGSRSGGGGSYGSGGGSRGGSSFGGGYSSGGKHSSGGGARGGSSSGGGTYGSERGASGSSEGGSSSEKSVVEGICTLESSFLKESPCH'},\n",
        "#     'KS1c9-Hom_sap' : {'species': 'Homo sapiens','offset': 0, 'sequence':'MSCRQFSSSYLSRSGGGGGGGLGSGGSIRSSYSRFSSSGGGGGGGRFSSSSGYGGGSSRVCGRGGGGSFGYSYGGGSGGGFSASSLGGGFGGGSRGFGGASGGGYSSSGGFGGGFGGGSGGGFGGGYGSGFGGFGGFGGGAGGGDGGILTANEKSTMQELNSRLASYLDKVQALEEANNDLENKIQDWYDKKGPAAIQKNYSPYYNTIDDLKDQIVDLTVGNNKTLLDIDNTRMTLDDFRIKFEMEQNLRQGVDADINGLRQVLDNLTMEKSDLEMQYETLQEELMALKKNHKEEMSQLTGQNSGDVNVEINVAPGKDLTKTLNDMRQEYEQLIAKNRKDIENQYETQITQIEHEVSSSGQEVQSSAKEVTQLRHGVQELEIELQSQLSKKAALEKSLEDTKNRYCGQLQMIQEQISNLEAQITDVRQEIECQNQEYSLLLSIKMRLEKEIETYHNLLEGGQEDFESSGAGKIGLGGRGGSGGSYGRGSRGGSGGSYGGGGSGGGYGGGSGSRGGSGGSYGGGSGSGGGSGGGYGGGSGGGHSGGSGGGHSGGSGGNYGGGSGSGGGSGGGYGGGSGSRGGSGGSHGGGSGFGGESGGSYGGGEEASGSGGGYGGGSGKSSHS'},\n",
        "\n",
        "\n",
        "\n",
        "#     'ALB-Bos_tau' : {'species': 'Bos taurus','offset': 0, 'sequence':'MKWVTFISLLLLFSSAYSRGVFRRDTHKSEIAHRFKDLGEEHFKGLVLIAFSQYLQQCPFDEHVKLVNELTEFAKTCVADESHAGCEKSLHTLFGDELCKVASLRETYGDMADCCEKQEPERNECFLSHKDDSPDLPKLKPDPNTLCDEFKADEKKFWGKYLYEIARRHPYFYAPELLYYANKYNGVFQECCQAEDKGACLLPKIETMREKVLASSARQRLRCASIQKFGERALKAWSVARLSQKFPKAEFVEVTKLVTDLTKVHKECCHGDLLECADDRADLAKYICDNQDTISSKLKECCDKPLLEKSHCIAEVEKDAIPENLPPLTADFAEDKDVCKNYQEAKDAFLGSFLYEYSRRHPEYAVSVLLRLAKEYEATLEECCAKDDPHACYSTVFDKLKHLVDEPQNLIKQNCDQFEKLGEYGFQNALIVRYTRKVPQVSTPTLVEVSRSLGKVGTRCCTKPESERMPCTEDYLSLILNRLCVLHEKTPVSEKVTKCCTESLVNRRPCFSALTPDETYVPKAFDEKLFTFHADICTLPDTEKQIKKQTALVELLKHKPKATEEQLKTVMENFVAFVDKCCAADDKEACFAVEGPKLVVSTQTALA'},\n",
        "#     'ALB-Equ_cab' : {'species': 'Equus caballus','offset': 0, 'sequence':'MKWVTFVSLLFLFSSAYSRGVLRRDTHKSEIAHRFNDLGEKHFKGLVLVAFSQYLQQCPFEDHVKLVNEVTEFAKKCAADESAENCDKSLHTLFGDKLCTVATLRATYGELADCCEKQEPERNECFLTHKDDHPNLPKLKPEPDAQCAAFQEDPDKFLGKYLYEVARRHPYFYGPELLFHAEEYKADFTECCPADDKLACLIPKLDALKERILLSSAKERLKCSSFQNFGERAVKAWSVARLSQKFPKADFAEVSKIVTDLTKVHKECCHGDLLECADDRADLAKYICEHQDSISGKLKACCDKPLLQKSHCIAEVKEDDLPSDLPALAADFAEDKEICKHYKDAKDVFLGTFLYEYSRRHPDYSVSLLLRIAKTYEATLEKCCAEADPPACYRTVFDQFTPLVEEPKSLVKKNCDLFEEVGEYDFQNALIVRYTKKAPQVSTPTLVEIGRTLGKVGSRCCKLPESERLPCSENHLALALNRLCVLHEKTPVSEKITKCCTDSLAERRPCFSALELDEGYVPKEFKAETFTFHADICTLPEDEKQIKKQSALAELVKHKPKATKEQLKTVLGNFSAFVAKCCGREDKEACFAEEGPKLVASSQLALA'},\n",
        "#     'ALB-Equ_asn' : {'species': 'Equus asinus','offset': 0, 'sequence':'MKWVTFVSLLFLFSSAYFRGVLRRDTHKSEIAHRFNDLGEKHFKGLVLVAFSQYLQQCPFEDHVKLVNEVTEFAKKCAADESAENCDKSLHTLFGDKLCTVATLRATYGELADCCEKQEPERNECFLTHKDDHPNLPKLKPEPDAQCAAFQEDPDKFLGKYLYEVARRHPYFYGPELLFHAEEYKADFTECCPADDKAGCLIPKLDALKERILLSSAKERLKCSSFQKFGERAFKAWSVARLSQKFPKADFAEVSKIVTDLTKVHKECCHGDLLECADDRADLTKYICEHQDSISGKLKACCDKPLLQKSHCIAEVKEDDLPSDLPALAADFAEDKEICKHYKDAKDVFLGTFLYEYSRRHPDYSVSLLLRIAKTYEATLEKCCAEADPPACYATVFDQFTPLVEEPKSLVKKNCDLFEEVGEYDFQNALIVRYTKKAPQVSTPTLVEIGRTLGKVGSRCCKLPESERLPCSENHLALALNRLCVLHEKTPVSEKITKCCTDSLAERRPCFSALELDEGYIPKEFKAETFTFHADICTLPEDEKQIKKQSALAELVKHKPKATKEQLKTVLGNFSAFVAKCCGAEDKEACFAEEGPKLVASSQLALA'},\n",
        "#     'ALB-Sus_scr' : {'species': 'Sus scrofa','offset': 0, 'sequence':'MKWVTFISLLFLFSSAYSRGVFRRDTYKSEIAHRFKDLGEQYFKGLVLIAFSQHLQQCPYEEHVKLVREVTEFAKTCVADESAENCDKSIHTLFGDKLCAIPSLREHYGDLADCCEKEEPERNECFLQHKNDNPDIPKLKPDPVALCADFQEDEQKFWGKYLYEIARRHPYFYAPELLYYAIIYKDVFSECCQAADKAACLLPKIEHLREKVLTSAAKQRLKCASIQKFGERAFKAWSLARLSQRFPKADFTEISKIVTDLAKVHKECCHGDLLECADDRADLAKYICENQDTISTKLKECCDKPLLEKSHCIAEAKRDELPADLNPLEHDFVEDKEVCKNYKEAKHVFLGTFLYEYSRRHPDYSVSLLLRIAKIYEATLEDCCAKEDPPACYATVFDKFQPLVDEPKNLIKQNCELFEKLGEYGFQNALIVRYTKKVPQVSTPTLVEVARKLGLVGSRCCKRPEEERLSCAEDYLSLVLNRLCVLHEKTPVSEKVTKCCTESLVNRRPCFSALTPDETYKPKEFVEGTFTFHADLCTLPEDEKQIKKQTALVELLKHKPHATEEQLRTVLGNFAAFVQKCCAAPDHEACFAVEGPKFVIEIRGILA'},\n",
        "\n",
        "#     'COL1A1-Gal_gal': {'species': 'Gallus gallus','offset': -7, 'sequence': 'SAGVAVPGPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGQRGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGQPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGAPGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGAVGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGAPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGEVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEK'},\n",
        "#     'COL1A1-Clu_har': {'species': 'Clupea harengus','offset': 0, 'sequence': 'GPMGPMGSRGSPGPPGNSGPQGFTGPPGEPGEAGAPGSLGPRGAAGPPGKNGEDGEPGKPGRGGERGSNGPQGARGFPGTPGLPGIKGHRGFSGLDGAKGDSGPAGPKGEGGASGENGTPGAMGPRGLPGERGRSGPNGAAGARGNDGSAGAAGPPGPTGPAGPPGFPGGPGSKGELGAQGARGGEGPAGARGEPGNPGPAGAAGPAGNNGADGAPGNKGAPGAAGVAGAPGFPGPRGPPGAAGATGAPGSKGNTGEAGAPGSKGESGAKGEAGVAGVQGPPGSAGEEGKRGARGEPGGVGTRGAPGERGAPGSRGFPGADGSAGPKGATGERGGAGVVGPKGSSGEPGRNGEPGMPGSKGMTGSPGSPGPDGKLGPSGASGQDGRPGAPGPVGARGQPGVMGFPGPKGAAGEGGKPGERGVAGPLGPGGAAGKDGDVGAPGAPGPSGPSGERGEQGPAGPPGFQGLPGPQGSTGEPGKAGEQGLSGEAGAVGPSGSRGDRGFPGERGGPGAAGPAGARGSPGSAGNDGAKGDVGAPGASGGQGAPGLQGMPGERGAAGLPGLKGERGEQGAKGTDGVTGKDGVRGMTGPIGPPGPAGAPGDKGEGGALGAAGPTGARGPPGERGESGPPGPAGFAGPPGSDGQAGGKGESGDTGSKGDSGSPGPAGATGAPGPQGPVGATGSKGARGAAGPPGATGFPGAVGRVGPPGPSGNSGPAGPPGPGGKEGPKGNRGDTGPAGRTGELGAAGPPGAPGEKGAPGGEGPSGASGTPGPSGIAGARGIVGLPGQRGERGFPGLSGQLGEPGKQGPAGPSGERGPNGPMGPAGLAGPPGEAGREGAPGQEGAVGRDGAAGQKGERGESGVSGAPGAPGPSGAPGPVGPAGKSGDRGESGPSGPAGSAGPAGPRGPAGPAGARGDKGEAGEAGERGMKGHRGFTGMQGPPGSPGSSGEAGPAGSSGPAGPRGPGGSAGSAGKDGMSGLPGPIGPPGPRGRNGEIGPAGPPGPPGPPGPPGAPGGGFDLGFIAQPQEKAPDPFR'},\n",
        "#     'COL1A1-Fal_tin': {'species': 'Falco tinnunculus','offset': 0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGAPGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGPAGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPQGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGDVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGGR'},\n",
        "#     'COL1A1-Apr_xxx': {'species': 'Apr xxx','offset': -0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGSPGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGGRGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGPAGPKGSPGESGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGAPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGIPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGSDGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGSAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPPGPAGARGPSGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGDVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQ'},\n",
        "#     'COL1A1-Caa_xxx': {'species': 'Caa xxx','offset': -0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGEDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGAPGAAGPPGPTGPAGPPGFPGAGGAKGETGPPGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNAGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGPVGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQGGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEVGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGETGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGAPGERGPPGPMGPPGLAGPPGEAGREGSPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPQGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGEAGPVGPPGPPGPPGPPGPPSGGFDFSFMPQ'},\n",
        "#     'COL1A1-Nop_xxx': {'species': 'Nop xxx','offset': -0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGEPGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGPSGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAAGAKGETGPQGARGSEGPQGARGEPGPPGPAGAAGPAGNPGADGQPGAKGATGAPGIAGAPGFPGARGPSGPQGPSGAPGPKGNSGEPGAPGNKGDTGAKGEPGPAGVQGPPGPAGEEGKRGARGEPGPAGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGSPGAVGPKGSPGEAGRPGEPGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKPGERGAPGPPGAVGAAGKDGEAGAQGPPGPTGPAGERGEQGPAGAPGFQGLPGPAGPPGEAGKPGEQGVPGDAGAPGPAGARGERGFPGERGVQGPPGPQGPRGANGAPGNDGAKGDAGAPGAPGNQGPPGLQGMPGERGAAGLPGAKGDRGDPGPKGADGAPGKDGLRGLTGPIGPPGPAGAPGDKGEAGPPGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGETGDAGAKGDAGPPGPAGPTGAPGPAGAVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGGKGPRGETGPAGRPGEPGPAGPPGPPGEKGSPGADGPIGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGSPGERGPPGPMGPPGLAGPPGEAGREGAPGAEGAPGRDGAAGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPPGPAGARGPAGPQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGQPGEQGPSGASGPAGPRGPPGSAGAAGKDGLNGLPGPIGPPGPRGRTGDVGPVGPPGPPGPPGPPGPPSGGFDFSFLPQ'},\n",
        "#     'COL1A1-Tha_sir': {'species': 'Thamnophis sirtalis','offset': -0, 'sequence': 'GPMGPAGPRGLPGPPGAPGPQGFQGPPGEPGEPGSSGPMGPRGPAGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQVGPRGLPGERGRPGPAGSAGARGNDGAAGAAGPTGPTGPAGPPGFPGPAGPKGETGPQGARGGEGAQGARGEPGAAGPAGAAGPAGNPGSDGQPGAKGAPGAPGISGAPGFPGARGAAGPQGPTGAPGPKGNSGEPGAPGNKGDAGAKGEAGPAGVQGPSGPPGEEGKRGSRGEPGPSGLPGPAGERGAPGSRGFPGADGIAGPKGPPGERGAPGPAGPKGSTGEAGRTGEPGLPGAKGLTGSPGSPGADGKTGPAGPAGQDGRPGPAGPPGARGQAGVMGFPGPKGSAGEPGKPGERGAPGATGANGAPGKDGDAGAQGPPGPAGPAGERGEQGPSGAPGFQGLPGPAGAPGESGKPGEQGVPGDVGAPGPSGARGERGFPGERGAQGPAGPSGPRGANGAPGNDGAKGDAGAPGAPGGQGPPGLQGMPGERGAAGLPGAKGDRGDPGAKGTDGSPGKDGPRGLTGPIGPPGPAGSPGDKGESGPSGPAGPTGARGAPGDRGEPGPAGPAGFAGPPGTDGQPGAKGEPGDAGAKGDAGPPGPAGATGPAGPAGPVGAPGPKGARGNAGPPGATGFPGAAGRVGPPGPSGNIGLPGPPGPSGKEGSKGPRGETGPAGRPGEVGPAGPPGPSGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPTGEPGKQGPSGASGERGPPGPSGPPGLAGPPGEAGREGSPGAEGAPGRDGAAGPKGDRGESGPAGAPGAPGAPGAPGPVGPAGKNGERGESGPAGPAGPAGPSGARGPAGAQGPRGDKGETGEQGDRGMKGHRGFSGLQGPPGPPGSPGEQGPSGSSGPAGPRGPPGSAGSSGKDGLNGLPGPIGPPGPRGRTGDVGPAGPAGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGGR'},\n",
        "#     'COL1A2-Chk_COL1A2?': {'species': 'Gallus gallus','offset': -4, 'sequence': 'AADFGPGPMGLMGPRGPPGASGPPGPPGFQGVPGEPGEPGQTGPQGPRGPPGPPGKAGEDGHPGKPGRPGERGVAGPQGARGFPGTPGLPGFKGIRGHNGLDGQKGQPGTPGTKGEPGAPGENGTPGQPGARGLPGERGRIGAPGPAGARGSDGSAGPTGPAGPIGAAGPPGFPGAPGAKGEIGPAGNVGPTGPAGPRGEIGLPGSSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPPGPAGPSGARGLVGEPGPAGAKGESGNKGEPGAAGPPGPPGPSGEEGKRGSNGEPGSAGPPGPAGLRGVPGSRGLPGADGRAGVMGPAGNRGASGPVGAKGPNGDAGRPGEPGLMGPRGLPGQPGSPGPAGKEGPVGFPGADGRVGPIGPAGNRGEPGNIGFPGPKGPTGEPGKPGEKGNVGLAGPRGAPGPEGNNGAQGPPGVTGNQGAKGETGPAGPPGFQGLPGPSGPAGEAGKPGERGLHGEFGVPGPAGPRGERGLPGESGAVGPAGPIGSRGPSGPPGPDGNKGEPGNVGPAGAPGPAGPGGIPGERGVAGVPGGKGEKGAPGLRGDTGATGRDGARGLPGAIGAPGPAGGAGDRGEGGPAGPAGPAGARGIPGERGEPGPVGPSGFAGPPGAAGQPGAKGERGPKGPKGETGPTGAIGPIGASGPPGPVGAAGPAGPRGDAGPPGMTGFPGAAGRVGPPGPAGITGPPGPPGPAGKDGPRGLRGDVGPVGRTGEQGIAGPPGFAGEKGPSGEAGAAGPPGTPGPQGILGAPGILGLPGSRGERGLPGIAGATGEPGPLGVSGPPGARGPSGPVGSPGPNGAPGEAGRDGNPGNDGPPGRDGAPGFKGERGAPGNPGPSGALGAPGPHGQVGPSGKPGNRGDPGPVGPVGPAGAFGPRGLAGPQGPRGEKGEPGDKGHRGLPGLKGHNGLQGLPGLAGQHGDQGPPGNNGPAGPRGPPGPSGPPGKDGRNGLPGPIGPAGVRGSHGSQGPAGPPGPPGPPGPPGPNGGGYEVGFDAEYYR'},\n",
        "#     'COL1A2-Gal_gal?': {'species': 'Gallus gallus','offset': -2, 'sequence': 'AGSVGMGYPPQPISGFPGPPGPSGPPGPPGHAGPPGSNGYQGPPGEPGQPGPSGPPGPAGMIGPAGPPGKDGEPGRPGRNGDRGIPGLPGHKGHPGMPGMPGMKGARGFDGKDGAKGDSGAPGPKGEAGQPGANGSPGQPGPRGPTGERGRPGNPGGPGAHGKDGAPGAAGPPGPPGPPGTAGFPGSPGFKGEAGPPGPAGASGSPGERGEPGPQGQAGPPGPQGPPGRAGSPGNKGEMGPSGIPGAPGLPGGRGLPGPPGTSGNPGAKGTPGEPGKNGAKGDPGPKGERGENGTPGAPGPPGEEGKRGANGEPGQNGVPGTPGERGSPGFRGLPGSNGLPGEKGPAGERGSPGPPGPSGPAGDRGQDGGPGLPGMRGLPGIPGSPGSDGKPGPPGNQGEPGRSGPPGPAGPRGQPGVMGFPGPKGNEGAPGKNGERGPGGPPGTPGPAGKNGDVGLPGPPGPAGPAGDRGEPGPSGSPGLQGLPGGPGPAGENGKPGEPGPKGDIGGPGFPGPKGENGIPGERGAQGPPGPTGARGGPGPAGSEGAKGPPGPPGAPGGTGLPGLQGMPGERGASGSPGPKGDKGEPGGKGADGLPGARGERGNVGPIGPPGPAGPPGDKGETGPAGAPGPAGSRGGPGERGEQGLPGPAGFPGAPGQNGEPGGKGERGPPGLRGEAGPPGAAGPQGGPGAPGPPGPQGVKGERGSPGGPGAAGFPGARGLPGPPGNNGSPGPPGNAGPPGKDGPPGPPGNTGPPGGSGPPGLRGEPGAPGEKGPPGARGERGTPGDPGPQGIIGSRGSTGLPGPRGLPGPAGMAGGKGEDGKPGVNGVPGERGAPGPQGPMGQRGLPGEPGRDGNPGSDGSPGRDGSPGGKGDRGESGPPGVPGPPGHPGPAGNNGAPGKAGERGFQGPPGPPGSAGPAGARGPAGPQGPRGDKGETGERGSAGIKGHRGFPGTPGLPGPPGPLGPQGAIGSPGASGARGPPGPAGPPGKDGRGGYPGPIGPPGPRGNRGESGPAGPPGQPGLPGPSGPPGPCCGGGVASLGAGEK'},\n",
        "#     'COL1A2-Ayf_COL1A2': {'species': 'Ayf xxx','offset': -0, 'sequence': ' GPMGLMGPRGPPGASGPPGPPGFQGVPGEPGEPGQTGPQGPRGPPGPPGKAGEDGHPGKPGRPGERGVAGPQGARGFPGTPGLPGFKGIRGHNGLDGQKGQPGTPGTKGEPGAPGENGTPGQPGARGLPGERGRVGAPGPAGARGSDGSAGPTGPAGPIGAAGPPGFPGAPGAKGEIGPAGNVGPTGPAGPRGEIGLPGSSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPPGPAGPSGARGLVGEPGPAGAKGESGNKGEPGAAGPPGPPGPSGEEGKRGSNGEPGSAGPPGPAGLRGVPGSRGLPGADGRAGVMGPAGNRGASGPVGAKGPNGDGGRPGEPGLMGPRGLPGQPGSPGPAGKEGPVGFPGADGRVGPIGPAGNRGEPGNIGFPGPKGPTGEPGKPGEKGNVGLAGPRGAPGPEGNNGAQGPPGVTGNQGAKGEQGPAGPPGFQGLPGPSGPAGEAGKPGERGLHGEFGVPGPAGPRGERGPPGESGAVGPAGPIGSRGPSGPPGPDGNKGEPGNVGAAGGPGPAGPGGIPGERGVAGVPGGKGEKGAPGLRGDTGATGRDGARGLPGAIGAPGPAGGAGDRGEGGPAGPAGPAGARGIPGERGEPGPVGPSGFAGPPGAAGQPGAKGERGPKGPKGETGPTGAIGPIGASGPPGPAGAAGPAGPRGDAGPPGMTGFPGAAGRVGPPGPAGITGPPGPPGPAGKDGARGLRGDVGPVGRTGEQGIAGPPGFAGEKGPSGEAGAPGPPGTPGPQGILGAPGILGLPGSRGERGLPGISGATGEPGPLGVSGPPGARGPSGPVGSPGPNGAPGEAGRDGNPGNDGPPGRDGAPGFKGERGAPGSPGPSGALGAPGPHGQVGPAGKPGNRGDPGPAGHVGPAGAFGPRGLAGPQGPRGEKGEPGDKGHRGLPGLKGHNGLQGLPGLAGQHGDQGPPGNNGPAGPRGPPGPSGPPGKDGRNGLPGPIGPAGARGSHGSQGPAGPPGPPGPPGPPGPNGGGYEVGYDAEYYRADQPSLRPK'},\n",
        "#     'COL1A2-Com_COL1A2': {'species': 'Com xxx','offset': -0, 'sequence': ' GPMGLMGPRGPPGASGPPGPPGFQGLPGEPGEPGQTGPQGPRGPPGPPGKAGEDGHPGKPGRPGERGVAGPQGARGFPGTPGLPGFKGIRGHNGLDGQKGQPGTPGSKGEPGAPGENGTPGQPGARGLPGERGRIGAPGPAGARGSDGSTGPTGPAGPIGAAGPPGFPGAPGAKGEIGPAGNVGPTGPAGPRGEIGLPGSSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGILGPPGPAGPSGARGLVGEPGPAGAKGESGMKGEPGAAGPAGPPGPSGEEGKRGSPGEPGSAGPPGPAGLRGVPGSRGLPGADGRAGVMGPAGNRGASGPVGAKGPSGDAGRPGEPGLMGPRGLPGQPGSPGPAGKEGPVGFPGADGRVGPIGPAGNRGEPGNIGFPGPKGPTGEPGKPGEKGNVGLAGPRGAPGPEGNNGAQGPPGVTGNPGGKGETGPAGPPGFQGLPGPSGPAGEAGKPGERGLHGEFGVPGPAGPRGERGPPGESGAVGPAGAIGSRGPSGPPGPDGNKGEPGNVGAAGAPGPAGPGGIPGERGVAGVPGGKGEKGAPGLRGDTGATGRDGARGLPGAIGAPGPAGGAGDRGEGGPAGPAGPAGARGIPGERGEPGPVGPNGFAGPPGAAGQPGAKGERGPKGPKGESGPTGAIGPIGASGPPGPVGAAGPAGPRGDAGPPGMTGFPGAAGRVGPPGPAGITGPPGPPGPAGKDGPRGLRGDVGPVGRTGEQGIAGPPGFAGEKGPSGEAGAAGPPGTPGPQGILGAPGILGLPGSRGERGLPGISGATGEPGPLGVSGPPGARGPSGPVGSPGPNGAPGEAGRDGNPGNDGPPGRDGAPGFKGERGAPGSPGPSGALGAPGPHGQVGPSGKPGNRGEPGPAGAVGPAGAFGPRGLAGPQGPRGEKGLPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGPPGNTGPAGPRGPPGPSGPPGKDGRNGLPGPIGPAGVRGSHGSQGPAGPPGPPGPPGPPGPNGGGYEVGFDAEYYRADQPSLRPK'},\n",
        "#     'COL1A2-Stc_COL1A2': {'species': 'Struthio camelus','offset': -0, 'sequence': ' GPMGLMGPRGPPGASGPPGPPGFQGLPGEPGEPGQTGPQGPRGPPGPPGKAGEDGHPGKPGRPGERGVAGPQGARGFPGTPGLPGFKGIRGHNGLDGQKGQPGTPGSKGEPGAPGENGTPGQPGARGLPGERGRIGAPGPAGARGSDGSTGPTGPAGPIGAAGPPGFPGAPGAKGEIGPAGNVGPTGPAGPRGEIGLPGSSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGILGPPGPAGPSGARGLVGEPGPAGAKGESGMKGEPGAAGPAGPPGPSGEEGKRGSPGEPGSAGPPGPAGLRGVPGSRGLPGADGRAGVMGPAGNRGASGPVGAKGPSGDAGRPGEPGLMGPRGLPGQPGSPGPAGKEGPVGFPGADGRVGPIGPAGNRGEPGNIGFPGPKGPTGEPGKPGEKGNVGLAGPRGAPGPEGNNGAQGPPGVTGNPGGKGETGPAGPPGFQGLPGPSGPAGEAGKPGERGLHGEFGVPGPAGPRGERGPPGESGAVGPAGAIGSRGPSGPPGPDGNKGEPGNVGAAGAPGPAGPGGIPGERGVAGVPGGKGEKGAPGLRGDTGATGRDGARGLPGAIGAPGPAGGAGDRGEGGPAGPAGPAGARGIPGERGEPGPVGPNGFAGPPGAAGQPGAKGERGPKGPKGESGPTGAIGPIGASGPPGPVGAAGPAGPRGDAGPPGMTGFPGAAGRVGPPGPAGITGPPGPPGPAGKDGPRGLRGDVGPVGRTGEQGIAGPPGFAGEKGPSGEAGAAGPPGTPGPQGILGAPGILGLPGSRGERGLPGISGATGEPGPLGVSGPPGARGPSGPVGSPGPNGAPGEAGRDGNPGNDGPPGRDGAPGFKGERGAPGSPGPSGALGAPGPHGQVGPSGKPGNRGEPGPAGAVGPAGAFGPRGLAGPQGPRGEKGLPGDKGPRGLPGLKGHNGLQGLPGLAGQHGDQGPPGNTGPAGPRGPPGPSGPPGKDGRNGLPGPIGPAGVRGSHGSQGPAGPPGPPGPPGPPGPNGGGYEVGFDAEYYRADQPSLRPK'},\n",
        "#     'COL1A2-EquxPrez': {'species': 'EquusxPrezwal','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPAGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQAGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLTGAKGAAGLPGVAGAPGLPGPRGIPGPAGAAGATGARGLVGEPGPAGSKGESGNKGEPGAAGPQGPPGPSGEEGKRGPNGEPGSTGPAGPPGLRGSPGSRGLPGADGRAGVMGPAGSRGASGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGEPGKPGDKGHAGLAGARGAPGPDGNNGAQGPPGPQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEVGKPGERGLPGEFGLPGPAGARGERGPPGESGAAGPAGPIGSRGPSGPPGPDGNKGEPGVLGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGEIGNPGRDGARGAPGAVGAPGPAGANGDRGEAGAAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGVTGFPGAAGRTGPPGPSGISGPPGPPGAAGKEGLRGPRGDQGPVGRAGETGASGPPGFAGEKGPSGEPGTAGPPGTPGPQGLLGAPGILGLPGSRGERGLPGVAGSLGEPGPLGIAGPPGARGPPGAVGAPGVNGAPGEAGRDGNPGSDGPPGRDGQPGHKGERGYPGNAGPVGAVGAPGPHGPVGPTGKHGHRGEPGPVGSVGPVGAVGPRGPSGPQGVRGDKGEPGDKGPRGLPGIKGHNGLQGLPGLAGQHGDQGAPGSVGPAGPRGPAGPTGPVGKDGRSGQPGTVGPAGVRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGYDGDFYRA'},\n",
        "#     'COL1A2-Bub_Bub': {'species': 'Bubalus bubalis','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGPAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPSGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPTGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGVRGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAVGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPAGISGPPGPPGPAGKEGLRGPRGDQGPVGRTGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGLPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGSVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYDFGFDGDFYRA'},\n",
        "#     'COL1A2-Bos_ind': {'species': 'Bos indicus','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPXGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A2-Bis_bis': {'species': 'Bison bison','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A2-Bosxbos': {'species': 'BosxBos','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A2-Phy_cas': {'species': 'Physter catodon','offset': -8, 'sequence': 'QYDGKGVGIGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKSGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGTPGVKGEPGAPGENGIPGQVGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPLGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGSRGEVGLPGVSGPVGPPGNPGVNGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGDSGNKGEPGAVGPTGPPGPSGEEGKRGTTGEIGSAGPPGPPGLRGNPGSRGLPGADGRAGVMGLHGSRGGTGPAGVRGPSGDSGRPGEPGLMGPRGFPGSPGNAGPAGKEGPTGLPGIDGRPGPIGPAGTRGEPGNIGFPGPKGPTGDPGKNGEKGHAGLAGPRGAPGPDGNNGAQGPPGLQGVSGGKGEQGPAGPPGFQGLPGPAGTAGEAGKAGERGIPGEFGLPGPAGPRGERGPPGESGAAGPAGPIGSRGPSGPAGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDVGSHGRDGARGAPGAVGAPGPAGANGDRGEAGPAGAAGPAGSRGSPGERGEVGPAGPNGFAGSAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPAGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGITGPPGPPGPAGKEGLRGPRGDQGPVGRTGETGASGPPGFVGEKGPSGEPGTAGSPGTPGPQGLLGSPGFLGLPGSRGERGLPGVAGSLGEPGPLGIAGPTGARGPPGAVGNPGVNGAPGEAGRDGNPGSDGPPGRDGQAGHKGERGYPGNAGPTGTAGAPGPQGPVGPTGKHGNRGESGPSGPVGLAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGTVGPAGPRGPAGPSGPSGKDGRTGHPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPAGPSGGGYDFGFEGDFYRA'},\n",
        "#     'COL1A2-Bis_bis': {'species': 'Bison bison','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGAAGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A2-Bos_mut': {'species': 'Bubalus bubalis','offset': -7, 'sequence': 'QFDAKGGGPGPMGLMGPRGPPGASGAPGPQGFQGPPGEPGEPGQTGPAGARGPPGPPGKAGEDGHPGKPGRPGERGVVGPQGARGFPGTPGLPGFKGIRGHNGLDGLKGQPGAPGVKGEPGAPGENGTPGQTGARGLPGERGRVGAPGPAGARGSDGSVGPVGPAGPIGSAGPPGFPGAPGPKGELGPVGNPGPAGPAGPRGEVGLPGLSGPVGPPGNPGANGLPGAKGAAGLPGVAGAPGLPGPRGIPGPVGASGATGARGLVGEPGPAGSKGESGNKGEPGAVGQPGPPGPSGEEGKRGSTGEIGPAGPPGPPGLRGNPGSRGLPGADGRAGVMGPAGSRGATGPAGVRGPNGDSGRPGEPGLMGPRGFPGSPGNIGPAGKEGPVGLPGIDGRPGPIGPAGARGEPGNIGFPGPKGPSGDPGKAGEKGHAGLAGARGAPGPDGNNGAQGPPGLQGVQGGKGEQGPAGPPGFQGLPGPAGTAGEAGKPGERGIPGEFGLPGPAGARGERGPPGESGAAGPTGPIGSRGPSGPPGPDGNKGEPGVVGAPGTAGPSGPSGLPGERGAAGIPGGKGEKGETGLRGDIGSPGRDGARGAPGAIGAPGPAGANGDRGEAGPAGPAGPAGPRGSPGERGEVGPAGPNGFAGPAGAAGQPGAKGERGTKGPKGENGPVGPTGPVGAAGPSGPNGPPGPAGSRGDGGPPGATGFPGAAGRTGPPGPSGISGPPGPPGPAGKEGLRGPRGDQGPVGRSGETGASGPPGFVGEKGPSGEPGTAGPPGTPGPQGLLGAPGFLGLPGSRGERGLPGVAGSVGEPGPLGIAGPPGARGPPGNVGNPGVNGAPGEAGRDGNPGNDGPPGRDGQPGHKGERGYPGNAGPVGAAGAPGPQGPVGPVGKHGNRGEPGPAGAVGPAGAVGPRGPSGPQGIRGDKGEPGDKGPRGLPGLKGHNGLQGLPGLAGHHGDQGAPGAVGPAGPRGPAGPSGPAGKDGRIGQPGAVGPAGIRGSQGSQGPAGPPGPPGPPGPPGPSGGGYEFGFDGDFYRA'},\n",
        "#     'COL1A1-Bos_mut': {'species': 'Bos mutus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEGGPQGPRGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGSKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGAPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGAPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPAGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGSKGPRGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGAKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPAGPPGPPGPPGPPGPPSGGYDLSFLPQPPQEKAHDGGR'},\n",
        "#     'COL1A1-Cam_fer': {'species': 'Bos taurus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGSSGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGVQGPPGPAGEEGKRGARGEPGPAGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGETGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGPPGPIGSVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPVGKEGSKGPRGETGPAGRPGEVGPPGPPGPAGEKGAPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGPNGERGPPGPMGPPGLAGPPGESGREGAPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKSGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGAPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAGGDGR'},\n",
        "#     'COL1A1-Vul_vul': {'species': 'Bos taurus','offset': 0, 'sequence': 'GPMGPSGPRGLPGPPGAPGPQGFQGPPGEPGEPGASGPMGPRGPPGPPGKNGDDGEAGKPGRPGERGPPGPQGARGLPGTAGLPGMKGHRGFSGLDGAKGDAGPAGPKGEPGSPGENGAPGQMGPRGLPGERGRPGAPGPAGARGNDGATGAAGPPGPTGPAGPPGFPGAVGAKGEAGPQGARGSEGPQGVRGEPGPPGPAGAAGPAGNPGADGQPGAKGANGAPGIAGAPGFPGARGPSGPQGPSGPPGPKGNSGEPGAPGNKGDTGAKGEPGPTGIQGPPGPAGEEGKRGARGEPGPTGLPGPPGERGGPGSRGFPGADGVAGPKGPAGERGSPGPAGPKGSPGEAGRPGEAGLPGAKGLTGSPGSPGPDGKTGPPGPAGQDGRPGPPGPPGARGQAGVMGFPGPKGAAGEPGKAGERGVPGPPGAVGPAGKDGEAGAQGPPGPAGPAGERGEQGPAGSPGFQGLPGPAGPPGEAGKPGEQGVPGDLGAPGPSGARGERGFPGERGVQGPPGPAGPRGANGAPGNDGAKGDAGAPGAPGSQGAPGLQGMPGERGAAGLPGPKGDRGDAGPKGADGSPGKDGVRGLTGPIGPPGPAGAPGDKGEAGPSGPAGPTGARGAPGDRGEPGPPGPAGFAGPPGADGQPGAKGEPGDAGAKGDAGPPGPAGPTGPPGPIGNVGAPGPKGARGSAGPPGATGFPGAAGRVGPPGPSGNAGPPGPPGPAGKEGGKGPRGETGPAGRPGEVGPPGPPGPAGEKGSPGADGPAGAPGTPGPQGIAGQRGVVGLPGQRGERGFPGLPGPSGEPGKQGPSGASGERGPPGPMGPPGLAGPPGESGREGSPGAEGSPGRDGSPGPKGDRGETGPAGPPGAPGAPGAPGPVGPAGKNGDRGETGPAGPAGPIGPVGARGPAGPQGPRGDKGETGEQGDRGIKGHRGFSGLQGPPGPPGSPGEQGPSGASGPAGPRGPPGSAGSPGKDGLNGLPGPIGPPGPRGRTGDAGPVGPPGPPGPPGPPGPPSGGFDFSFLPQPPQEKAHDGGR'},\n",
        "#     'COL3A1-BosxBos': {'species': 'BosxBos','offset': -17, 'sequence': 'QYEAYDVKSGIAGGGIAGYPGPAGPPGPPGPPGTSGHPGAPGAPGYQGPPGEPGQAGPAGPPGPPGAIGPSGPAGKDGESGRPGRPGERGFPGPPGMKGPAGMPGFPGMKGHRGFDGRNGEKGETGAPGLKGENGVPGENGAPGPMGPRGAPGERGRPGLPGAAGARGNDGARGSDGQPGPPGPPGTAGFPGSPGAKGEVGPAGSPGSSGAPGQRGEPGPQGHAGAPGPPGPPGSNGSPGGKGEMGPAGIPGAPGLIGARGPPGPPGTNGVPGQRGAAGEPGKNGAKGDPGPRGERGEAGSPGIAGPKGEDGKDGSPGEPGANGLPGAAGERGVPGFRGPAGANGLPGEKGPPGDRGGPGPAGPRGVAGEPGRDGLPGGPGLRGIPGSPGGPGSDGKPGPPGSQGETGRPGPPGSPGPRGQPGVMGFPGPKGNDGAPGKNGERGGPGGPGPQGPAGKNGETGPQGPPGPTGPSGDKGDTGPPGPQGLQGLPGTSGPPGENGKPGEPGPKGEAGAPGIPGGKGDSGAPGERGPPGAGGPPGPRGGAGPPGPEGGKGAAGPPGPPGSAGTPGLQGMPGERGGPGGPGPKGDKGEPGSSGVDGAPGKDGPRGPTGPIGPPGPAGQPGDKGESGAPGVPGIAGPRGGPGERGEQGPPGPAGFPGAPGQNGEPGAKGERGAPGEKGEGGPPGAAGPAGGSGPAGPPGPQGVKGERGSPGGPGAAGFPGGRGPPGPPGSNGNPGPPGSSGAPGKDGPPGPPGSNGAPGSPGISGPKGDSGPPGERGAPGPQGPPGAPGPLGIAGLTGARGLAGPPGMPGARGSPGPQGIKGENGKPGPSGQNGERGPPGPQGLPGLAGTAGEPGRDGNPGSDGLPGRDGAPGAKGDRGENGSPGAPGAPGHPGPPGPVGPAGKSGDRGETGPAGPSGAPGPAGSRGPPGPQGPRGDKGETGERGAMGIKGHRGFPGNPGAPGSPGPAGHQGAVGSPGPAGPRGPVGPSGPPGKDGASGHPGPIGPPGPRGNRGERGSEGSPGHPGQPGPPGPPGAPGPCCGAGGVAAIAGVGAEKAGGFAPYYG'},\n",
        "#     'AHSG-Bos_tau' : {'species': 'Bos taurus','offset': 0, 'sequence':'IPLDPVAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTQHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNDSRVVHAVEVALATFNAESNGSYLQLVEISRAQFVPLPVSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVRVTCTLFQTQPVIPQPQPDGAEAEAPSAVPDAAGPTPSAAGPPVASVVVGPSVVAVPLPLHRAHYDLRHTFSGVASVESSSGEAFHVGKTPIVGQPSIPGGPVRLCPGRIRYFKI'},\n",
        "#     'AHSG-Bub_bub' : {'species': 'Bubalus bubalis','offset': 0, 'sequence':'IPLDPVAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNDSRVVHAVEVALATFNAQSNGSYLQLVEISRAQFVPLPASVSVEFAVAATDCIAKDVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVAVTCTLFQTQPVILQPQPDGAEAGAPSAVPDAAGPAPSAAGPPVASVVVGPSVVAVPLPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVRLCPGRIRYFKI'},\n",
        "#     'AHSG-Equ_cab' : {'species': 'Equus caballus','offset': 1, 'sequence':' LPNGLSPAYRQLNCDDPETEQAALLAVDYINSHIHQGYKHVLNQIDKVQVWAQPTGESFKLEIDTLETTCHALDPTPLANCSVRQLTQHAVEGDCDVRLLKQNGQFSVSFVKCKSSPDSAEDVRKVCLDCPLLAPLNDTRVVHAVEAALAAFNAQNNGSYFQLVEISRAQLVPLPVSVHVEFAVAATDCVAKEVIDPAKCNLLAEKQYGFCKATLTEKVGGEDVAVTCTVFQTQPVVLLPQPDGPDVGVPGPVADAVTPAPSPADLPVASLVVGPVVVAASQLPPPVHRAHYDLRHAFAGVGSGESASGEAFHVEKPPKVAHPNTAAAAGPVVRPCPGRIRYFKII'},\n",
        "#     'AHSG-Ovi_are' : {'species': 'Ovis aries','offset': 1, 'sequence':'IPLDPIAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLVNCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNNSQVVHAAEVALATFNAQNNGSYFQLVEISRAQFVPLPGSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVTVTCTLFQTQPVIPQPQPEGAEAGAPSAVPDAAVPDAAVPAPSAAGLPVGSVVAGPSVVAVPLPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVHLCPGRIRYFKI'},\n",
        "#     'AHSG-Cap_hir' : {'species': 'Capra hircus','offset': 1, 'sequence':'IPLDPIAGYKEPACDDPDTEQAALAAVDYINKHLPRGYKHTLNQIDSVKVWPRRPTGEVYDIEIDTLETTCHVLDPTPLANCSVRQQTEHAVEGDCDIHVLKQDGQFSVLFTKCDSSPDSAEDVRKLCPDCPLLAPLNNSQVVHAAEVALATFNAQNNGSYFQLVEISRAQFVPLPVSVSVEFAVAATDCIAKEVVDPTKCNLLAEKQYGFCKGSVIQKALGGEDVAVTCTLFQTQPVIPQPQPEGAEAGAPSAVPDAAVPAPSAAGLPVGSVVAGPSVVAVALPLHRAHYDLRHTFSGVASVESASGEAFHVGKTPIVGQPSVPGGPVHLCPGRIRYFKI'},\n",
        "#     'AHSG-Equ_asn' : {'species': 'Equus asinus','offset': 1, 'sequence':' LPNGLSPAYRQLNCDDPETEQAALLAVDYINSHIHQGYKHVLNQIDKVQVWAQPTGESFKLEIDTLETTCHALDPTPLANCSVRQLTQHAVEGDCDVRLLKQNGQFSVSFVKCKSSPDSAEDVRKVCLDCPLLAPLNDTRVVHAVEAALAAFNAQNNGSYFQLVEISRAQLVPLPVSVHVEFAVAATDCVAKEVIDPAKCNLLAEKQYGFCKATLTEKVGGEDVAVTCTVFQTQPVVPLPQPDGPDVGVPGPVADAATPEPSPADLPVASLVVGPVVVAAPQLPPPVHRAHYDLRHAFAGVGSGESASGEAFHVEKPPKVAHPNTAAAAGPVVRPCPGRIRYFKI'},\n",
        "#     'yNFM-Pse_aer' : {'species': 'Pseudomonas aeruginosa','offset': 1, 'sequence':'MALAAPFVGRLSGGRERSRVIAFAASLLCVPMLLAAHAGSFAAFLLWRFAEGMLIPLLFASSVAYIGERWSGGAVIELTSLYVAGTILGGFAGRFLTGLLTEEWDWRVAFRILAGLTLLIAVAIRALLPANSGRHAPAPGKRPLVSRELFRAPLLSSYAVGFCVLFSQVAMFTYIGIHLSRPPYSLDTAQLGSIYAVFLLALLVIPASGRLARTRPHRQLLLGASLLGVCGSLLTLAPGLPTILVGLALSATGVFLAQSTVNAFTASNAGTDKAGAVGLYLTCYYAGGSLGAVLPAPFWERWGWPGCLGLILLAQLLPLLLTRYGWSAPQAERAALRASQGRP'},\n",
        "#     'PIV-Pse_aer' : {'species': 'Pseudomonas keratitis','offset': 1, 'sequence':'MHKRTYLNACLVLALAAGASQASAAPGASEMAGDVAVLQASPASTGHARFANPNAATSAAGIHFAAPPARRVARAAPLAPKPGTPLQVGVGLKTATPEIDLATLEWIDTPDGRHTARFPISAAGAASLRAAIRLETRSGSLPDDVLLHFAGAGKEIFEASGKDLSXNRPYWSPVIEGDTLTVELVLPAXLQPGDLRLSVPQVSYFADSLYKAGYRDGFGASGSCEVDAVCATQSGTRAYDNATAAVAKMVFTSSADGGSYICTGTLLNNGNSPKRQLFWSAAHCIEDQATAATLQTIWFYNTTQCYGDASTINQSVTVLTGGANILHRDAKRDTLLLELKRTPPAGVFYQGWSATPIANGSLGHDIHHPRGDAKKYSQGNVSAVGVTYDGHTALTRVDWPSAVVEGGSSGSGLLTVAGDGSYQLRGGLYGGPSYCGAPTSQRNDYFSDFSGVYSQISRYFAP'}\n",
        "\n",
        "\n",
        "#     }\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Enzymatic Cleavage Sites"
      ],
      "metadata": {
        "id": "4dPQ8oEPuRyd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def find_cleavage_sites_with_offset(collagen_sequences: dict) -> dict:\n",
        "    \"\"\"\n",
        "    Find the positions of the letters 'K' and 'R' in given protein sequences stored in a dictionary,\n",
        "    adjusting for an offset provided for each sequence.\n",
        "\n",
        "    Parameters:\n",
        "        collagen_sequences (dict): A dictionary where each key is a unique identifier for a protein,\n",
        "                                   and each value is another dictionary with keys 'offset' and 'sequence'\n",
        "                                   representing the offset from the start of the protein and the amino\n",
        "                                   acid sequence, respectively.\n",
        "\n",
        "    Returns:\n",
        "        dict: A dictionary where each key is the protein name (GNC) and each value is a list of adjusted\n",
        "              positions (1-based) where either 'K' or 'R' occurs in the sequence, considering the offset.\n",
        "    \"\"\"\n",
        "    cleavage_sites_results = {}\n",
        "\n",
        "    for key, value in collagen_sequences.items():\n",
        "        # Use the new extract_gene_identifier function to get the gene name\n",
        "        protein_name = extract_gene_identifier(key)\n",
        "\n",
        "        sequence = value['sequence']\n",
        "        offset = value['offset']\n",
        "        cleavage_sites = []\n",
        "\n",
        "        for i, amino_acid in enumerate(sequence):\n",
        "            if amino_acid in ['K', 'R']:\n",
        "                position_with_offset = i + 1 + offset\n",
        "                cleavage_sites.append(position_with_offset)\n",
        "\n",
        "        cleavage_sites_results[protein_name] = cleavage_sites\n",
        "\n",
        "    return cleavage_sites_results\n",
        "\n",
        "\n",
        "def extract_gene_identifier(full_gene_name):\n",
        "    \"\"\"\n",
        "    Extracts the gene identifier from the full gene name.\n",
        "\n",
        "    Parameters:\n",
        "    - full_gene_name (str): The full gene name, e.g., 'COL3A1-Bos_mut'.\n",
        "\n",
        "    Returns:\n",
        "    - str: The extracted gene identifier, e.g., 'COL3A1'.\n",
        "    \"\"\"\n",
        "    # Split the full gene name at the hyphen and return the first part\n",
        "    return full_gene_name.split('-')[0]\n"
      ],
      "metadata": {
        "id": "Y2hTpTb7-vLn"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "GKnxUUhickhM",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a4070e10-cbb4-4672-a511-6e2ba66f9032"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "COL1A1: [9, 42, 50, 59, 62, 66, 75, 87, 90, 99, 108, 126, 132, 134, 144, 174, 183, 192, 219, 237, 252, 264, 270, 290, 291, 294, 309, 315, 327, 333, 342, 350, 360, 374, 386, 396, 408, 416, 420, 434, 453, 479, 498, 501, 507, 519, 531, 555, 564, 567, 573, 581, 585, 603, 618, 624, 648, 657, 684, 687, 704, 725, 729, 732, 740, 756, 780, 789, 792, 806, 816, 836, 848, 855, 858, 884, 888, 906, 915, 918, 927, 930, 933, 963, 974, 990, 992, 1030, 1036]\n",
            "COL1A2: [-3, 12, 45, 53, 62, 65, 69, 78, 90, 93, 102, 111, 129, 135, 137, 147, 177, 195, 222, 240, 255, 267, 273, 293, 294, 312, 318, 326, 336, 345, 353, 363, 377, 389, 399, 411, 419, 423, 432, 456, 482, 486, 501, 504, 522, 534, 558, 567, 570, 576, 584, 588, 606, 621, 627, 651, 654, 657, 660, 690, 707, 728, 732, 735, 743, 759, 792, 795, 819, 839, 851, 858, 861, 887, 891, 909, 918, 921, 927, 930, 936, 966, 977, 980, 993, 1031]\n",
            "COL3A1: [-9, 59, 65, 68, 72, 81, 93, 96, 101, 105, 114, 132, 138, 140, 150, 156, 180, 198, 225, 243, 258, 266, 270, 276, 279, 291, 296, 315, 321, 333, 339, 348, 356, 366, 380, 392, 402, 414, 422, 426, 440, 459, 485, 492, 504, 513, 525, 537, 561, 570, 573, 587, 591, 609, 624, 630, 654, 657, 663, 690, 693, 708, 731, 753, 762, 786, 798, 807, 812, 822, 842, 854, 861, 864, 890, 894, 912, 921, 924, 930, 936, 939, 969, 980, 996, 999, 1002, 1042]\n",
            "COL5A1: [0, 13, 21, 22, 25, 32, 35, 44, 63, 65, 66, 90, 102, 112, 120, 123, 130, 131, 140, 141, 142, 145, 149, 166, 187, 207, 244, 250, 257, 285, 333, 370, 409, 412, 415, 458, 463, 485, 495, 514, 518, 529, 542, 554, 565, 568, 569, 571, 578, 587, 590, 602, 605, 623, 632, 641, 647, 668, 704, 709, 727, 734, 752, 755, 761, 764, 767, 770, 779, 785, 788, 797, 806, 808, 824, 826, 838, 842, 857, 860, 865, 869, 872, 878, 881, 884, 889, 893, 908, 923, 931, 941, 949, 974, 994, 998, 1009, 1016, 1022, 1034, 1052, 1066, 1079, 1085, 1093, 1118, 1127, 1130, 1133, 1163, 1172, 1178, 1199, 1217, 1244, 1265, 1268, 1271, 1286, 1295, 1325, 1351, 1352, 1363, 1367, 1370, 1381, 1393, 1400, 1433, 1439, 1442, 1460, 1463, 1475, 1502, 1505, 1514, 1542, 1544, 1545, 1579, 1586, 1587, 1597, 1600, 1624, 1628, 1645, 1646, 1651, 1657, 1668, 1669, 1671, 1692, 1722, 1725, 1743, 1752, 1753, 1757, 1766, 1786]\n",
            "COL6A1: [-7, -6, 7, 10, 13, 16, 28, 30, 46, 52, 55, 63, 64, 72, 79, 81, 84, 91, 97, 112, 118, 123, 126, 127, 136, 139, 154, 157, 163, 166, 171, 172, 178, 187, 190, 199, 208, 211, 214, 220, 231, 235, 244, 253, 258, 267, 273, 285, 291, 296, 297, 300, 315, 318, 345, 355, 356, 381, 390, 396, 399, 405]\n",
            "COL6A2: [6, 9, 12, 15, 27, 29, 45, 51, 54, 62, 63, 71, 78, 80, 83, 90, 96, 111, 117, 122, 125, 126, 135, 138, 153, 156, 162, 165, 170, 171, 177, 186, 189, 198, 207, 210, 213, 219, 230, 234, 243, 252, 257, 266, 272, 284, 290, 295, 296, 299, 314, 317, 344, 354, 355, 380, 389, 395, 398, 404, 425, 433, 437, 453, 458, 461, 464, 465, 467, 469, 479, 483, 490, 511, 524, 529, 541, 575, 588, 596, 598, 599, 606, 607, 612, 613, 621, 631, 654, 678, 682, 683, 710, 711, 732, 743, 745, 758, 761]\n",
            "COL12A1: [-6, -4, 26, 48, 60, 111, 115, 116, 119, 124, 145, 149, 168, 170, 181, 200, 201, 205, 220, 230, 235, 244, 254, 257, 269, 274, 277, 340, 343, 357, 368, 384, 400, 413, 418, 425, 430, 450, 452, 460, 468, 478, 487, 488, 491, 505, 513, 519, 521, 527, 530, 535, 544, 550, 555, 557, 569, 573, 599, 611, 620, 621, 622, 628, 641, 659, 684, 715, 719, 722, 732, 741, 744, 746, 750, 758, 767, 768, 769, 779, 806, 810, 813, 823, 828, 832, 834, 854, 863, 897, 906, 925, 928, 943, 962, 964, 992, 995, 1005, 1009, 1014, 1019, 1023, 1025, 1029, 1034, 1045, 1046, 1062, 1067, 1069, 1078, 1080, 1083, 1086, 1096, 1107, 1110, 1120, 1121, 1140, 1146, 1187, 1204, 1208, 1211, 1216, 1226, 1227, 1240, 1249, 1251, 1252, 1264, 1278, 1283, 1289, 1291, 1293, 1294, 1303, 1313, 1314, 1316, 1328, 1336, 1358, 1372, 1388, 1391, 1394, 1406, 1408, 1418, 1419, 1426, 1429, 1434, 1437, 1459, 1463, 1509, 1511, 1514, 1552, 1562, 1565, 1567, 1584, 1586, 1587, 1591, 1593, 1600, 1606, 1608, 1614, 1664, 1678, 1689, 1734, 1737, 1747, 1751, 1766, 1773, 1776, 1778, 1798, 1806, 1808, 1828, 1832, 1834, 1836, 1842, 1845, 1857, 1866, 1869, 1895, 1918, 1924, 1928, 1932, 1947, 1959, 1969, 2015, 2019, 2023, 2026, 2050, 2070, 2071, 2086, 2106, 2125, 2127, 2141, 2145, 2205, 2217, 2222, 2228, 2230, 2238, 2245, 2281, 2283, 2287, 2308, 2312, 2315, 2335, 2338, 2367, 2371, 2378, 2389, 2391, 2396, 2399, 2405, 2407, 2416, 2417, 2421, 2430, 2436, 2437, 2468, 2472, 2485, 2516, 2526, 2548, 2551, 2578, 2596, 2599, 2611, 2618, 2621, 2635, 2644, 2652, 2655, 2666, 2669, 2684, 2687, 2690, 2691, 2711, 2713, 2720, 2721, 2725, 2755, 2758, 2761, 2773, 2801, 2805, 2816, 2835, 2840, 2847, 2867, 2870, 2876, 2883, 2886, 2889, 2899, 2903, 2916, 2932, 2950, 2958, 2974, 2980, 2983, 2991, 3004, 3014, 3022, 3029]\n",
            "COLunk: [-6, -5, -3, 16, 56, 65, 76, 87, 95, 121, 126, 148, 155, 171, 191, 210, 228, 229, 240, 250, 255, 259, 263, 279, 290, 291, 298, 299, 301, 318, 329, 336, 345, 358, 359, 370, 394, 410, 419, 422, 423, 444, 462, 484, 492, 523, 553, 568, 620, 621, 641, 646, 654, 659, 670, 674, 681, 689, 690, 692, 694, 697, 710, 715, 720, 736, 740, 749, 761, 786, 799, 813, 826, 833, 838, 849, 851, 860, 862, 881, 902, 912, 928, 934, 941, 942, 977, 989, 995, 1000, 1016, 1017, 1037, 1042, 1055, 1066, 1077, 1085, 1086, 1088, 1090, 1093, 1106, 1110, 1111, 1116, 1132, 1145, 1156, 1165, 1180, 1188, 1209, 1233, 1241, 1246, 1249, 1250, 1257, 1259, 1276, 1279, 1289, 1290, 1292, 1294, 1303, 1310, 1315, 1320, 1336, 1343, 1349, 1361, 1365, 1376, 1382, 1385, 1393, 1417, 1420, 1421, 1422, 1436, 1437, 1445, 1479, 1484, 1493, 1497, 1500, 1505, 1512, 1513, 1523, 1527, 1539, 1554, 1557, 1564, 1573, 1583, 1618, 1641, 1646, 1652, 1653, 1659, 1671, 1698, 1703, 1716, 1721, 1723, 1731, 1749, 1755, 1760, 1769, 1783, 1789, 1792, 1812, 1820, 1825, 1828, 1837, 1846, 1858, 1876, 1879, 1882, 1888, 1906, 1915, 1920, 1921, 1924, 1927, 1930, 1933, 1939, 1952, 1955, 1969, 1981, 1987, 1988, 1994, 2012, 2036, 2048, 2050, 2056, 2057, 2066, 2072, 2075, 2083, 2091, 2094, 2095, 2098, 2107, 2122, 2125, 2127, 2128, 2140, 2152, 2155, 2169, 2171, 2202, 2207, 2212, 2226, 2229, 2244, 2249, 2250, 2251, 2257, 2259, 2268, 2281, 2285, 2286, 2288, 2294, 2295, 2303, 2306, 2312, 2317, 2334, 2364, 2388, 2392, 2394, 2395, 2424, 2425, 2432, 2441, 2448, 2469, 2480, 2482, 2490, 2498, 2502, 2519, 2522, 2534, 2542, 2548, 2550, 2561, 2562, 2566, 2581, 2585, 2596, 2599, 2608, 2618, 2619, 2631, 2639, 2654, 2660, 2667, 2678, 2683, 2688, 2693, 2698, 2703, 2708, 2713, 2718, 2723, 2728, 2733, 2738, 2743, 2748, 2753, 2758, 2763, 2768, 2773, 2778, 2784, 2789, 2794, 2800, 2805, 2810, 2815, 2820, 2825, 2829, 2832, 2845, 2850, 2874, 2882, 2902, 2906, 2916, 2917, 2927, 2942, 2947, 2952, 2955, 2961, 2966, 2978, 2990, 2999, 3007]\n",
            "FibMod: [3, 6, 8, 9, 15, 26, 31, 33, 37, 44, 59, 65, 80, 81, 84, 89, 97, 102, 104, 136, 178, 199, 202, 210, 212, 249, 252, 253, 257, 259, 263, 273, 280, 283, 294, 324, 327, 340, 341, 369, 373, 378, 411, 428, 448, 453, 460, 461, 473]\n",
            "TRYUP: [43, 45, 49, 69, 89, 99, 107, 125, 139, 149, 170, 200, 202, 208]\n",
            "TPM1: [5, 6, 7, 12, 15, 21, 29, 30, 35, 37, 48, 49, 51, 59, 65, 70, 76, 77, 90, 91, 101, 105, 112, 118, 125, 128, 133, 136, 140, 149, 152, 160, 161, 167, 168, 178, 182, 189, 191, 198, 205, 263, 272, 279, 287, 291, 294, 300, 305, 307, 312, 318, 322, 325, 338, 340, 342]\n",
            "KS1: [8, 13, 17, 37, 38, 45, 46, 53, 75, 143, 147, 152, 159, 161, 173, 185, 201, 202, 208, 211, 213, 219, 229, 231, 238, 239, 250, 251, 260, 266, 278, 305, 317, 326, 328, 338, 354, 357, 365, 369, 371, 378, 379, 394, 400, 405, 415, 417, 422, 425, 434, 445, 454, 480, 521, 524, 529, 532, 538, 546, 567, 576, 598, 617, 647, 667, 668, 672, 686, 700, 708, 722, 736, 751]\n",
            "KS9: [4, 13, 29, 34, 46, 59, 63, 95, 154, 163, 170, 184, 191, 192, 199, 212, 224, 233, 240, 242, 250, 261, 271, 289, 290, 293, 317, 321, 327, 336, 338, 339, 368, 374, 390, 391, 396, 402, 404, 427, 444, 446, 449, 472, 478, 487, 490, 513, 579, 619]\n",
            "KS2: [2, 7, 11, 28, 29, 36, 37, 44, 64, 73, 81, 85, 174, 176, 178, 180, 184, 196, 198, 210, 222, 238, 239, 240, 245, 250, 256, 266, 268, 275, 276, 287, 288, 297, 303, 342, 354, 363, 365, 375, 385, 391, 394, 402, 406, 408, 415, 416, 431, 437, 440, 442, 454, 459, 462, 471, 482, 491, 517, 548, 587, 601, 615, 623, 632, 643, 644]\n",
            "KS10: [4, 9, 15, 35, 40, 59, 86, 147, 156, 163, 165, 177, 179, 184, 194, 198, 201, 207, 228, 235, 237, 245, 256, 257, 266, 284, 285, 292, 295, 322, 333, 334, 343, 345, 362, 369, 370, 386, 399, 422, 439, 441, 450, 464, 567, 580, 583]\n",
            "KS1c9: [4, 13, 29, 34, 46, 59, 63, 95, 154, 163, 170, 184, 191, 192, 199, 212, 224, 233, 240, 242, 250, 261, 271, 289, 290, 293, 317, 321, 327, 336, 338, 339, 368, 374, 390, 391, 396, 402, 404, 427, 444, 446, 449, 472, 478, 487, 490, 513, 579, 619]\n",
            "ALB: [2, 19, 23, 24, 28, 34, 36, 44, 65, 68, 75, 88, 97, 105, 117, 122, 130, 138, 140, 156, 160, 167, 168, 185, 197, 204, 209, 211, 218, 220, 222, 228, 232, 235, 241, 245, 248, 256, 263, 266, 280, 285, 297, 299, 304, 309, 317, 318, 336, 340, 343, 346, 359, 360, 371, 374, 386, 399, 408, 412, 420, 433, 436, 437, 451, 452, 459, 462, 463, 468, 482, 489, 495, 498, 507, 508, 521, 523, 544, 547, 548, 557, 559, 568, 580, 597, 603]\n",
            "AHSG: [12, 41, 49, 62, 86, 100, 103, 114, 116, 126, 127, 142, 170, 194, 201, 208, 214, 220, 291, 297, 318, 321, 336, 341, 343, 346]\n",
            "yNFM: [11, 16, 18, 20, 49, 70, 95, 108, 112, 126, 135, 142, 143, 148, 152, 182, 211, 214, 216, 219, 274, 302, 324, 334, 338, 343]\n",
            "PIV: [4, 5, 50, 71, 72, 75, 82, 94, 114, 118, 130, 134, 138, 155, 163, 169, 197, 212, 216, 238, 249, 275, 276, 329, 332, 333, 341, 342, 371, 375, 376, 397, 426, 443, 459]\n"
          ]
        }
      ],
      "source": [
        "# Merge sequence dictionaries\n",
        "all_sequences = {**sequences_dict,\n",
        "                # **non_sequences_dict\n",
        "                 }\n",
        "# Find cleavage sites considering the offsets\n",
        "cleavage_sites_with_offset = find_cleavage_sites_with_offset(all_sequences)\n",
        "\n",
        "for protein, sites in cleavage_sites_with_offset.items():\n",
        "    print(f\"{protein}: {sites}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Sample Names"
      ],
      "metadata": {
        "id": "tlfu4_3TWpk7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define a dictionary to map old names to new names\n",
        "rename_dict = {\n",
        "    'M3_20231109-1713': 'M3',\n",
        "    'M1_20231109-1625': 'M1',\n",
        "    'M4_20231109-1737': 'M4',\n",
        "    'P3_20231109-1848': 'P3',\n",
        "    'P4_20231109-1912': 'P4',\n",
        "    'P2_20231109-1824': 'P2',\n",
        "    'P1_20231109-1801': 'P1',\n",
        "    'M2_20231109-1649': 'M2',\n",
        "    '20230614-0225_QEHF3_1005194_ONJ_TR_MatthewEvo_13_LLM_C_WhiteP_02_OBi.raw' :'02_OBi_Pak_P',\n",
        "    '20230614-0314_QEHF3_1005195_ONJ_TR_MatthewEvo_14_LLM_C_PinkP_12_ICib.raw' :' 12_ICib_Pak_P',\n",
        "    '20230614-0402_QEHF3_1005196_ONJ_TR_MatthewEvo_15_LLM_C_Glue_14_OL.raw' :' 14_OL_Pak_FG',\n",
        "    '20231109-1138_QEHF2_1007125_ONJ_TR_MJC_2_D7_Cz2.raw' :' Cz2_CFrag_FG',\n",
        "    '20231109-1203_QEHF2_1007126_ONJ_TR_MJC_2_D8_03_OBO.raw' :'03_OBO_Pak_P',\n",
        "    '20231109-1227_QEHF2_1007127_ONJ_TR_MJC_2_D9_16_OL.raw' :'16_OL_Pak_FG',\n",
        "    '20231109-1251_QEHF2_1007128_ONJ_TR_MJC_2_D10_13_ICib.raw' :'13_ICib_Pak_W',\n",
        "    '20231109-1315_QEHF2_1007129_ONJ_TR_MJC_2_D11_05_ICB.raw' :'05_ICB_Pak_P',\n",
        "    '20231109-1339_QEHF2_1007130_ONJ_TR_MJC_2_D12_06_ICB.raw' :'06_ICB_Pak_FG',\n",
        "    '20231109-1402_QEHF2_1007131_ONJ_TR_MJC_2_E1_09_ICB.raw' :'09_ICB_Pak_FG',\n",
        "    '20231109-1426_QEHF2_1007132_ONJ_TR_MJC_2_E2_10_ICB.raw' :'10_ICB_Pak_Fiber',\n",
        "    '20231109-1450_QEHF2_1007133_ONJ_TR_MJC_2_E3_Cz17.raw' :'Cz17_CFrag_W',\n",
        "    '20231109-1514_QEHF2_1007134_ONJ_TR_MJC_2_E4_Cz11.raw' :'Cz11_CFrag_W',\n",
        "    '20231109-1538_QEHF2_1007135_ONJ_TR_MJC_2_E5_Cz5b.raw' :'Cz5b_CFrag_W',\n",
        "    '20231109-1538_QEHF2_1007135_ONJ_TR_MJC_2_E5_Cz5b.raw' :'Cz5b_CFrag_W',\n",
        "    # '20240209-0622_QEHF2_1007825_ONJ_TR_MC_CZ21.raw' :'Cz21',\n",
        "    # '20240209-0558 QEHF2_1007824_ONJ_TR_MC_CZ20.raw' :'Cz20',\n",
        "    # '20240209-0534 QEHF2_1007823_ONJ_TR_MC_CZ19.raw' :'Cz19',\n",
        "    # '20240209-0510_QEHF2_1007822_ONJ_TR_MC_CZ18.raw' :'Cz18',\n",
        "    # '20240209-0422_QEHF2_1007820_ONJ_TR_MC_CZ17.raw' :'Cz17',\n",
        "    # '20240209-0359 QEHF2_1007819_ONJ_TR_MC_CZ16.raw' :'Cz16',\n",
        "    # '20240209-0335_QEHF2_1007818_ONJ_TR_MC_CZ15.raw' :'Cz15',\n",
        "    # '20240209-0311_QEHF2_1007817_ONJ_TR_MC_CZ14.raw' :'Cz14',\n",
        "    # '20240209-0247_QEHF2_1007816_ONJ_TR_MC_CZ13.raw' :'Cz13',\n",
        "    # '20240209-0223_QEHF2_1007815_ONJ_TR_MC_CZ12.raw' :'Cz12',\n",
        "    # '20240209-0159_QEHF2_1007814_ONJ_TR_MC_CZ11.raw' :'Cz11',\n",
        "    # '20240209-0136_QEHF2_1007813_ONJ_TR_MC_CZ10.raw' :'Cz10',\n",
        "    # '20240209-0112_QEHF2_1007812_ONJ_TR_MC_CZ9.raw' :'Cz9',\n",
        "    # '20240209-0048_QEHF2_1007811_ONJ_TR_MC_CZ8.raw' :'Cz8',\n",
        "    # '20240209-0024_QEHF2_1007810_ONJ_TR_MC_CZ7.raw' :'Cz7',\n",
        "    # '20240209-0000_QEHF2_1007809_ONJ_TR_MC_CZ6.raw' :'Cz6',\n",
        "    # '20240208-2337_QEHF2_1007808_ONJ_TR_MC_CZ5.raw' :'Cz5',\n",
        "    # '20240208-2313_QEHF2_1007807_ONJ_TR_MC_CZ4.raw' :'Cz4',\n",
        "    # '20240208-2249 QEHF2_1007806_ONJ_TR_MC_CZ3.raw' :'Cz3',\n",
        "    # '20240208-2225 QEHF2_1007805_ONJ_TR_MC_CZ2.raw' :'Cz2',\n",
        "    # '20240208-2201_QEHF2_1007804_ONJ_TR_MC_CZ1.raw' :'Cz1',\n",
        "    '20240209-0645_QEHF2' :'Cz22',\n",
        "    '20240209-0622_QEHF2' :'Cz21',\n",
        "    '20240209-0558_QEHF2' :'Cz20',\n",
        "    '20240209-0534_QEHF2' :'Cz19',\n",
        "    '20240209-0510_QEHF2' :'Cz18',\n",
        "    '20240209-0422_QEHF2' :'Cz17',\n",
        "    '20240209-0359_QEHF2' :'Cz16',\n",
        "    '20240209-0335_QEHF2' :'Cz15',\n",
        "    '20240209-0311_QEHF2' :'Cz14',\n",
        "    '20240209-0247_QEHF2' :'Cz13',\n",
        "    '20240209-0223_QEHF2' :'Cz12',\n",
        "    '20240209-0159_QEHF2' :'Cz11',\n",
        "    '20240209-0136_QEHF2' :'Cz10',\n",
        "    '20240209-0112_QEHF2' :'Cz9',\n",
        "    '20240209-0048_QEHF2' :'Cz8',\n",
        "    '20240209-0024_QEHF2' :'Cz7',\n",
        "    '20240209-0000_QEHF2' :'Cz6',\n",
        "    '20240208-2337_QEHF2' :'Cz5',\n",
        "    '20240208-2313_QEHF2' :'Cz4',\n",
        "    '20240208-2249_QEHF2' :'Cz3',\n",
        "    '20240208-2225_QEHF2' :'Cz2',\n",
        "    '20240208-2201_QEHF2' :'Cz1',\n",
        "    '20240208-2201_QEHF2' :'Cz1',\n",
        "    '20231109-1050_QEHF2' :'MACHINE_BSA',\n",
        "    '20231109-1114_QEHF2' :'INJECTION_BLANK',\n",
        "    '20231109-1602_QEHF2' :'INJECTION_BLANK',\n",
        "\n",
        "}"
      ],
      "metadata": {
        "id": "ED0bSqg_WoOQ"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Amino Acid cand PTM colours"
      ],
      "metadata": {
        "id": "e7WQaPHDugIm"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "hIUs0Yn-PeTB"
      },
      "outputs": [],
      "source": [
        "# Amino acids colors\n",
        "amino_acids_colors = {\n",
        "    \"I\": \"#009688\", \"V\": \"#8bc34a\", \"B\": \"009688\", \"L\": \"#009688\",\n",
        "    \"F\": \"#507351\", \"C\": \"#ffeb3b\", \"M\": \"#ffeb3b\", \"A\": \"#bdd54e\",\n",
        "    \"G\": \"#9e9e9e\", \"T\": \"#ffc75e\", \"W\": \"#f49272\", \"S\": \"#ffc107\",\n",
        "    \"Y\": \"#30802f\", \"P\": \"#607d8b\", \"H\": \"#673ab7\", \"Z\": \"average\",\n",
        "    \"Q\": \"#f44336\", \"E\": \"#f44336\", \"N\": \"#e81e63\", \"D\": \"#f44336\",\n",
        "    \"X\": \"#9d9e9e\", \"K\": \"#701637\", \"R\": \"#bd3e04\"\n",
        "}\n",
        "\n",
        "# PTM colors - if a colour is not defined, the PTM is not plotted\n",
        "\n",
        "ptm_colors = {\n",
        "    \"Hydroxylation\": \"#DFDEE3\",\n",
        "    \"Deamidated\": \"#8F0030\",\n",
        "    \"DeamidationN\" :\"#FF0008\",\n",
        "    \"DeamidationQ\" :\"#C41E23\",\n",
        "    \"Oxidation\": \"#A1CF6B\",\n",
        "    \"Dioxidation\": \"#7AC74F\",\n",
        "    \"Trioxidation \": \"#61AB36\",\n",
        "    \"Carboxy\": \"#f44336\",\n",
        "    \"Dehydrated\": \"#c6d8f5\",\n",
        "    \"Ammonia-loss\": \"#9A879D\",\n",
        "    \"Phospho\": \"#F4845F\",\n",
        "    \"Thiophospho\": \"#F25C54\",\n",
        "    \"Carboxymethyl\": \"#7A3B69\",\n",
        "    \"Formyl\": \"#9A879D\",\n",
        "    \"Pentose\": \"#CE4257\",\n",
        "    \"Arg->Orn\": \"#563F1B\",\n",
        "    \"Sulfide\": \"#F7F052\"\n",
        "}"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "McS7x8aK0MLa"
      },
      "source": [
        "## Functions"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Directory functions"
      ],
      "metadata": {
        "id": "8f2pCtQusDdL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# def combine_spectra_files(base_paths):\n",
        "#   \"\"\"\n",
        "#   Combines spectra files from multiple base paths into a single DataFrame.\n",
        "\n",
        "#   Args:\n",
        "#       base_paths: A list of base paths containing the \"pFind-Filtered.spectra\" files.\n",
        "\n",
        "#   Returns:\n",
        "#       A pandas DataFrame containing the combined spectra data (or None if no files found).\n",
        "#   \"\"\"\n",
        "#   all_data = []\n",
        "#   for base_path in base_paths:\n",
        "#     spectra_file = f'{base_path}/result/pFind-Filtered.spectra'\n",
        "#     try:\n",
        "#       # Read the spectra file using pandas.read_csv\n",
        "#       data = pd.read_csv(spectra_file)\n",
        "#       all_data.append(data)\n",
        "#     except FileNotFoundError:\n",
        "#       print(f\"File not found: {spectra_file}\")\n",
        "\n",
        "#   # Combine all dataframes into a single one (if any files were found)\n",
        "#   if all_data:\n",
        "#     combined_df = pd.concat(all_data, ignore_index=True)\n",
        "#     return combined_df\n",
        "#   else:\n",
        "#     print(\"No spectra files found in provided paths.\")\n",
        "#     return None\n",
        "\n",
        "# import pandas as pd\n",
        "\n",
        "def combine_spectra_files(base_paths):\n",
        "    \"\"\"\n",
        "    Combines spectra files from multiple base paths into a single DataFrame.\n",
        "    Adjusted to handle tab-separated values (TSV) by specifying the sep parameter.\n",
        "\n",
        "    Args:\n",
        "        base_paths: A list of base paths containing the \"pFind-Filtered.spectra\" files.\n",
        "\n",
        "    Returns:\n",
        "        A pandas DataFrame containing the combined spectra data (or None if no files found).\n",
        "    \"\"\"\n",
        "    all_data = []\n",
        "    for base_path in base_paths:\n",
        "        spectra_file = f'{base_path}result/pFind-Filtered.spectra'\n",
        "        try:\n",
        "            # Read the spectra file using pandas.read_csv with sep='\\t' for tab-separated files\n",
        "            data = pd.read_csv(spectra_file, sep='\\t')\n",
        "            all_data.append(data)\n",
        "        except FileNotFoundError:\n",
        "            print(f\"File not found: {spectra_file}\")\n",
        "\n",
        "    # Combine all dataframes into a single one (if any files were found)\n",
        "    if all_data:\n",
        "        combined_df = pd.concat(all_data, ignore_index=True)\n",
        "        return combined_df\n",
        "    else:\n",
        "        print(\"No spectra files found in provided paths.\")\n",
        "        return None\n",
        "\n",
        "\n",
        "def find_spectra_files(base_path):\n",
        "    \"\"\"\n",
        "    Walk through the directory structure starting from base_path to find all\n",
        "    pFind-Filtered.spectra files.\n",
        "\n",
        "    Parameters:\n",
        "    - base_path (str): The base directory to start the search from.\n",
        "\n",
        "    Returns:\n",
        "    - list: A list of paths to pFind-Filtered.spectra files.\n",
        "    \"\"\"\n",
        "    spectra_files = []\n",
        "    for root, dirs, files in os.walk(base_path):\n",
        "        for file in files:\n",
        "            if file == 'pFind-Filtered.spectra':\n",
        "                spectra_files.append(os.path.join(root, file))\n",
        "    return spectra_files\n",
        "\n",
        "def combine_spectra_files_from_single_directory(spectra_files):\n",
        "    \"\"\"\n",
        "    Combine spectra files into a single pandas DataFrame.\n",
        "\n",
        "    Parameters:\n",
        "    - spectra_files (list): A list of paths to pFind-Filtered.spectra files.\n",
        "\n",
        "    Returns:\n",
        "    - DataFrame: A combined DataFrame of all spectra files.\n",
        "    \"\"\"\n",
        "    df_list = []\n",
        "    for file in spectra_files:\n",
        "        df = pd.read_csv(file, sep='\\t')  # Assuming the file is tab-separated\n",
        "        df_list.append(df)\n",
        "    combined_df = pd.concat(df_list, ignore_index=True)\n",
        "    return combined_df"
      ],
      "metadata": {
        "id": "R828OCRSuvt0"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Dictionary Manipulations"
      ],
      "metadata": {
        "id": "cBn6SrFCdp_g"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# def levenshtein_distance(s1, s2):\n",
        "#     \"\"\"\n",
        "#     Calculate the Levenshtein distance between two strings.\n",
        "#     The Levenshtein distance measures the minimum number of single-character edits\n",
        "#     (insertions, deletions, or substitutions) required to change one word into the other.\n",
        "\n",
        "#     Parameters:\n",
        "#     - s1 (str): The first string to compare.\n",
        "#     - s2 (str): The second string to compare.\n",
        "\n",
        "#     Returns:\n",
        "#     - int: The Levenshtein distance between the two strings.\n",
        "#     \"\"\"\n",
        "#     if len(s1) < len(s2):\n",
        "#         return levenshtein_distance(s2, s1)\n",
        "\n",
        "#     if len(s2) == 0:\n",
        "#         return len(s1)\n",
        "\n",
        "#     previous_row = range(len(s2) + 1)\n",
        "#     for i, c1 in enumerate(s1):\n",
        "#         current_row = [i + 1]\n",
        "#         for j, c2 in enumerate(s2):\n",
        "#             insertions = previous_row[j + 1] + 1\n",
        "#             deletions = current_row[j] + 1\n",
        "#             substitutions = previous_row[j] + (c1 != c2)\n",
        "#             current_row.append(min(insertions, deletions, substitutions))\n",
        "#         previous_row = current_row\n",
        "\n",
        "#     return previous_row[-1]\n",
        "\n",
        "\n",
        "\n",
        "def find_match_and_update(row, sequence_position_dict, NCP_peptide_positions_dict):\n",
        "    \"\"\"\n",
        "    Update the row based on matches found in sequence_position_dict and NCP_peptide_positions_dict.\n",
        "\n",
        "    Parameters:\n",
        "    - row: A pandas Series representing a row in a DataFrame.\n",
        "    - sequence_position_dict: A dictionary containing matched sequences with gene data.\n",
        "    - NCP_peptide_positions_dict: A dictionary containing NCP peptide positions.\n",
        "\n",
        "    Returns:\n",
        "    - Updated row with match data from sequence_position_dict or NCP_peptide_positions_dict.\n",
        "    \"\"\"\n",
        "    sequence = row['Sequence']\n",
        "    # Check for a match in the sequence_position_dict\n",
        "    if sequence in sequence_position_dict:\n",
        "        # Use the extract_gene_identifier function to simplify gene name extraction\n",
        "        gene = extract_gene_identifier(sequence_position_dict[sequence]['gene'])\n",
        "        # Update 'GNC&spp' with the extracted gene, append if already exists\n",
        "        row['GNC&spp'] = f\"{gene}\" if pd.isnull(row['GNC&spp']) else f\"{row['GNC&spp']},{gene}\"\n",
        "\n",
        "    if sequence in NCP_peptide_positions_dict:\n",
        "        gene = extract_gene_identifier(NCP_peptide_positions_dict[sequence]['gene'])\n",
        "        # Append or set the gene identifier for NCP matches\n",
        "        row['GNC&spp'] = f\"{gene}\" if pd.isnull(row['GNC&spp']) else f\"{row['GNC&spp']},{gene}\"\n",
        "\n",
        "    return row\n",
        "\n"
      ],
      "metadata": {
        "id": "rPPNDAI5dm-L"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Find positions and create Sequence_postion dictionary Functions"
      ],
      "metadata": {
        "id": "TTeI6z2Mr9_m"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def match_peptides(peptides, sequences):\n",
        "    \"\"\"\n",
        "    Match peptides exactly against sequences in the provided dictionary.\n",
        "\n",
        "    Parameters:\n",
        "    - peptides (list): List of peptide sequences to match.\n",
        "    - sequences (dict): Dictionary of gene sequences with details.\n",
        "    \"\"\"\n",
        "    for peptide in peptides:\n",
        "        for gene, details in sequences.items():\n",
        "            sequence = details['sequence']\n",
        "            offset = details['offset']\n",
        "            position = sequence.find(peptide)\n",
        "            if position != -1:  # Peptide found in the sequence\n",
        "                corrected_position = position + offset  # Adjust for sequence offset\n",
        "                sequence_position_dict[peptide] = {\n",
        "                    'gene': gene,\n",
        "                    'position': corrected_position,\n",
        "                    'length': len(peptide)  # Add length of peptide\n",
        "                }\n",
        "                matched_peptides.add(peptide)\n",
        "                break\n",
        "\n",
        "def is_one_amino_acid_difference(peptide1, peptide2):\n",
        "    \"\"\"\n",
        "    Check if two peptides of the same length differ by exactly one amino acid.\n",
        "    \"\"\"\n",
        "    if len(peptide1) != len(peptide2):\n",
        "        return False\n",
        "\n",
        "    diff_count = sum(1 for a, b in zip(peptide1, peptide2) if a != b)\n",
        "    return diff_count == 1\n",
        "\n",
        "def is_within_mismatch_limit(peptide1, peptide2, max_mismatches):\n",
        "    \"\"\"\n",
        "    Check if two peptides of the same length differ by a number of amino acids up to the max_mismatches limit.\n",
        "    \"\"\"\n",
        "    if len(peptide1) != len(peptide2):\n",
        "        return False\n",
        "\n",
        "    diff_count = sum(1 for a, b in zip(peptide1, peptide2) if a != b)\n",
        "    return diff_count <= max_mismatches\n",
        "\n",
        "def match_unmatched_peptides(remaining_unmatched_peptides, sequence_position_dict):\n",
        "    \"\"\"\n",
        "    Attempt to match remaining unmatched peptides with those in sequence_position_dict,\n",
        "    allowing for one amino acid substitution.\n",
        "    \"\"\"\n",
        "    newly_matched_peptides = set()\n",
        "\n",
        "    for unmatched_peptide in remaining_unmatched_peptides:\n",
        "        for matched_peptide in sequence_position_dict.keys():\n",
        "            if is_one_amino_acid_difference(unmatched_peptide, matched_peptide):\n",
        "                newly_matched_peptides.add(unmatched_peptide)\n",
        "                break\n",
        "\n",
        "    remaining_unmatched_peptides -= newly_matched_peptides\n",
        "    return remaining_unmatched_peptides\n",
        "\n",
        "def iterative_match_unmatched_peptides(unmatched_peptides, sequence_position_dict, sequences, iterations=1): #change the number of interations here)\n",
        "    \"\"\"\n",
        "    Iteratively match remaining unmatched peptides with those in sequence_position_dict,\n",
        "    allowing for an increasing number of amino acid substitutions up to the specified iterations.\n",
        "    Each newly matched peptide is added to sequence_position_dict, broadening the search in subsequent iterations.\n",
        "    \"\"\"\n",
        "    for i in range(1, iterations + 1):  # Allow mismatches up to 'iterations'\n",
        "        newly_matched_peptides = {}\n",
        "\n",
        "        for unmatched_peptide in unmatched_peptides:\n",
        "            for matched_peptide, details in sequence_position_dict.items():\n",
        "                if is_within_mismatch_limit(unmatched_peptide, matched_peptide, i):\n",
        "                    # Record newly matched peptide with the details of the matched_peptide it compares to\n",
        "                    newly_matched_peptides[unmatched_peptide] = details\n",
        "                    break\n",
        "\n",
        "        # Add newly matched peptides to sequence_position_dict and remove from unmatched_peptides\n",
        "        for peptide, details in newly_matched_peptides.items():\n",
        "            sequence_position_dict[peptide] = details\n",
        "            unmatched_peptides.remove(peptide)\n",
        "\n",
        "        # Optional: Report progress after each iteration\n",
        "        print(f\"Iteration {i}: Found {len(newly_matched_peptides)} new matches.\")\n",
        "\n",
        "    return unmatched_peptides\n",
        "\n",
        "def annotate_row(row, positions_dict):\n",
        "    \"\"\"\n",
        "    Annotate a single row with gene and start position based on a given positions dictionary.\n",
        "\n",
        "    Parameters:\n",
        "    - row: A row of a pandas DataFrame.\n",
        "    - positions_dict: A dictionary with sequences as keys and dictionaries with 'gene' and 'position' as values.\n",
        "\n",
        "    Returns:\n",
        "    - row: The modified row with 'GNC&spp' and 'start_position' annotated if the sequence is found in positions_dict.\n",
        "    \"\"\"\n",
        "    sequence = row['Sequence']\n",
        "    if sequence in positions_dict:\n",
        "        row['GNC&spp'] = positions_dict[sequence].get('gene')\n",
        "        row['start_position'] = positions_dict[sequence].get('position')\n",
        "    return row\n"
      ],
      "metadata": {
        "id": "m5kFlstexegj"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Identify PTMS and Parse peptide"
      ],
      "metadata": {
        "id": "ePK5tpSjxOsD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "Tlmd4BGsjYvK"
      },
      "outputs": [],
      "source": [
        "def parse_modifications(mod_str):\n",
        "    \"\"\"\n",
        "    Parses a string containing modification data into a structured list.\n",
        "\n",
        "    This function extracts the position, type of modification, and the amino acid\n",
        "    involved from a given modification string. The expected format of the modification\n",
        "    string is \"position,Modification[AminoAcid];\".\n",
        "\n",
        "    Parameters:\n",
        "        mod_str (str): A string representing modifications, or NaN/None for no modifications.\n",
        "\n",
        "    Returns:\n",
        "        list of tuples: A list where each tuple contains (position, modification type, amino acid).\n",
        "    \"\"\"\n",
        "    # Check if mod_str is not a string (e.g., NaN represented as float in pandas)\n",
        "    if not isinstance(mod_str, str):\n",
        "        return []\n",
        "\n",
        "    parsed_mods = []\n",
        "    mods = mod_str.strip(';').split(';')\n",
        "    for mod in mods:\n",
        "        parts = mod.split(',')\n",
        "        if len(parts) == 2:\n",
        "            pos, mod_detail = parts\n",
        "            try:\n",
        "                mod_type, aa = mod_detail.split('[')\n",
        "                aa = aa.rstrip(']')\n",
        "                parsed_mods.append((int(pos), mod_type, aa))\n",
        "            except ValueError:\n",
        "                # Skip modification if parsing fails\n",
        "                continue\n",
        "\n",
        "    return parsed_mods\n",
        "\n",
        "def parse_peptide(peptide):\n",
        "    \"\"\"\n",
        "    Parses peptide strings to identify amino acids and extract them along with their positions.\n",
        "\n",
        "    :param peptide: A string representation of a peptide.\n",
        "    :return: A list of tuples, each representing an amino acid and its position in the peptide.\n",
        "    \"\"\"\n",
        "    # A simple pattern that matches each amino acid in the sequence.\n",
        "    aa_pattern = re.compile(r'[A-Z]')\n",
        "    parsed_peptides = []\n",
        "    pos = 1  # Position initialization\n",
        "\n",
        "    for match in aa_pattern.finditer(peptide):\n",
        "        aa = match.group()  # Extract the amino acid\n",
        "        parsed_peptides.append((aa, pos))\n",
        "        pos += 1  # Increment position for each amino acid\n",
        "\n",
        "    return parsed_peptides"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Dataframe manipulation"
      ],
      "metadata": {
        "id": "xFfgmqw0wcXQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def parse_and_update_df(sequence, sequence_position_dict): #combined_df\n",
        "    \"\"\"\n",
        "    Check if the sequence matches an entry in the sequence_position_dict and return gene and position.\n",
        "\n",
        "    Parameters:\n",
        "    - sequence (str): The peptide sequence to check.\n",
        "    - sequence_position_dict (dict): Dictionary with peptide sequences as keys and details as values.\n",
        "\n",
        "    Returns:\n",
        "    - tuple: (gene, position) if match is found; (None, None) otherwise.\n",
        "    \"\"\"\n",
        "    if sequence in sequence_position_dict:\n",
        "        # Extract gene and position from the matched entry\n",
        "        gene = sequence_position_dict[sequence]['gene']\n",
        "        position = sequence_position_dict[sequence]['start_position']\n",
        "        return gene, position\n",
        "    else:\n",
        "        # Return None or placeholders if no match is found\n",
        "        return None, None\n",
        "\n",
        "def update_NCP_matches(row, NCP_peptide_positions_dict):\n",
        "    \"\"\"\n",
        "    Update the row with NCP peptide positions if the sequence matches an NCP peptide.\n",
        "\n",
        "    Parameters:\n",
        "    - row: A pandas Series representing a row in a DataFrame.\n",
        "    - NCP_peptide_positions_dict: A dictionary containing NCP peptide positions.\n",
        "\n",
        "    Returns:\n",
        "    - Updated row with NCP peptide match data.\n",
        "    \"\"\"\n",
        "    sequence = row['Sequence']\n",
        "    if sequence in NCP_peptide_positions_dict:\n",
        "        match_data = NCP_peptide_positions_dict[sequence]\n",
        "        # Check and process the gene name specifically if it exists\n",
        "        if 'gene' in match_data:\n",
        "            gene = match_data['gene'].split('-', 1)[0]  # Extract gene name before the underscore\n",
        "            row['GNC&spp'] = gene  # Assuming you're updating to a gene-specific column\n",
        "        else:\n",
        "            for key, value in match_data.items():\n",
        "                row[key] = value\n",
        "    return row\n",
        "\n",
        "\n",
        "def extract_protein_frequencies_and_associations(df, column_name):\n",
        "    \"\"\"\n",
        "    Extracts the frequency of each protein ID and the five most common proteins\n",
        "    listed with each ID from a specified column in a DataFrame.\n",
        "\n",
        "    Parameters:\n",
        "    df (pd.DataFrame): The DataFrame containing protein data.\n",
        "    column_name (str): The name of the column with protein IDs separated by '/'.\n",
        "\n",
        "    Returns:\n",
        "    pd.DataFrame: A DataFrame with each protein's frequency and its five most common associations.\n",
        "    \"\"\"\n",
        "    # Parse the specified column to create a list of proteins per row\n",
        "    #df['Proteins_List'] = df[column_name].apply(lambda x: x.split('/'))\n",
        "    df['Proteins_List'] = df[column_name].apply(lambda x: [p.split('&') for p in x.split('/')])\n",
        "\n",
        "    # Initialize a dictionary to keep count of each protein's frequency\n",
        "    protein_counts = Counter()\n",
        "    # Initialize a dictionary to keep track of proteins listed together\n",
        "    protein_associations = defaultdict(lambda: Counter())\n",
        "\n",
        "    # Count frequencies and associations\n",
        "    for proteins in df['Proteins_List']:\n",
        "        for protein in proteins:\n",
        "            protein_counts[protein] += 1\n",
        "            # Add counts for other proteins in the list to the associations\n",
        "            for associated_protein in proteins:\n",
        "                if associated_protein != protein:\n",
        "                    protein_associations[protein][associated_protein] += 1\n",
        "\n",
        "    # Prepare the final report data\n",
        "    report_data = [{\n",
        "        'Protein': protein,\n",
        "        'Frequency': count,\n",
        "        'Most_Common_Associations': [assoc for assoc, _ in protein_associations[protein].most_common(5)]\n",
        "    } for protein, count in protein_counts.items()]\n",
        "\n",
        "    # Convert report_data to a DataFrame\n",
        "    report_df = pd.DataFrame(report_data)\n",
        "\n",
        "    return report_df\n",
        "\n",
        "def expand_df_to(df):\n",
        "    \"\"\"\n",
        "    Expands a DataFrame by processing each peptide sequence and constructing detailed rows\n",
        "    for each amino acid within the peptide sequences, including modification information and\n",
        "    simplified sample name.\n",
        "\n",
        "    Parameters:\n",
        "        df (pd.DataFrame): A pandas DataFrame containing peptide information.\n",
        "\n",
        "    Returns:\n",
        "        pd.DataFrame: An expanded DataFrame with detailed amino acid and modification information.\n",
        "    \"\"\"\n",
        "    expanded_rows = []\n",
        "    # for _, row in df.iterrows():\n",
        "    #     try:\n",
        "    #         start_position = float(row['start_position'])\n",
        "    #         if not start_position.is_integer():\n",
        "    #             continue  # Skip if start_position is not a whole number\n",
        "    #         start_position = int(start_position)\n",
        "    #     except (ValueError, TypeError):\n",
        "    #         continue  # Skip if start_position is not convertible to float\n",
        "    for _, row in df.iterrows():\n",
        "        try:\n",
        "            # Directly convert start_position to an integer\n",
        "            start_position = int(row['start_position'])\n",
        "        except ValueError:\n",
        "            # Skip this row if start_position is not a valid integer\n",
        "            continue\n",
        "\n",
        "        modifications = parse_modifications(row.get('Modification', ''))\n",
        "        mod_dict = {mod[0]: {'type': mod[1], 'aa': mod[2]} for mod in modifications}\n",
        "        amino_acids = parse_peptide(row['Sequence'])\n",
        "\n",
        "        gnc_spp_split = row.get('GNC&spp', '').split('-', maxsplit=1)\n",
        "        if len(gnc_spp_split) != 2:\n",
        "            continue  # Skip if the split does not result in two parts\n",
        "\n",
        "        GNC, spp = gnc_spp_split\n",
        "        simplified_file_name = \"_\".join(row.get('File_Name', '').split(\"_\")[:2])\n",
        "\n",
        "        for aa, pos in amino_acids:\n",
        "            mod_info = mod_dict.get(pos, {'type': '', 'aa': ''})\n",
        "            expanded_row = {\n",
        "                'id': None,\n",
        "                'sample_name': simplified_file_name,\n",
        "                'GNC': GNC,  #Genome Nomenclature\n",
        "                'spp': spp,  #Species\n",
        "                'aa': aa,    #amino acid\n",
        "                'position': pos - 1 + start_position, #-1 as python counts from 0\n",
        "                'ptm': mod_info['type'], # infomation on any post translational modification\n",
        "                'confidence': None,  #confidence score for the amino acid\n",
        "                'spectraId': row['Scan_No'], #the ID of the MS/MS scan number\n",
        "                'mz': row['Exp.MH+'],  #mass of the peptide detected (m/z, mass over charge)\n",
        "                'z': row['Charge'],   #charge\n",
        "                'pepMass': row['Calc.MH+'],  #mass\n",
        "                'err': row['Mass_Shift(Exp.-Calc.)'],  #error between the observed and calculated mass\n",
        "                'score': row['Final_Score'], # score for the peptide - note this is difference from the 'confidence' which is for the individual amion acid\n",
        "                'scanNum': row['Scan_No'], #scan number\n",
        "                'peptide_gene_name': row['Proteins'], #full information about the ID from mapping algorithm (e.g. MaxQuant, pFind, Orthrus)\n",
        "                'RT': None, #Retention time\n",
        "                'ppm': None, #error in ppm\n",
        "            }\n",
        "            expanded_rows.append(expanded_row)\n",
        "\n",
        "    return pd.DataFrame(expanded_rows)"
      ],
      "metadata": {
        "id": "U7sV67w8whff"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Export functions"
      ],
      "metadata": {
        "id": "oYn3pAs88i6X"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def generate_sequence_with_gaps(group):\n",
        "    \"\"\"\n",
        "    Generate a sequence with gaps for missing positions from a sorted group of sequences.\n",
        "\n",
        "    Parameters:\n",
        "    - group: A pandas DataFrame containing 'position' and 'aa' (amino acid) columns.\n",
        "\n",
        "    Returns:\n",
        "    - A string representing the sequence with gaps for missing positions.\n",
        "    \"\"\"\n",
        "    group_sorted = group.sort_values('position')\n",
        "    full_range = np.arange(group_sorted['position'].min(), group_sorted['position'].max() + 1)\n",
        "    sequence_parts = []\n",
        "\n",
        "    for position in full_range:\n",
        "        if position not in group_sorted['position'].values:\n",
        "            sequence_parts.append(\"-\")\n",
        "        else:\n",
        "            aa_list = group_sorted.loc[group_sorted['position'] == position, 'aa'].unique()\n",
        "            sequence_parts.append(f\"({''.join(aa_list)})\" if len(aa_list) > 1 else ''.join(aa_list))\n",
        "\n",
        "    return ''.join(sequence_parts)\n",
        "\n",
        "def format_fasta_header(row):\n",
        "    \"\"\"\n",
        "    Format a FASTA header string based on the sample name and GNC from a DataFrame row.\n",
        "\n",
        "    Parameters:\n",
        "    - row: A pandas Series representing a row in a DataFrame.\n",
        "\n",
        "    Returns:\n",
        "    - A string representing the formatted FASTA header.\n",
        "    \"\"\"\n",
        "    return f\">{row['sample_name']}_{row['GNC'].replace(' ', '_')}\"\n",
        "\n"
      ],
      "metadata": {
        "id": "xv4_gRZ18hCR"
      },
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZUkf1aZXZzad"
      },
      "source": [
        "###Plotting Functions"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Protein aggregation functions"
      ],
      "metadata": {
        "id": "4nuMqd6_jPLe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def process_group(group):\n",
        "    \"\"\"\n",
        "    Process a pandas DataFrame group to extract specific analytics and return results as a pd.Series.\n",
        "\n",
        "    This function computes various metrics including the unique GNC, position, and\n",
        "    sample_name values, counts for each amino acid, and detailed statistics for the\n",
        "    top 3 amino acids. It includes mean, max, and standard deviation of scores,\n",
        "    top scores and spectraIds, and PTM analysis. The result is returned as a pd.Series\n",
        "    to facilitate easy integration with pandas data structures.\n",
        "\n",
        "    Parameters:\n",
        "    group (pandas.DataFrame): The DataFrame group to process.\n",
        "\n",
        "    Returns:\n",
        "    pandas.Series: A Series containing the computed metrics and statistics.\n",
        "    \"\"\"\n",
        "\n",
        "    gnc, position, sample_name = group['GNC'].iloc[0], group['position'].iloc[0], group['sample_name'].iloc[0]\n",
        "\n",
        "    aa_counts = group['aa'].value_counts()\n",
        "    top_3_aas = aa_counts.index[:3]\n",
        "\n",
        "    result = {'GNC': gnc, 'position': position, 'sample_name': sample_name}\n",
        "\n",
        "    for i, aa in enumerate(top_3_aas, start=1):\n",
        "        aa_group = group[group['aa'] == aa]\n",
        "        scores = aa_group['score']\n",
        "        top_scores = aa_group.nlargest(3, 'score')[['score', 'spectraId']]\n",
        "\n",
        "        ptm_counts = aa_group['ptm'].value_counts()\n",
        "        top_ptms = ptm_counts.index[:2]\n",
        "\n",
        "        aa_info = {\n",
        "            f'aa{i}': aa,\n",
        "            f'n_aa{i}': aa_counts[aa],\n",
        "            f'aa{i}_score_mean': scores.mean(),\n",
        "            f'aa{i}_score_max': scores.max(),\n",
        "            f'aa{i}_score_std': scores.std(),\n",
        "        }\n",
        "\n",
        "        aa_info.update({\n",
        "            **{f'score{j}_aa{i}': score for j, (score, _) in enumerate(top_scores['score'], start=1)},\n",
        "            **{f'spectraId{j}_aa{i}': spectra_id for j, (_, spectra_id) in enumerate(top_scores['spectraId'], start=1)}\n",
        "        })\n",
        "\n",
        "        for j, ptm in enumerate(top_ptms, start=1):\n",
        "            ptm_group = aa_group[aa_group['ptm'] == ptm]\n",
        "            scores = ptm_group['score']\n",
        "            top_ptm_score = ptm_group.nlargest(1, 'score')[['score', 'spectraId']].iloc[0]\n",
        "\n",
        "            aa_info.update({\n",
        "                f'ptm{j}_aa{i}': ptm,\n",
        "                f'%ptm{j}_aa{i}': (ptm_group.shape[0] / aa_group.shape[0]) * 100,\n",
        "                f'ptm{j}_aa{i}_score_mean': scores.mean(),\n",
        "                f'ptm{j}_aa{i}_score_max': scores.max(),\n",
        "                f'ptm{j}_aa{i}_score_std': scores.std(),\n",
        "                f'score1_ptm{j}_aa{i}': top_ptm_score['score'],\n",
        "                f'spectraId1_ptm{j}_aa{i}': top_ptm_score['spectraId']\n",
        "            })\n",
        "\n",
        "        result.update(aa_info)\n",
        "\n",
        "    return pd.Series(result)"
      ],
      "metadata": {
        "id": "06S5sUCAjURY"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_top_n(df, column, n=3):\n",
        "    \"\"\"\n",
        "    Returns the top N items from a dataframe based on the count of a specified column.\n",
        "    \"\"\"\n",
        "    return df[column].value_counts().nlargest(n).reset_index().rename(columns={'index': column, column: f'n_{column}'})\n",
        "\n",
        "def get_scores_info(df, aa, prefix):\n",
        "    \"\"\"\n",
        "    Aggregate score information for a given amino acid.\n",
        "    \"\"\"\n",
        "    filtered_df = df[df['aa'] == aa]\n",
        "    top_scores = filtered_df.nlargest(3, 'score')[['score', 'spectraId']]\n",
        "    score_info = {\n",
        "        f'{prefix}_score_mean': filtered_df['score'].mean(),\n",
        "        f'{prefix}_score_max': filtered_df['score'].max(),\n",
        "        f'{prefix}_score_SD': filtered_df['score'].std(),\n",
        "    }\n",
        "    for i, (score, spectraId) in enumerate(zip(top_scores['score'], top_scores['spectraId']), start=1):\n",
        "        score_info.update({\n",
        "            f'{prefix}_score{i}': score,\n",
        "            f'{prefix}_spectraId{i}': spectraId,\n",
        "        })\n",
        "    return score_info\n",
        "\n",
        "def get_ptm_info(df, aa, prefix):\n",
        "    \"\"\"\n",
        "    Aggregate PTM information for a given amino acid.\n",
        "    \"\"\"\n",
        "    # Filtering for non-null PTMs\n",
        "    filtered_df = df[(df['aa'] == aa) & (df['ptm'].notnull())]\n",
        "    if filtered_df.empty:\n",
        "        return {f'{prefix}_ptm': None}  # Return None or appropriate value if no PTMs are found\n",
        "\n",
        "    # Assuming PTM data is categorical and can be directly counted\n",
        "    top_ptms = filtered_df['ptm'].value_counts().nlargest(2).index\n",
        "    ptm_info = {}\n",
        "    for i, ptm in enumerate(top_ptms, start=1):\n",
        "        ptm_group = filtered_df[filtered_df['ptm'] == ptm]\n",
        "        ptm_info.update({\n",
        "            f'{prefix}_ptm{i}': ptm,\n",
        "            f'{prefix}_ptm{i}_count': ptm_group.shape[0],\n",
        "            f'{prefix}_ptm{i}_percentage': (ptm_group.shape[0] / filtered_df.shape[0]) * 100,\n",
        "            # Additional PTM-specific stats can be added here\n",
        "        })\n",
        "\n",
        "    return ptm_info\n",
        "\n",
        "def aggregate_info(group):\n",
        "    \"\"\"\n",
        "    Aggregates detailed information for each group of data.\n",
        "    \"\"\"\n",
        "    top_aas = get_top_n(group, 'aa')\n",
        "    aggregated_data = []\n",
        "\n",
        "    for _, row in top_aas.iterrows():\n",
        "        aa = row['aa']\n",
        "        aa_info = get_scores_info(group, aa, 'aa1')\n",
        "        ptm_info = get_ptm_info(group, aa, 'aa1')\n",
        "\n",
        "        record = {\n",
        "            'aa': aa,\n",
        "            'n_aa': row['n_aa'],\n",
        "            **aa_info,\n",
        "            **ptm_info\n",
        "        }\n",
        "\n",
        "        aggregated_data.append(record)\n",
        "\n",
        "    return pd.DataFrame(aggregated_data)"
      ],
      "metadata": {
        "id": "dCO2RT3PmX0v"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3DtwTX5xGJBx"
      },
      "source": [
        "# Load files"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0JtblkZ90zf4"
      },
      "source": [
        "##Data Path"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Muli-directories"
      ],
      "metadata": {
        "id": "P0VAxuxqNUl4"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {
        "id": "4ntSKQ06E3Mw"
      },
      "outputs": [],
      "source": [
        "# # Find pFind-Filtered.spectra files\n",
        "# spectra_files = find_spectra_files(BASE_PATH)\n",
        "# # Combine into a single DataFrame\n",
        "# combined_df = combine_spectra_files(spectra_files)\n",
        "# #Add the length of the peptide\n",
        "# combined_df['peptide_length'] = combined_df['Sequence'].apply(len)\n",
        "# #----------------------Reporting\n",
        "# #combined_df.shape\n",
        "# #combined_df.info()\n",
        "# ##combined_df.describe()\n",
        "# #combined_df['Sequence'].nunique()\n",
        "# #combined_df['Miss.Clv.Sites'].nunique()\n",
        "# combined_df.head()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Set the base path as per your Google Drive structure\n",
        "STUDY_NAME1 = 'CE&Tuuli'\n",
        "STUDY_NAME2 = 'CATE'\n",
        "STUDY_NAMES = [STUDY_NAME1, STUDY_NAME2]\n",
        "\n",
        "PFIND_FOLDER1 = f'pFind_{STUDY_NAME1}'  # Folder path for study 1\n",
        "PFIND_FOLDER2 = f'pFind_{STUDY_NAME2}'  # Folder path for study 2\n",
        "\n",
        "BASE_PATH1 = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/{STUDY_NAME1}/{PFIND_FOLDER1}/'\n",
        "BASE_PATH2 = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/{STUDY_NAME2}/{PFIND_FOLDER2}/'\n",
        "\n",
        "def find_base_paths(study_names):\n",
        "  base_paths = []\n",
        "  for study_name in study_names:\n",
        "    PFIND_FOLDER = f'pFind_{study_name}'\n",
        "    base_path = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/{study_name}/{PFIND_FOLDER}/'\n",
        "    base_paths.append(base_path)\n",
        "  return base_paths\n",
        "\n",
        "# Find base paths for both studies\n",
        "base_paths = find_base_paths(STUDY_NAMES)\n",
        "\n",
        "# Load dataframes from each study (assuming 'combine_spectra_files' works as intended)\n",
        "df_study1 = combine_spectra_files([base_paths[0]])  # Pass only the base path for study 1\n",
        "df_study2 = combine_spectra_files([base_paths[1]])  # Pass only the base path for study 2\n",
        "combined_df = pd.concat([df_study1, df_study2], ignore_index=True)\n",
        "\n",
        "\n",
        "# Check the shape of each DataFrame\n",
        "print(f\"Shape of df_study1: {df_study1.shape}\")\n",
        "print(f\"Shape of df_study2: {df_study2.shape}\")\n",
        "\n",
        "# Reporting (assuming combined_df is not None)\n",
        "if combined_df is not None:\n",
        "  print(combined_df.shape)\n",
        "  print(combined_df.info())\n",
        "  print(f\"Number of unique sequences: {combined_df['Sequence'].nunique()}\")\n",
        "#----------------------Reporting\n",
        "#combined_df.shape\n",
        "#combined_df.info()\n",
        "##combined_df.describe()\n",
        "combined_df['Sequence'].nunique()\n",
        "#combined_df['Miss.Clv.Sites'].nunique()\n",
        "#combined_df.head()"
      ],
      "metadata": {
        "id": "SH0RZRzWutOK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "86501c0f-d8c1-4d19-90ec-eded9477e626"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Shape of df_study1: (48116, 19)\n",
            "Shape of df_study2: (30429, 19)\n",
            "(78545, 19)\n",
            "<class 'pandas.core.frame.DataFrame'>\n",
            "RangeIndex: 78545 entries, 0 to 78544\n",
            "Data columns (total 19 columns):\n",
            " #   Column                  Non-Null Count  Dtype  \n",
            "---  ------                  --------------  -----  \n",
            " 0   File_Name               78545 non-null  object \n",
            " 1   Scan_No                 78545 non-null  int64  \n",
            " 2   Exp.MH+                 78545 non-null  float64\n",
            " 3   Charge                  78545 non-null  int64  \n",
            " 4   Q-value                 78545 non-null  float64\n",
            " 5   Sequence                78545 non-null  object \n",
            " 6   Calc.MH+                78545 non-null  float64\n",
            " 7   Mass_Shift(Exp.-Calc.)  78545 non-null  float64\n",
            " 8   Raw_Score               78545 non-null  float64\n",
            " 9   Final_Score             78545 non-null  float64\n",
            " 10  Modification            56710 non-null  object \n",
            " 11  Specificity             78545 non-null  int64  \n",
            " 12  Proteins                78545 non-null  object \n",
            " 13  Positions               78545 non-null  object \n",
            " 14  Label                   78545 non-null  object \n",
            " 15  Target/Decoy            78545 non-null  object \n",
            " 16  Miss.Clv.Sites          78545 non-null  int64  \n",
            " 17  Avg.Frag.Mass.Shift     78545 non-null  float64\n",
            " 18  Others                  78545 non-null  int64  \n",
            "dtypes: float64(7), int64(5), object(7)\n",
            "memory usage: 11.4+ MB\n",
            "None\n",
            "Number of unique sequences: 7569\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "7569"
            ]
          },
          "metadata": {},
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# # Find base paths for both studies\n",
        "# base_paths = find_base_paths(STUDY_NAMES)\n",
        "\n",
        "# # Load dataframes from each study (assuming 'combine_spectra_files' works as intended)\n",
        "# df_study1 = combine_spectra_files([base_paths[0]])  # Pass only the base path for study 1\n",
        "# df_study2 = combine_spectra_files([base_paths[1]])  # Pass only the base path for study 2\n",
        "# combined_df = pd.concat([df_study1, df_study2], ignore_index=True)\n",
        "\n",
        "\n",
        "# # Check the shape of each DataFrame\n",
        "# print(f\"Shape of df_study1: {df_study1.shape}\")\n",
        "# print(f\"Shape of df_study2: {df_study2.shape}\")\n",
        "\n",
        "# # Reporting (assuming combined_df is not None)\n",
        "# if combined_df is not None:\n",
        "#   print(combined_df.shape)\n",
        "#   print(combined_df.info())\n",
        "#   print(f\"Number of unique sequences: {combined_df['Sequence'].nunique()}\")\n",
        "# #----------------------Reporting\n",
        "# #combined_df.shape\n",
        "# #combined_df.info()\n",
        "# ##combined_df.describe()\n",
        "# combined_df['Sequence'].nunique()\n",
        "# #combined_df['Miss.Clv.Sites'].nunique()\n",
        "# #combined_df.head()"
      ],
      "metadata": {
        "id": "wEkHlmiyrXEn"
      },
      "execution_count": 33,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Single directory"
      ],
      "metadata": {
        "id": "p6HYYv7ANb-_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# STUDY_NAME = 'CE&Tuuli'\n",
        "# PFIND_FOLDER = f'pFind_{STUDY_NAME}'\n",
        "# #Find pFind-Filtered.spectra files\n",
        "# #spectra_files = find_spectra_files(BASE_PATH)\n",
        "\n",
        "# spectra_files = [f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/{STUDY_NAME}/{PFIND_FOLDER}/result/pFind-Filtered.spectra']\n",
        "# # Combine into a single DataFrame\n",
        "# combined_df = combine_spectra_files_from_single_directory(spectra_files)\n",
        "# #Add the length of the peptide\n",
        "\n",
        "# combined_df['peptide_length'] = combined_df['Sequence'].apply(len)\n",
        "# #----------------------Reporting\n",
        "# #combined_df.shape\n",
        "# #combined_df.info()\n",
        "# ##combined_df.describe()\n",
        "# combined_df['Sequence'].nunique()\n",
        "# #combined_df['Miss.Clv.Sites'].nunique()\n",
        "# #combined_df.head()"
      ],
      "metadata": {
        "id": "M51comiRGK7J"
      },
      "execution_count": 34,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nZo29vtF3BKm"
      },
      "source": [
        "###Find Collagen motif\n",
        "Iterates through each peptide in unique_peptides.\n",
        "Uses a generator expression within all() to check if every third character, starting from the offset determined by len(peptide) % 3, is a 'G'.\n",
        "This method directly reflects the requirement for 'G' to be every third character after an initial offset and accounts for sequences that may not start with 'G' or might be truncated."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Dictionary Matching"
      ],
      "metadata": {
        "id": "8atLJDXUsI8m"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Initialize dictionary to store matched peptide details with sequence length\n",
        "sequence_position_dict = {}\n",
        "\n",
        "# Initialize set for matched peptides\n",
        "matched_peptides = set()\n",
        "\n",
        "# Extract unique peptides from DataFrame\n",
        "unique_peptides = combined_df['Sequence'].unique().tolist()\n",
        "\n",
        "# First round of matching\n",
        "match_peptides(unique_peptides, sequences_dict)\n",
        "\n",
        "# Calculate unmatched peptides after the first round\n",
        "remaining_unmatched_unique_found_peptides = set(unique_peptides) - matched_peptides\n",
        "\n",
        "# Iterative second round of matching, including sequences_dict for gene and position info\n",
        "remaining_unmatched_after_iterations = iterative_match_unmatched_peptides(remaining_unmatched_unique_found_peptides, sequence_position_dict, sequences_dict, iterations=1)\n",
        "\n",
        "# Reporting the outcome\n",
        "total_matched_after_iterations = len(sequence_position_dict)  # Updated total matches include initial and iterative matches\n",
        "total_unmatched_after_iterations = len(remaining_unmatched_after_iterations)\n",
        "\n",
        "\n",
        "print(f\"Total matched peptides after iterations: {total_matched_after_iterations}\")\n",
        "print(f\"Total unmatched peptides after iterations: {total_unmatched_after_iterations}\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ZGXLjS61m0kV",
        "outputId": "5df2c7af-bd56-49a2-8a36-3d2ea7bd954c"
      },
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Iteration 1: Found 804 new matches.\n",
            "Total matched peptides after iterations: 2911\n",
            "Total unmatched peptides after iterations: 4658\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Add Gene IDs and start postion to peptides"
      ],
      "metadata": {
        "id": "5oFITi57DcuE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Initialize new columns to avoid KeyError\n",
        "combined_df['GNC&spp'] = None\n",
        "combined_df['start_position'] = None\n",
        "\n",
        "# Annotate Rols\n",
        "combined_df = combined_df.apply(annotate_row, positions_dict=sequence_position_dict, axis=1)\n",
        "\n",
        "# Check for non-null entries if nulls are np.nan (standard missing data marker in pandas)\n",
        "non_null_entries = combined_df['GNC&spp'].notnull().sum()\n",
        "\n",
        "# If empty strings are used to represent missing data, you might want to adjust the method to something like:\n",
        "# non_empty_entries = combined_df[combined_df['GNC&spp'] != ''].shape[0]\n",
        "\n",
        "print(f\"Number of non-null/empty entries in 'GNC&spp': {non_null_entries}\")\n",
        "\n",
        "# Most common entries in 'GNC&spp'\n",
        "most_common_entries = combined_df['GNC&spp'].value_counts().head(15)  # Adjust the number inside head() as needed\n",
        "print(\"Most common entries in 'GNC&spp':\")\n",
        "print(most_common_entries)\n",
        "\n",
        "# Calculating the percentage of rows with an entry in 'GNC&spp'\n",
        "total_rows = combined_df.shape[0]\n",
        "percentage_with_entry = (non_null_entries / total_rows) * 100\n",
        "\n",
        "print(f\"Percentage of rows with an entry in 'GNC&spp': {percentage_with_entry:.2f}%\")"
      ],
      "metadata": {
        "id": "-6feqVwn-nLx",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "729ec7a2-e837-4d9c-9581-d77301a79d25"
      },
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of non-null/empty entries in 'GNC&spp': 62006\n",
            "Most common entries in 'GNC&spp':\n",
            "COL1A1-Bos_tau    16629\n",
            "COL1A2-Bos_mut     7888\n",
            "TRYUP-Sus_sco      4435\n",
            "COL1A2-Lox_afr     3796\n",
            "COL3A1-Bos_tau     3357\n",
            "COL1A1-Hip_amp     2682\n",
            "K2C1-Hom_sap       2220\n",
            "COL1A2-Equ_cab     2154\n",
            "KS10-Hom_sap       1875\n",
            "PIV-Pse_aer        1865\n",
            "COL3A1-Equ_cab     1763\n",
            "COL1A2-Hip_amp     1438\n",
            "KS9-Hom_sap        1418\n",
            "COL1A2-Equ_asi     1374\n",
            "COLunk-Bos_tau     1072\n",
            "Name: GNC&spp, dtype: int64\n",
            "Percentage of rows with an entry in 'GNC&spp': 78.94%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter the DataFrame for rows where 'GNC&spp' or 'start_position' is missing\n",
        "missing_annotations_df = combined_df[combined_df['GNC&spp'].isna() | combined_df['start_position'].isna()]\n",
        "\n",
        "# Print the top 100 entries for the 'Proteins' column from these rows\n",
        "print(missing_annotations_df['Proteins'].head(400))\n",
        "\n",
        "# Extract all protein IDs, split by '/', from the 'Proteins' column\n",
        "protein_list = missing_annotations_df['Proteins'].str.split('/').explode()\n",
        "\n",
        "# Find the 100 most common protein strings\n",
        "top_400_proteins = protein_list.value_counts().head(400)\n",
        "\n",
        "\n",
        "top_400_proteins.to_csv('top_400_proteins.txt', header=False, index=True)\n",
        "\n",
        "\n",
        "# Get protein names and their counts\n",
        "protein_counts = protein_list.value_counts()\n",
        "\n",
        "# Ensure at least one protein per count (using nlargest)\n",
        "one_protein_per_count = protein_counts.nlargest(len(protein_counts.unique()))\n",
        "\n",
        "# Combine protein name and count into a DataFrame\n",
        "result_df = pd.DataFrame({'protein_id': one_protein_per_count.index, 'count': one_protein_per_count.values})\n",
        "\n",
        "# Print the results (protein ID and count)\n",
        "print(result_df)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VVpbBLr5s1Fu",
        "outputId": "1b6066f0-1c13-4c14-fa93-d41c973412fa"
      },
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "429     C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "463                                       XP_041562691.1/\n",
            "498     KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "504     KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "519     KAF1529257.1/WP_168397204.1/3B63_L/5ADX_H/KAG8...\n",
            "599                                       XP_026294502.1/\n",
            "637     KAF6085594.1/KAF6085593.1/XP_006178597.1/XP_01...\n",
            "667     KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "735     C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "766                          KAE8693106.1/XP_039012742.1/\n",
            "779     MCP6497603.1/MCP6564868.1/ABA42866.1/3V03_A/5Y...\n",
            "798     KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "826     KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "878     XP_038174782.1/XP_038174780.1/XP_038174781.1/X...\n",
            "880     XP_038174782.1/XP_038174780.1/XP_038174781.1/X...\n",
            "899     KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "905     MCP6497603.1/MCP6564868.1/ABA42866.1/3V03_A/5Y...\n",
            "906                            XP_050038206.1/TGZ36829.1/\n",
            "920                                       XP_032647404.1/\n",
            "932     ELU01212.1/KAI5695485.1/XP_030569544.1/EDV4505...\n",
            "937     KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "1030                             TNM97480.1/KAF5894526.1/\n",
            "1073    CAF2094928.1/CAF1812475.1/VDC70085.1/CAG787432...\n",
            "1118    KAB0391484.1/AAB35421.1/BAG63152.1/XP_00801089...\n",
            "1136                                          AAT06139.1/\n",
            "1139    XP_008056148.1/XP_014446155.1/XP_052052481.1/X...\n",
            "1162    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1237    C0HLI1.1/C0HLJ7.1/XP_035756024.1/ERE69281.1/BA...\n",
            "1256    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "1265                                          AAT06139.1/\n",
            "1272                               PHS77156.1/MBQ49997.1/\n",
            "1304                                          AAT06139.1/\n",
            "1314    KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "1330    XP_036691630.1/XP_036691628.1/XP_036691631.1/X...\n",
            "1336    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1344                                          AAT06139.1/\n",
            "1369    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1372    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "1375    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1401    XP_038174782.1/XP_038174780.1/XP_038174781.1/X...\n",
            "1404    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1462                                      XP_023985979.1/\n",
            "1484                               PHS77156.1/MBQ49997.1/\n",
            "1511                               PHS77156.1/MBQ49997.1/\n",
            "1550    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "1557    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "1560    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "1562    KFW82933.1/XP_016402454.1/XP_016334116.1/TEU55...\n",
            "1596    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1603    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "1623    XP_008056148.1/XP_014446155.1/XP_052052481.1/X...\n",
            "1647                       XP_039189090.1/XP_039189089.1/\n",
            "1671    WP_265319884.1/MBG7006293.1/WP_034081609.1/WP_...\n",
            "1678    2FT3_A/TEA10467.1/XP_011857624.1/KAI4545195.1/...\n",
            "1683                                          AAT06139.1/\n",
            "1692    KAF1529257.1/WP_168397204.1/3B63_L/5ADX_H/KAG8...\n",
            "1696                                          AAT06139.1/\n",
            "1705    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "1707    C0HLI1.1/C0HLJ7.1/XP_035756024.1/ERE69281.1/BA...\n",
            "1711    KAF1529257.1/WP_168397204.1/3B63_L/5ADX_H/KAG8...\n",
            "1712    WP_265319884.1/MBG7006293.1/WP_034081609.1/WP_...\n",
            "1726    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "1763    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1806                                            C0HJP5.1/\n",
            "1811                       XP_039189090.1/XP_039189089.1/\n",
            "1828    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1838    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "1841    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1844    ELU01212.1/KAI5695485.1/XP_030569544.1/EDV4505...\n",
            "1863                                          AAT06139.1/\n",
            "1893    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "1902    KAH8255364.1/KAH8381007.1/KAH8308344.1/AGB9157...\n",
            "1921    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "1947                                          ALM23467.1/\n",
            "1951    QQP38306.1/CAA50332.1/ABW71259.1/KAG8129429.1/...\n",
            "1994    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2026    KAI9558932.1/XP_032783644.1/GFT72627.1/XP_0226...\n",
            "2033    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2047    KAF1529257.1/WP_168397204.1/3B63_L/5ADX_H/KAG8...\n",
            "2050    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "2055    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "2063    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "2079    KAI9558932.1/XP_032783644.1/GFT72627.1/XP_0226...\n",
            "2091    MCP6497603.1/MCP6564868.1/ABA42866.1/3V03_A/5Y...\n",
            "2120                                          ALM23467.1/\n",
            "2135                                      XP_007195818.1/\n",
            "2173                           XP_050038206.1/TGZ36829.1/\n",
            "2184                       XP_041553566.1/XP_040707018.1/\n",
            "2193    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "2197    KAI9558932.1/XP_032783644.1/GFT72627.1/XP_0226...\n",
            "2200          XP_034232182.1/XP_052124545.1/KAE8745731.1/\n",
            "2216    QQP38306.1/CAA50332.1/ABW71259.1/KAG8129429.1/...\n",
            "2231    CAA50332.1/ABW71259.1/AFJ64472.1/AAQ67451.1/AA...\n",
            "2238    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2239                       XP_039189090.1/XP_039189089.1/\n",
            "2245    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "2250    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "2251    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "2255                                          ALM23467.1/\n",
            "2262    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "2263                                      XP_026294502.1/\n",
            "2287    CAF2094928.1/CAF1812475.1/VDC70085.1/CAG787432...\n",
            "2302    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2308    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2309                       XP_039189090.1/XP_039189089.1/\n",
            "2311    KAF0873703.1/KFO27958.1/XP_004628686.1/XP_0391...\n",
            "2322    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2345                    GFW60454.1/GFR19359.1/GFY51967.1/\n",
            "2350                           XP_050038206.1/TGZ36829.1/\n",
            "2357                                          AAT06139.1/\n",
            "2370    ELU01212.1/KAI5695485.1/XP_030569544.1/EDV4505...\n",
            "2372    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "2374    CAH8852664.1/CAH8858534.1/CAI2729635.1/CAI2729...\n",
            "2396                               MAJ21524.1/ACB13096.1/\n",
            "2429    PNI33057.1/PNJ25491.1/AAA36145.1/XP_031319480....\n",
            "2437                               MAJ21524.1/ACB13096.1/\n",
            "2445    BAD12426.1/XP_026281316.1/KAI5722962.1/KAI5725...\n",
            "2448    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2449                                      XP_023840994.1/\n",
            "2457    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2466    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "2489    KAI9558932.1/XP_032783644.1/GFT72627.1/XP_0226...\n",
            "2490                       XP_039189090.1/XP_039189089.1/\n",
            "2498    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "2501                                      XP_023840994.1/\n",
            "2507    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2509                                          ALM23467.1/\n",
            "2517    KAA0199347.1/XP_026281587.1/XP_026272714.1/XP_...\n",
            "2522    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "2527    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2534                               ASU08726.1/AKQ06662.1/\n",
            "2539    XP_043202016.1/XP_043202010.1/XP_043202018.1/K...\n",
            "2545    ELU01212.1/KAI5695485.1/XP_030569544.1/EDV4505...\n",
            "2556                                          ABX75382.1/\n",
            "2566                                            C0HJP5.1/\n",
            "2570    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "2572                               PHS77156.1/MBQ49997.1/\n",
            "2598                                      XP_023840994.1/\n",
            "2613    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "2645    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2647                                      XP_007195818.1/\n",
            "2654                               MAJ21524.1/ACB13096.1/\n",
            "2703    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "2718                                      XP_023840994.1/\n",
            "2721                               ASU08726.1/AKQ06662.1/\n",
            "2723    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "2727                               PHS77156.1/MBQ49997.1/\n",
            "2736                                      XP_023840994.1/\n",
            "2740    XP_003358093.2/XP_005653958.1/XP_047613433.1/X...\n",
            "2749    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "2756    KFQ57848.1/NWX43241.1/NWI86726.1/AEG78351.1/EH...\n",
            "2768    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "2769    XP_043202016.1/XP_043202010.1/XP_043202018.1/K...\n",
            "2770    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "2772                               ASU08726.1/AKQ06662.1/\n",
            "2780    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "2786    KAE8752147.1/CAD7196728.1/KAF4514053.1/CAD7568...\n",
            "2793    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "2797    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "2808                               PHS77156.1/MBQ49997.1/\n",
            "2819    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "2833                               ASU08726.1/AKQ06662.1/\n",
            "2844                       XP_039189090.1/XP_039189089.1/\n",
            "2851    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "2865    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "2869                                      XP_023840994.1/\n",
            "2878    KAB1733990.1/CAB64662.1/GBL77556.1/GBL77561.1/...\n",
            "2901    ELU09075.1/RKO90579.1/XP_037897848.1/EOA94351....\n",
            "2908                               ASU08726.1/AKQ06662.1/\n",
            "2913                         XP_013067610.1/KAI8751318.1/\n",
            "2915    ELU09075.1/RKO90579.1/XP_037897848.1/EOA94351....\n",
            "2919                                      XP_023840994.1/\n",
            "2948    KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "2955    XP_032004443.1/XP_032004442.1/WP_252511980.1/X...\n",
            "2991                                      XP_026274744.1/\n",
            "2995    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3006                                      XP_023840994.1/\n",
            "3037    QGX86366.1/3B63_A/3B63_C/KAF1529257.1/ACB41835...\n",
            "3040                               PHS77156.1/MBQ49997.1/\n",
            "3041                       XP_051007496.1/XP_051007495.1/\n",
            "3062    CRL25778.1/XP_040813404.1/CAG8109423.1/CAG7928...\n",
            "3074    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3098    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "3101                               PHS77156.1/MBQ49997.1/\n",
            "3106    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3145        XP_040662007.1/XP_040705940.1/XP_041551256.1/\n",
            "3148    XP_030664052.1/XP_003279481.2/XP_026304479.1/X...\n",
            "3159                       XP_026347034.1/XP_026347037.1/\n",
            "3161    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "3166    C0HJP1.1/XP_019493986.1/XP_023445022.1/XP_0289...\n",
            "3180    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3184    XP_043202016.1/XP_043202010.1/XP_043202018.1/K...\n",
            "3192    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3193                               MAJ21524.1/ACB13096.1/\n",
            "3195    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3197    KAB1733990.1/CAB64662.1/GBL77556.1/GBL77561.1/...\n",
            "3206    KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "3208    QNT18286.1/QNT18283.1/QNT18294.1/XP_044376900....\n",
            "3211    VFV38247.1/CAA29248.1/ACU45491.1/WP_254515363....\n",
            "3215                                      XP_023840994.1/\n",
            "3224    C0HLI1.1/C0HLJ7.1/XP_035756024.1/ERE69281.1/BA...\n",
            "3228                               PHS77156.1/MBQ49997.1/\n",
            "3241    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3242                               MAJ21524.1/ACB13096.1/\n",
            "3243    PNI33057.1/PNJ25491.1/AAA36145.1/XP_031319480....\n",
            "3245    KAF1529257.1/WP_168397204.1/3B63_L/5ADX_H/KAG8...\n",
            "3251                               ASU08726.1/AKQ06662.1/\n",
            "3257    KFV57992.1/MBN3304060.1/KFV70405.1/MBN3306880....\n",
            "3264    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3289                                          AAT06139.1/\n",
            "3305    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3322    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "3346    XP_006560146.1/XP_003689891.1/XP_017883478.1/X...\n",
            "3349    KAB1733990.1/CAB64662.1/GBL77556.1/GBL77561.1/...\n",
            "3357                               MAJ21524.1/ACB13096.1/\n",
            "3364    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3386    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3396    KAG8512871.1/XP_033618937.1/NP_001008750.1/XP_...\n",
            "3399                                      XP_004608270.1/\n",
            "3402                               MAJ21524.1/ACB13096.1/\n",
            "3404    KAJ0162024.1/KZL64837.1/KAF7898128.1/XP_038727...\n",
            "3422    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "3425    KAJ0162024.1/KZL64837.1/KAF7898128.1/XP_038727...\n",
            "3430    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "3453    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "3462    XP_006560146.1/XP_003689891.1/XP_017883478.1/X...\n",
            "3474    MBZ3891101.1/EFB27495.1/P04258.1/AAB59383.1/KA...\n",
            "3477    MBZ3891101.1/EFB27495.1/P04258.1/AAB59383.1/KA...\n",
            "3478    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "3480    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "3486    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3494    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3516    WP_254515363.1/MBO8905120.1/EDL02573.1/XP_0482...\n",
            "3519    KFV57992.1/MBN3304060.1/KFV70405.1/MBN3306880....\n",
            "3524    KAH8255364.1/KAH8381007.1/KAH8308344.1/AGB9157...\n",
            "3530                                          ALM23467.1/\n",
            "3541    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3549                               MAJ21524.1/ACB13096.1/\n",
            "3555    BAG63158.1/XP_026308655.1/XP_024113011.1/XP_02...\n",
            "3560                               MAJ21524.1/ACB13096.1/\n",
            "3567    KAB1733990.1/CAB64662.1/GBL77556.1/GBL77561.1/...\n",
            "3572    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3604    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "3606    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3640    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "3641                               MAJ21524.1/ACB13096.1/\n",
            "3651    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "3659    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "3666                                      XP_023840994.1/\n",
            "3667    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3675    KAJ0162024.1/KZL64837.1/KAF7898128.1/XP_038727...\n",
            "3678    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3680    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "3686                                      XP_034233054.1/\n",
            "3694    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "3695                                          ABA39179.1/\n",
            "3709    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3717    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "3720    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "3761    CAA50332.1/ABW71259.1/AFJ64472.1/AAQ67451.1/AA...\n",
            "3766    QGX86366.1/3B63_A/3B63_C/KAF1529257.1/ACB41835...\n",
            "3772    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3776    XP_027728615.1/XP_020853290.1/KAI5128820.1/XP_...\n",
            "3779                               MAJ21524.1/ACB13096.1/\n",
            "3789                               MAJ21524.1/ACB13096.1/\n",
            "3808                               PHS77156.1/MBQ49997.1/\n",
            "3836    KAB0389662.1/PNI33057.1/PNJ25491.1/XP_00675716...\n",
            "3838    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "3842    GBL77556.1/GBL77561.1/GBL77558.1/GBL77559.1/GF...\n",
            "3847                                          AAT06139.1/\n",
            "3893                                      XP_040147273.1/\n",
            "3896    CAA50332.1/ABW71259.1/AFJ64472.1/AAQ67451.1/AA...\n",
            "3916    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "3921            XP_041556663.1/ADE74975.1/XP_040703360.1/\n",
            "3925    CAF4368892.1/KAF1529257.1/ACB41835.1/ACB41865....\n",
            "3952    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "3967                                      XP_040147273.1/\n",
            "3983                                      XP_007195818.1/\n",
            "4001                                        KAI1308738.1/\n",
            "4019    QNT18286.1/QNT18283.1/QNT18294.1/XP_044376900....\n",
            "4024                         KAG3269228.1/XP_040126450.1/\n",
            "4025    XP_014256321.1/XP_014288054.1/CAH1393132.1/XP_...\n",
            "4029                                      XP_004608270.1/\n",
            "4041    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4046    XP_049738935.1/XP_003417915.1/XP_006888799.1/X...\n",
            "4053    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4072                                      WP_138519551.1/\n",
            "4089    MBT5160811.1/BBD34522.1/MBT6387688.1/WP_097154...\n",
            "4090    ODM23861.1/XP_043132635.1/XP_022403087.1/OOF97...\n",
            "4094    KAB1733990.1/GBN68585.1/GBN68584.1/GBN68582.1/...\n",
            "4113                           AFS50718.1/WP_251340670.1/\n",
            "4132    XP_035206481.1/XP_035206482.1/KFM60571.1/XP_01...\n",
            "4136    XP_028700691.1/XP_047607991.1/XP_005550260.1/N...\n",
            "4144    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4145    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "4148    5XW1_A/P00761.1/CAB58178.1/1AKS_A/1AVW_A/1FMG_...\n",
            "4173    XP_051525967.1/XP_014001029.1/XP_014001028.1/X...\n",
            "4174                                            C0HJP5.1/\n",
            "4175    CRL25778.1/XP_040813404.1/CAG8109423.1/CAG7928...\n",
            "4184                                          ABZ81054.1/\n",
            "4201                               MAJ21524.1/ACB13096.1/\n",
            "4204                                      XP_004608270.1/\n",
            "4215    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4220    EEB93724.1/KAH9607788.1/ABG49391.1/AKM54887.1/...\n",
            "4232                       XP_039189090.1/XP_039189089.1/\n",
            "4238    KAE8750270.1/AAM29636.1/GBM05080.1/XP_02628953...\n",
            "4244    5XW1_A/P00761.1/CAB58178.1/1AKS_A/1AVW_A/1FMG_...\n",
            "4262    QGX86366.1/3B63_A/3B63_C/KAF1529257.1/ACB41835...\n",
            "4270    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "4286                           AFS50718.1/WP_251340670.1/\n",
            "4290    KAB0391484.1/AAB35421.1/BAG63152.1/XP_00801089...\n",
            "4291    AGR28078.1/QBB19875.1/AEU03868.1/ANS56518.1/AW...\n",
            "4311    NVI70683.1/NVI70698.1/AUW83286.1/AUW83219.1/AU...\n",
            "4326                                      XP_004608270.1/\n",
            "4335         QWE91166.1/QWE91165.1/TDH69765.1/RMX67487.1/\n",
            "4337                               ASU08726.1/AKQ06662.1/\n",
            "4340    KFO23903.1/XP_004632680.1/EHA98618.1/XP_004862...\n",
            "4355                                      XP_003791605.1/\n",
            "4367    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4396    XP_036346229.1/GBN68585.1/GBN68584.1/GBN68582....\n",
            "4397                               PHS77156.1/MBQ49997.1/\n",
            "4400                                      WP_138519551.1/\n",
            "4402                                          ABA39179.1/\n",
            "4403    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4424    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "4427    KAG7174224.1/XP_042212113.1/XP_042212114.1/XP_...\n",
            "4432    KAE8752147.1/KYB26155.1/XP_014282769.1/XP_0142...\n",
            "4437                                      XP_041551953.1/\n",
            "4461    KAG7174224.1/XP_042212113.1/XP_042212114.1/XP_...\n",
            "4479    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "4485                                      XP_015269157.1/\n",
            "4490                                      XP_041560428.1/\n",
            "4492    XP_028522549.1/XP_031365952.1/XP_043796388.1/X...\n",
            "4496                                          ABZ81054.1/\n",
            "4498    WP_264214829.1/XP_025773236.1/NWX92503.1/NXA40...\n",
            "4499    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "4501    KAG8512871.1/XP_029783500.1/XP_010984311.2/KAB...\n",
            "4505                                      XP_004608270.1/\n",
            "4508                               PHS77156.1/MBQ49997.1/\n",
            "4522    MBU2514434.1/MBV8451037.1/MBV8356511.1/MCJ1442...\n",
            "4557    KAG8512871.1/XP_033618937.1/XP_004390384.1/XP_...\n",
            "4562    EFB27495.1/P04258.1/KAF6113246.1/XP_035879080....\n",
            "4567    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "4583                         XP_013067610.1/KAI8751318.1/\n",
            "4585                                      XP_041484083.1/\n",
            "4590    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "4592                                      XP_023840994.1/\n",
            "4607                                      XP_003791605.1/\n",
            "4616    XP_022900593.1/XP_037008559.1/XP_006560146.1/X...\n",
            "4619    prf||0705172A/1JTI_A/1UHG_A/AUD54643.1/AUD5479...\n",
            "4620    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4621                                      WP_138519551.1/\n",
            "4633    XP_007397011.1/WP_076346033.1/MCL6489439.1/WP_...\n",
            "4637                                      XP_026294502.1/\n",
            "4639    CAA49832.1/BAB31724.1/CAA06510.1/ERE84830.1/XP...\n",
            "4642    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "4644    XP_022900593.1/XP_037008559.1/XP_006560146.1/X...\n",
            "4648         QWE91166.1/QWE91165.1/TDH69765.1/RMX67487.1/\n",
            "4666    XP_014256321.1/XP_014288054.1/CAH1393132.1/XP_...\n",
            "4669                                      XP_040147273.1/\n",
            "4677    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4681    XP_012286526.1/XP_023287532.1/XP_012286534.1/X...\n",
            "4685                                          ABZ81054.1/\n",
            "4687                       XP_020388689.1/XP_020388690.1/\n",
            "4693    KTF78707.1/KAG1967375.1/XP_051571265.1/XP_0425...\n",
            "4696    KAA1478665.1/KAF9224509.1/XP_010550599.1/CAA73...\n",
            "4698                                      WP_138519551.1/\n",
            "4699                                          CAC94011.1/\n",
            "4703                                      XP_023840994.1/\n",
            "4706    CAF4614105.1/UBT01889.1/AFM38192.1/CAF4224705....\n",
            "4715    KAG7174224.1/XP_042212113.1/XP_042212114.1/XP_...\n",
            "4716    KAG8512871.1/XP_033618937.1/XP_004390384.1/XP_...\n",
            "4719    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4730    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "4732    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4737    KAE8752147.1/KYB26155.1/XP_014282769.1/XP_0142...\n",
            "4770                                      WP_138519551.1/\n",
            "4775            XP_041556663.1/ADE74975.1/XP_040703360.1/\n",
            "4779                                      XP_026294502.1/\n",
            "4781    MBU2514434.1/MBV8451037.1/MBV8356511.1/MCJ1442...\n",
            "4784    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "4813    MBZ3891101.1/EFB27495.1/P04258.1/AAB59383.1/KA...\n",
            "4820    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "4829                                      XP_007195818.1/\n",
            "4832    C0HJP6.1/KAF5927812.1/C0HJN8.1/XP_032498639.1/...\n",
            "4838                                      XP_023840994.1/\n",
            "4853                         XP_026272535.1/KAE8740084.1/\n",
            "4858    EFB27495.1/KAF6113246.1/XP_035879080.1/XP_0358...\n",
            "4864    C0HLI7.1/XP_027728615.1/XP_020853290.1/XP_0103...\n",
            "4881    MCP6497603.1/MCP6564868.1/ABA42866.1/3V03_A/5Y...\n",
            "4882    QQP38306.1/NXU13765.1/VDI74957.1/1NBM_E/1NBM_D...\n",
            "4886    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4891    MBU2514434.1/MBV8451037.1/MBV8356511.1/MCJ1442...\n",
            "4892                                      XP_035108944.1/\n",
            "4902    KAI5769851.1/XP_031534140.1/XP_006210056.1/XP_...\n",
            "4906                           AFS50718.1/WP_251340670.1/\n",
            "4910    XP_004591119.1/XP_040856287.1/XP_035889661.1/X...\n",
            "4912                                      WP_138519551.1/\n",
            "4917    WP_264214829.1/XP_025773236.1/NWX92503.1/NXA40...\n",
            "4931    MBZ3891101.1/XP_012519093.1/XP_042544240.1/XP_...\n",
            "Name: Proteins, dtype: object\n",
            "         protein_id  count\n",
            "0                    16539\n",
            "1        NWV22044.1    422\n",
            "2    XP_005570993.2    416\n",
            "3    XP_024113020.1    415\n",
            "4    XP_014748466.1    415\n",
            "5    XP_030678472.1    415\n",
            "6    XP_032029551.1    415\n",
            "7    XP_011857542.1    414\n",
            "8    XP_028685340.1    414\n",
            "9        AAG41947.1    414\n",
            "10   XP_011727181.1    414\n",
            "11      NP_006112.3    414\n",
            "12       AFA52007.1    414\n",
            "13       AAB47721.1    414\n",
            "14       AFA52006.1    414\n",
            "15   XP_003845964.2    414\n",
            "16   XP_003906478.2    414\n",
            "17       AFA52002.1    414\n",
            "18       AFA52005.1    414\n",
            "19   XP_025258843.1    414\n",
            "20   XP_011857532.1    414\n",
            "21   XP_001098182.2    414\n",
            "22   XP_025258844.1    414\n",
            "23   XP_050604771.1    414\n",
            "24   XP_011857566.1    414\n",
            "25   XP_011857549.1    414\n",
            "26   XP_011903561.1    414\n",
            "27   XP_025258845.1    414\n",
            "28   XP_011857559.1    414\n",
            "29       AKI71766.1    414\n",
            "30   XP_004053220.3    414\n",
            "31       PNI33041.1    414\n",
            "32   NP_001104288.1    414\n",
            "33       AFA52004.1    414\n",
            "34       AFA52003.1    414\n",
            "35   XP_037853004.1    413\n",
            "36   XP_004429138.2    406\n",
            "37   XP_011854305.1    404\n",
            "38   XP_051624887.1    402\n",
            "39   XP_050193369.1    398\n",
            "40       NWZ27654.1    397\n",
            "41       NXP47142.1    397\n",
            "42       NXN75400.1    397\n",
            "43       EOA93537.1    397\n",
            "44   XP_027301265.2    397\n",
            "45     KAI6062500.1    397\n",
            "46   XP_032060635.1    397\n",
            "47   XP_032060634.1    397\n",
            "48   XP_032060632.1    397\n",
            "49   XP_032060633.1    397\n",
            "50       ERE83599.1    397\n",
            "51       NXD88147.1    392\n",
            "52     KAF5928480.1    391\n",
            "53   XP_004429134.1    391\n",
            "54   XP_015742492.1    390\n",
            "55   XP_003405511.1    389\n",
            "56   XP_049737514.1    389\n",
            "57       OWK51526.1    388\n",
            "58   XP_005495880.1    387\n",
            "59       NXH02614.1    387\n",
            "60       NXE00558.1    387\n",
            "61       NWT31592.1    387\n",
            "62       NXP93400.1    387\n",
            "63       NXQ29038.1    387\n",
            "64   XP_009097282.1    387\n",
            "65       KFO58538.1    387\n",
            "66       NXI30281.1    387\n",
            "67       NXB19268.1    387\n",
            "68       NWI73244.1    387\n",
            "69       NXR40224.1    387\n",
            "70   XP_030822960.1    387\n",
            "71       TRZ11655.1    387\n",
            "72   XP_005061498.1    387\n",
            "73       NWY67390.1    387\n",
            "74       NXL23130.1    387\n",
            "75       NXD29585.1    387\n",
            "76       NWI04967.1    387\n",
            "77       KFQ06759.1    387\n",
            "78       NXA82061.1    387\n",
            "79   XP_009945031.1    387\n",
            "80       NXH24614.1    387\n",
            "81   XP_008102284.2    387\n",
            "82       NXB64763.1    387\n",
            "83   XP_050770042.1    387\n",
            "84       KFP50929.1    387\n",
            "85   XP_037364045.1    387\n",
            "86       NXP39835.1    387\n",
            "87   XP_017588854.1    387\n",
            "88       NXE69913.1    387\n",
            "89       NXM91717.1    387\n",
            "90       NXD60288.1    387\n",
            "91       NXM40672.1    387\n",
            "92   XP_031949989.1    387\n",
            "93       NXL79903.1    387\n",
            "94       NXI81646.1    387\n",
            "95       NXX67124.1    387\n",
            "96       NWZ90145.1    387\n",
            "97       NXF27162.1    387\n",
            "98       NWW57911.1    387\n",
            "99       NWT87927.1    387\n",
            "100      NXQ69854.1    387\n",
            "101      NWY48079.1    387\n",
            "102      NWV91774.1    387\n",
            "103      NXA66614.1    387\n",
            "104  XP_030920491.1    387\n",
            "105      NWV57476.1    387\n",
            "106      NXU44835.1    387\n",
            "107      NWY30292.1    387\n",
            "108  XP_048183764.1    387\n",
            "109      NWW01856.1    387\n",
            "110      NXB06142.1    387\n",
            "111  XP_036255724.1    387\n",
            "112      NWZ09271.1    387\n",
            "113      NXV61559.1    387\n",
            "114  XP_041881672.1    387\n",
            "115      NWR23915.1    387\n",
            "116      NWY21840.1    387\n",
            "117      NWW21823.1    387\n",
            "118      NXB38165.1    387\n",
            "119      NXH80539.1    387\n",
            "120      NXC54092.1    387\n",
            "121      NXC15233.1    386\n",
            "122      NWZ02125.1    386\n",
            "123      NXO82978.1    385\n",
            "124      KFQ06829.1    383\n",
            "125  XP_009920628.1    383\n",
            "126      NWZ59689.1    383\n",
            "127  XP_010569819.1    383\n",
            "128  XP_026723422.1    382\n",
            "129  XP_009878356.1    382\n",
            "130  XP_039247220.1    381\n",
            "131  XP_009817097.1    381\n",
            "132      KFU91935.1    380\n",
            "133  XP_051497339.1    380\n",
            "134  XP_009999013.1    380\n",
            "135      NXI37888.1    380\n",
            "136      NXX34546.1    379\n",
            "137      NXS82830.1    379\n",
            "138      NXW71464.1    379\n",
            "139      RMB94097.1    379\n",
            "140      NWS86668.1    379\n",
            "141  XP_039944603.1    379\n",
            "142  XP_028929007.1    378\n",
            "143  XP_027525625.1    377\n",
            "144  XP_010299348.1    376\n",
            "145      NXY60886.1    375\n",
            "146  XP_002711050.2    374\n",
            "147  XP_013810921.1    372\n",
            "148  XP_010130474.1    372\n",
            "149      KFO87299.1    372\n",
            "150      NWH75695.1    372\n",
            "151  XP_025934002.1    372\n",
            "152      NWR62631.1    372\n",
            "153      NWR54235.1    368\n",
            "154  XP_010076830.1    367\n",
            "155  XP_038608307.1    366\n",
            "156      NXY21196.1    365\n",
            "157      NXI03193.1    365\n",
            "158  XP_042306046.1    365\n",
            "159      NXY37478.1    365\n",
            "160      NXS84474.1    365\n",
            "161      NXS58026.1    363\n",
            "162      NWH55029.1    363\n",
            "163      NWI69766.1    363\n",
            "164      KFV19286.1    363\n",
            "165  XP_009986657.1    363\n",
            "166      NXM71247.1    363\n",
            "167  XP_047910643.1    362\n",
            "168  XP_009697509.1    362\n",
            "169      NXJ72000.1    362\n",
            "170      NXI70994.1    362\n",
            "171  XP_010280753.1    362\n",
            "172      NXU81193.1    362\n",
            "173  XP_040395671.1    362\n",
            "174  XP_030324757.1    362\n",
            "175      NXT01409.1    362\n",
            "176  XP_030324758.1    362\n",
            "177      KFV59393.1    362\n",
            "178  XP_049675457.1    362\n",
            "179      NXH76878.1    362\n",
            "180  XP_009574574.1    362\n",
            "181      NXL39337.1    362\n",
            "182      NXV00436.1    362\n",
            "183      NXQ61729.1    362\n",
            "184      NWI43337.1    362\n",
            "185      NWS32225.1    362\n",
            "186      NXT80952.1    362\n",
            "187      NWW33750.1    362\n",
            "188      NXN45711.1    362\n",
            "189      NXT86479.1    362\n",
            "190  XP_009919366.1    362\n",
            "191      NXO50624.1    362\n",
            "192      NXW35724.1    362\n",
            "193  XP_009490386.1    362\n",
            "194      NWT92430.1    362\n",
            "195  XP_010167948.1    362\n",
            "196  XP_035426924.1    362\n",
            "197      KFZ50558.1    362\n",
            "198      NXF69371.1    362\n",
            "199      NWQ94818.1    362\n",
            "200  XP_039561682.1    362\n",
            "201  XP_032939078.1    362\n",
            "202      NXQ43329.1    362\n",
            "203      NWH28513.1    362\n",
            "204      NXW49134.1    362\n",
            "205      NXF51489.1    362\n",
            "206  XP_005532067.1    362\n",
            "207      KGL89196.1    362\n",
            "208      NXT31425.1    362\n",
            "209      NXV51313.1    362\n",
            "210      NWS63256.1    362\n",
            "211      NWX79868.1    362\n",
            "212      NXG82988.1    362\n",
            "213      NXS65342.1    362\n",
            "214      KFO72669.1    362\n",
            "215      KFW03584.1    362\n",
            "216      NXU29430.1    362\n",
            "217  XP_009554461.1    362\n",
            "218      NWX44293.1    362\n",
            "219      KFP64556.1    362\n",
            "220      NXQ92926.1    362\n",
            "221      NXK13987.1    362\n",
            "222      NXA21194.1    362\n",
            "223      NWY54188.1    362\n",
            "224      NXW07197.1    362\n",
            "225  XP_015507558.1    362\n",
            "226  XP_023798930.1    362\n",
            "227      NXY73501.1    362\n",
            "228      NXL65353.1    362\n",
            "229      NXR88354.1    362\n",
            "230      NWZ70274.1    362\n",
            "231      NXC11577.1    362\n",
            "232  XP_013056216.2    362\n",
            "233      NXD17333.1    362\n",
            "234      NXA56313.1    362\n",
            "235      NXN94489.1    362\n",
            "236  XP_014811193.1    362\n",
            "237      NXJ55985.1    362\n",
            "238      NWX62634.1    362\n",
            "239  XP_035164756.1    362\n",
            "240      KFP10620.1    362\n",
            "241  XP_029893226.1    362\n",
            "242  XP_009685948.1    362\n",
            "243      NXW23696.1    362\n",
            "244  XP_009646524.1    362\n",
            "245      KFQ65975.1    362\n",
            "246      KFV74254.1    362\n",
            "247      NWZ87796.1    362\n",
            "248      NXV24593.1    362\n",
            "249      KFP01288.1    362\n",
            "250      NXN50447.1    362\n",
            "251      NXT65075.1    362\n",
            "252      NWS76984.1    362\n",
            "253      NXO61779.1    362\n",
            "254  XP_031363095.1    362\n",
            "255      NWU87417.1    362\n",
            "256  XP_002194692.2    362\n",
            "257      NXE26984.1    362\n",
            "258      NXX35035.1    362\n",
            "259      NXR65036.1    362\n",
            "260      NWU10361.1    362\n",
            "261      NXR45048.1    362\n",
            "262      NXJ43630.1    362\n",
            "263      NWT11413.1    362\n",
            "264      NXQ17543.1    362\n",
            "265      RLV87413.1    362\n",
            "266      NXI18813.1    362\n",
            "267      NXE15907.1    362\n",
            "268      KFP94478.1    362\n",
            "269      NXE98628.1    362\n",
            "270  XP_005233187.1    362\n",
            "271  XP_010566796.1    362\n",
            "272  XP_038020260.1    362\n",
            "273      NWR96740.1    362\n",
            "274  XP_027667004.1    362\n",
            "275      NXR18961.1    362\n",
            "276      NWZ42847.1    362\n",
            "277      KFV58043.1    362\n",
            "278      NXS36724.1    362\n",
            "279      NWR39750.1    362\n",
            "280  XP_032853365.2    362\n",
            "281      NXP65750.1    362\n",
            "282      NXO09429.1    362\n",
            "283      NXA93616.1    362\n",
            "284      KFQ86222.1    362\n",
            "285      NXT11935.1    362\n",
            "286  XP_021399612.1    362\n",
            "287      NXM18226.1    362\n",
            "288      NWX29647.1    362\n",
            "289      NXS49946.1    362\n",
            "290      NWH36650.1    362\n",
            "291      NWR47737.1    362\n",
            "292      NXO92448.1    362\n",
            "293      NXD02861.1    362\n",
            "294      KFQ57790.1    362\n",
            "295      NXJ99619.1    362\n",
            "296      NWZ59663.1    362\n",
            "297      NXX20886.1    362\n",
            "298      NWH65177.1    362\n",
            "299      NXC89629.1    362\n",
            "300      NWV85180.1    362\n",
            "301      NXW68802.1    362\n",
            "302      NXB70105.1    362\n",
            "303      KFP45310.1    362\n",
            "304      NXQ03214.1    362\n",
            "305      NXB85087.1    362\n",
            "306      NXX01961.1    362\n",
            "307  XP_039944610.1    362\n",
            "308      NXV37421.1    362\n",
            "309      NWT42756.1    362\n",
            "310  XP_010125683.1    362\n",
            "311      NXC72574.1    362\n",
            "312      NXK68810.1    362\n",
            "313  XP_010178727.1    361\n",
            "314      NXJ84474.1    361\n",
            "315  XP_009874961.1    361\n",
            "316      NXN88074.1    361\n",
            "317      NWX10944.1    361\n",
            "318      KFP92381.1    361\n",
            "319      KFQ31819.1    361\n",
            "320      NWW71569.1    361\n",
            "321      NXH50036.1    361\n",
            "322      NXY43963.1    361\n",
            "323      NXB47356.1    360\n",
            "324      NXR75426.1    360\n",
            "325  XP_014748469.1    360\n",
            "326      KFV03779.1    360\n",
            "327  XP_009942899.1    360\n",
            "328  XP_010077830.1    360\n",
            "329      NXT21589.1    360\n",
            "330      NXH61634.1    360\n",
            "331      KFR03096.1    360\n",
            "332    KAF4803609.1    360\n",
            "333      NXX96489.1    360\n",
            "334      NXV78972.1    360\n",
            "335      NXA43338.1    360\n",
            "336      NXA15758.1    360\n",
            "337      NXW83112.1    360\n",
            "338      PKK18628.1    360\n",
            "339      NWR78708.1    360\n",
            "340      NWU71585.1    360\n",
            "341  XP_021137901.1    360\n",
            "342      NWS83030.1    360\n",
            "343      NXY90174.1    357\n",
            "344      NXK55103.1    357\n",
            "345      NXW59306.1    357\n",
            "346      NXI54610.1    357\n",
            "347      NXG71957.1    357\n",
            "348      NWI14245.1    356\n",
            "349      NXI92620.1    356\n",
            "350      KFZ65243.1    356\n",
            "351      NWJ00939.1    356\n",
            "352      NWQ71910.1    356\n",
            "353      NXE83877.1    356\n",
            "354    KAF1627081.1    356\n",
            "355      NXL00820.1    356\n",
            "356  XP_032569451.1    356\n",
            "357      KFM01488.1    356\n",
            "358  XP_009284198.1    356\n",
            "359      KFW72893.1    356\n",
            "360  XP_009330519.1    356\n",
            "361    KAF1398520.1    356\n",
            "362    KAF1487781.1    356\n",
            "363  XP_010222750.1    356\n",
            "364    KAF1492208.1    356\n",
            "365    KAF1520764.1    356\n",
            "366    KAG5211924.1    356\n",
            "367  XP_009470699.1    356\n",
            "368      NWV70057.1    356\n",
            "369  XP_032569452.1    356\n",
            "370      NWX23664.1    356\n",
            "371      NXL51931.1    356\n",
            "372      NXI44186.1    356\n",
            "373      NWI23940.1    356\n",
            "374    KAF1481622.1    356\n",
            "375    KAF1660500.1    356\n",
            "376    KAF1530849.1    356\n",
            "377    KAF1566084.1    356\n",
            "378      KFR05398.1    356\n",
            "379      KGL84003.1    356\n",
            "380      NWV30877.1    355\n",
            "381      NWV01384.1    355\n",
            "382      OXB82930.1    354\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get one protein ID for each unique count\n",
        "unique_counts = top_400_proteins.sort_values(ascending=False)\n",
        "# Get the first occurrence for each unique count using idxmin\n",
        "one_protein_per_count = top_400_proteins.groupby(top_400_proteins.values).agg(lambda x: x.iloc[0])\n",
        "\n",
        "# Print the results (protein ID and count)\n",
        "top_400_proteins\n",
        "\n",
        "# Print the results (protein ID and count)\n",
        "#print(one_protein_per_count(50))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KEKlH36pcZFI",
        "outputId": "af33582c-cb61-403f-cd27-6c8351b6b433"
      },
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                  16539\n",
              "NWV22044.1          422\n",
              "XP_005570993.2      416\n",
              "XP_024113020.1      415\n",
              "XP_014748466.1      415\n",
              "XP_030678472.1      415\n",
              "XP_032029551.1      415\n",
              "XP_011857542.1      414\n",
              "XP_028685340.1      414\n",
              "AAG41947.1          414\n",
              "XP_011727181.1      414\n",
              "NP_006112.3         414\n",
              "AFA52007.1          414\n",
              "AAB47721.1          414\n",
              "AFA52006.1          414\n",
              "XP_003845964.2      414\n",
              "XP_003906478.2      414\n",
              "AFA52002.1          414\n",
              "AFA52005.1          414\n",
              "XP_025258843.1      414\n",
              "XP_011857532.1      414\n",
              "XP_001098182.2      414\n",
              "XP_025258844.1      414\n",
              "XP_050604771.1      414\n",
              "XP_011857566.1      414\n",
              "XP_011857549.1      414\n",
              "XP_011903561.1      414\n",
              "XP_025258845.1      414\n",
              "XP_011857559.1      414\n",
              "AKI71766.1          414\n",
              "XP_004053220.3      414\n",
              "PNI33041.1          414\n",
              "NP_001104288.1      414\n",
              "AFA52004.1          414\n",
              "AFA52003.1          414\n",
              "XP_037853004.1      413\n",
              "XP_004429138.2      406\n",
              "XP_011854305.1      404\n",
              "XP_051624887.1      402\n",
              "XP_050193369.1      398\n",
              "NWZ27654.1          397\n",
              "NXP47142.1          397\n",
              "NXN75400.1          397\n",
              "EOA93537.1          397\n",
              "XP_027301265.2      397\n",
              "KAI6062500.1        397\n",
              "XP_032060635.1      397\n",
              "XP_032060634.1      397\n",
              "XP_032060632.1      397\n",
              "XP_032060633.1      397\n",
              "ERE83599.1          397\n",
              "NXD88147.1          392\n",
              "KAF5928480.1        391\n",
              "XP_004429134.1      391\n",
              "XP_015742492.1      390\n",
              "XP_003405511.1      389\n",
              "XP_049737514.1      389\n",
              "OWK51526.1          388\n",
              "XP_005495880.1      387\n",
              "NXH02614.1          387\n",
              "NXE00558.1          387\n",
              "NWT31592.1          387\n",
              "NXP93400.1          387\n",
              "NXQ29038.1          387\n",
              "XP_009097282.1      387\n",
              "KFO58538.1          387\n",
              "NXI30281.1          387\n",
              "NXB19268.1          387\n",
              "NWI73244.1          387\n",
              "NXR40224.1          387\n",
              "XP_030822960.1      387\n",
              "TRZ11655.1          387\n",
              "XP_005061498.1      387\n",
              "NWY67390.1          387\n",
              "NXL23130.1          387\n",
              "NXD29585.1          387\n",
              "NWI04967.1          387\n",
              "KFQ06759.1          387\n",
              "NXA82061.1          387\n",
              "XP_009945031.1      387\n",
              "NXH24614.1          387\n",
              "XP_008102284.2      387\n",
              "NXB64763.1          387\n",
              "XP_050770042.1      387\n",
              "KFP50929.1          387\n",
              "XP_037364045.1      387\n",
              "NXP39835.1          387\n",
              "XP_017588854.1      387\n",
              "NXE69913.1          387\n",
              "NXM91717.1          387\n",
              "NXD60288.1          387\n",
              "NXM40672.1          387\n",
              "XP_031949989.1      387\n",
              "NXL79903.1          387\n",
              "NXI81646.1          387\n",
              "NXX67124.1          387\n",
              "NWZ90145.1          387\n",
              "NXF27162.1          387\n",
              "NWW57911.1          387\n",
              "NWT87927.1          387\n",
              "NXQ69854.1          387\n",
              "NWY48079.1          387\n",
              "NWV91774.1          387\n",
              "NXA66614.1          387\n",
              "XP_030920491.1      387\n",
              "NWV57476.1          387\n",
              "NXU44835.1          387\n",
              "NWY30292.1          387\n",
              "XP_048183764.1      387\n",
              "NWW01856.1          387\n",
              "NXB06142.1          387\n",
              "XP_036255724.1      387\n",
              "NWZ09271.1          387\n",
              "NXV61559.1          387\n",
              "XP_041881672.1      387\n",
              "NWR23915.1          387\n",
              "NWY21840.1          387\n",
              "NWW21823.1          387\n",
              "NXB38165.1          387\n",
              "NXH80539.1          387\n",
              "NXC54092.1          387\n",
              "NXC15233.1          386\n",
              "NWZ02125.1          386\n",
              "NXO82978.1          385\n",
              "KFQ06829.1          383\n",
              "XP_009920628.1      383\n",
              "NWZ59689.1          383\n",
              "XP_010569819.1      383\n",
              "XP_026723422.1      382\n",
              "XP_009878356.1      382\n",
              "XP_039247220.1      381\n",
              "XP_009817097.1      381\n",
              "KFU91935.1          380\n",
              "XP_051497339.1      380\n",
              "XP_009999013.1      380\n",
              "NXI37888.1          380\n",
              "NXX34546.1          379\n",
              "NXS82830.1          379\n",
              "NXW71464.1          379\n",
              "RMB94097.1          379\n",
              "NWS86668.1          379\n",
              "XP_039944603.1      379\n",
              "XP_028929007.1      378\n",
              "XP_027525625.1      377\n",
              "XP_010299348.1      376\n",
              "NXY60886.1          375\n",
              "XP_002711050.2      374\n",
              "XP_013810921.1      372\n",
              "XP_010130474.1      372\n",
              "KFO87299.1          372\n",
              "NWH75695.1          372\n",
              "XP_025934002.1      372\n",
              "NWR62631.1          372\n",
              "NWR54235.1          368\n",
              "XP_010076830.1      367\n",
              "XP_038608307.1      366\n",
              "NXY21196.1          365\n",
              "NXI03193.1          365\n",
              "XP_042306046.1      365\n",
              "NXY37478.1          365\n",
              "NXS84474.1          365\n",
              "NXS58026.1          363\n",
              "NWH55029.1          363\n",
              "NWI69766.1          363\n",
              "KFV19286.1          363\n",
              "XP_009986657.1      363\n",
              "NXM71247.1          363\n",
              "XP_047910643.1      362\n",
              "XP_009697509.1      362\n",
              "NXJ72000.1          362\n",
              "NXI70994.1          362\n",
              "XP_010280753.1      362\n",
              "NXU81193.1          362\n",
              "XP_040395671.1      362\n",
              "XP_030324757.1      362\n",
              "NXT01409.1          362\n",
              "XP_030324758.1      362\n",
              "KFV59393.1          362\n",
              "XP_049675457.1      362\n",
              "NXH76878.1          362\n",
              "XP_009574574.1      362\n",
              "NXL39337.1          362\n",
              "NXV00436.1          362\n",
              "NXQ61729.1          362\n",
              "NWI43337.1          362\n",
              "NWS32225.1          362\n",
              "NXT80952.1          362\n",
              "NWW33750.1          362\n",
              "NXN45711.1          362\n",
              "NXT86479.1          362\n",
              "XP_009919366.1      362\n",
              "NXO50624.1          362\n",
              "NXW35724.1          362\n",
              "XP_009490386.1      362\n",
              "NWT92430.1          362\n",
              "XP_010167948.1      362\n",
              "XP_035426924.1      362\n",
              "KFZ50558.1          362\n",
              "NXF69371.1          362\n",
              "NWQ94818.1          362\n",
              "XP_039561682.1      362\n",
              "XP_032939078.1      362\n",
              "NXQ43329.1          362\n",
              "NWH28513.1          362\n",
              "NXW49134.1          362\n",
              "NXF51489.1          362\n",
              "XP_005532067.1      362\n",
              "KGL89196.1          362\n",
              "NXT31425.1          362\n",
              "NXV51313.1          362\n",
              "NWS63256.1          362\n",
              "NWX79868.1          362\n",
              "NXG82988.1          362\n",
              "NXS65342.1          362\n",
              "KFO72669.1          362\n",
              "KFW03584.1          362\n",
              "NXU29430.1          362\n",
              "XP_009554461.1      362\n",
              "NWX44293.1          362\n",
              "KFP64556.1          362\n",
              "NXQ92926.1          362\n",
              "NXK13987.1          362\n",
              "NXA21194.1          362\n",
              "NWY54188.1          362\n",
              "NXW07197.1          362\n",
              "XP_015507558.1      362\n",
              "XP_023798930.1      362\n",
              "NXY73501.1          362\n",
              "NXL65353.1          362\n",
              "NXR88354.1          362\n",
              "NWZ70274.1          362\n",
              "NXC11577.1          362\n",
              "XP_013056216.2      362\n",
              "NXD17333.1          362\n",
              "NXA56313.1          362\n",
              "NXN94489.1          362\n",
              "XP_014811193.1      362\n",
              "NXJ55985.1          362\n",
              "NWX62634.1          362\n",
              "XP_035164756.1      362\n",
              "KFP10620.1          362\n",
              "XP_029893226.1      362\n",
              "XP_009685948.1      362\n",
              "NXW23696.1          362\n",
              "XP_009646524.1      362\n",
              "KFQ65975.1          362\n",
              "KFV74254.1          362\n",
              "NWZ87796.1          362\n",
              "NXV24593.1          362\n",
              "KFP01288.1          362\n",
              "NXN50447.1          362\n",
              "NXT65075.1          362\n",
              "NWS76984.1          362\n",
              "NXO61779.1          362\n",
              "XP_031363095.1      362\n",
              "NWU87417.1          362\n",
              "XP_002194692.2      362\n",
              "NXE26984.1          362\n",
              "NXX35035.1          362\n",
              "NXR65036.1          362\n",
              "NWU10361.1          362\n",
              "NXR45048.1          362\n",
              "NXJ43630.1          362\n",
              "NWT11413.1          362\n",
              "NXQ17543.1          362\n",
              "RLV87413.1          362\n",
              "NXI18813.1          362\n",
              "NXE15907.1          362\n",
              "KFP94478.1          362\n",
              "NXE98628.1          362\n",
              "XP_005233187.1      362\n",
              "XP_010566796.1      362\n",
              "XP_038020260.1      362\n",
              "NWR96740.1          362\n",
              "XP_027667004.1      362\n",
              "NXR18961.1          362\n",
              "NWZ42847.1          362\n",
              "KFV58043.1          362\n",
              "NXS36724.1          362\n",
              "NWR39750.1          362\n",
              "XP_032853365.2      362\n",
              "NXP65750.1          362\n",
              "NXO09429.1          362\n",
              "NXA93616.1          362\n",
              "KFQ86222.1          362\n",
              "NXT11935.1          362\n",
              "XP_021399612.1      362\n",
              "NXM18226.1          362\n",
              "NWX29647.1          362\n",
              "NXS49946.1          362\n",
              "NWH36650.1          362\n",
              "NWR47737.1          362\n",
              "NXO92448.1          362\n",
              "NXD02861.1          362\n",
              "KFQ57790.1          362\n",
              "NXJ99619.1          362\n",
              "NWZ59663.1          362\n",
              "NXX20886.1          362\n",
              "NWH65177.1          362\n",
              "NXC89629.1          362\n",
              "NWV85180.1          362\n",
              "NXW68802.1          362\n",
              "NXB70105.1          362\n",
              "KFP45310.1          362\n",
              "NXQ03214.1          362\n",
              "NXB85087.1          362\n",
              "NXX01961.1          362\n",
              "XP_039944610.1      362\n",
              "NXV37421.1          362\n",
              "NWT42756.1          362\n",
              "XP_010125683.1      362\n",
              "NXC72574.1          362\n",
              "NXK68810.1          362\n",
              "XP_010178727.1      361\n",
              "NXJ84474.1          361\n",
              "XP_009874961.1      361\n",
              "NXN88074.1          361\n",
              "NWX10944.1          361\n",
              "KFP92381.1          361\n",
              "KFQ31819.1          361\n",
              "NWW71569.1          361\n",
              "NXH50036.1          361\n",
              "NXY43963.1          361\n",
              "NXB47356.1          360\n",
              "NXR75426.1          360\n",
              "XP_014748469.1      360\n",
              "KFV03779.1          360\n",
              "XP_009942899.1      360\n",
              "XP_010077830.1      360\n",
              "NXT21589.1          360\n",
              "NXH61634.1          360\n",
              "KFR03096.1          360\n",
              "KAF4803609.1        360\n",
              "NXX96489.1          360\n",
              "NXV78972.1          360\n",
              "NXA43338.1          360\n",
              "NXA15758.1          360\n",
              "NXW83112.1          360\n",
              "PKK18628.1          360\n",
              "NWR78708.1          360\n",
              "NWU71585.1          360\n",
              "XP_021137901.1      360\n",
              "NWS83030.1          360\n",
              "NXY90174.1          357\n",
              "NXK55103.1          357\n",
              "NXW59306.1          357\n",
              "NXI54610.1          357\n",
              "NXG71957.1          357\n",
              "NWI14245.1          356\n",
              "NXI92620.1          356\n",
              "KFZ65243.1          356\n",
              "NWJ00939.1          356\n",
              "NWQ71910.1          356\n",
              "NXE83877.1          356\n",
              "KAF1627081.1        356\n",
              "NXL00820.1          356\n",
              "XP_032569451.1      356\n",
              "KFM01488.1          356\n",
              "XP_009284198.1      356\n",
              "KFW72893.1          356\n",
              "XP_009330519.1      356\n",
              "KAF1398520.1        356\n",
              "KAF1487781.1        356\n",
              "XP_010222750.1      356\n",
              "KAF1492208.1        356\n",
              "KAF1520764.1        356\n",
              "KAG5211924.1        356\n",
              "XP_009470699.1      356\n",
              "NWV70057.1          356\n",
              "XP_032569452.1      356\n",
              "NWX23664.1          356\n",
              "NXL51931.1          356\n",
              "NXI44186.1          356\n",
              "NWI23940.1          356\n",
              "KAF1481622.1        356\n",
              "KAF1660500.1        356\n",
              "KAF1530849.1        356\n",
              "KAF1566084.1        356\n",
              "KFR05398.1          356\n",
              "KGL84003.1          356\n",
              "NWV30877.1          355\n",
              "NWV01384.1          355\n",
              "OXB82930.1          354\n",
              "XP_033722538.1      354\n",
              "KFQ18418.1          354\n",
              "OXB56691.1          354\n",
              "XP_008937613.1      354\n",
              "NXP03931.1          352\n",
              "AAA36153.1          347\n",
              "XP_048219940.1      346\n",
              "XP_012887899.1      346\n",
              "NXJ58318.1          344\n",
              "XP_029893225.2      344\n",
              "XP_017525321.2      343\n",
              "XP_036771076.1      343\n",
              "KAI5931539.1        342\n",
              "XP_011785995.1      341\n",
              "XP_010372265.1      341\n",
              "XP_033077867.1      341\n",
              "XP_011785997.1      341\n",
              "Name: Proteins, dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 58
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Print the results (protein ID and count)\n",
        "print(one_protein_per_count)"
      ],
      "metadata": {
        "id": "ur2U1jos613a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "top_400_proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "X3SGayWDqleD",
        "outputId": "028fd87e-252d-4115-e536-a5e230af59d0"
      },
      "execution_count": 49,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                  27100\n",
              "XP_045876752.1     1172\n",
              "XP_002918275.1     1124\n",
              "XP_022358799.1     1115\n",
              "XP_049497431.1     1115\n",
              "XP_007076757.1     1115\n",
              "XP_045350756.1     1115\n",
              "XP_032730864.1     1114\n",
              "XP_045350757.1     1112\n",
              "XP_032161267.1     1111\n",
              "XP_004767919.1     1111\n",
              "XP_044101981.1     1111\n",
              "XP_047551750.1     1111\n",
              "XP_039092171.1     1053\n",
              "KAF0880087.1       1046\n",
              "XP_022441863.1      912\n",
              "XP_029089490.1      908\n",
              "XP_028700691.1      880\n",
              "XP_005550260.1      880\n",
              "NP_001253266.1      880\n",
              "XP_004632680.1      814\n",
              "XP_048195999.1      807\n",
              "AKI72178.1          797\n",
              "AKI72177.1          797\n",
              "NP_000414.2         797\n",
              "AKI72621.1          795\n",
              "AKI72180.1          793\n",
              "KAI4066109.1        793\n",
              "AAB81946.1          793\n",
              "AFA52006.1          780\n",
              "PNI33042.1          779\n",
              "XP_030856700.1      779\n",
              "XP_034790894.1      776\n",
              "XP_034790893.1      776\n",
              "XP_008954692.2      776\n",
              "XP_003475044.2      774\n",
              "XP_005006094.1      774\n",
              "XP_005570993.2      767\n",
              "XP_032029551.1      766\n",
              "XP_030678472.1      766\n",
              "XP_011903561.1      765\n",
              "XP_011857549.1      765\n",
              "XP_011857542.1      765\n",
              "XP_001098182.2      765\n",
              "XP_050604771.1      765\n",
              "XP_025258844.1      765\n",
              "XP_011857559.1      765\n",
              "XP_011727181.1      765\n",
              "XP_011857532.1      765\n",
              "XP_025258843.1      765\n",
              "XP_025258845.1      765\n",
              "XP_003906478.2      765\n",
              "XP_028685340.1      765\n",
              "XP_011857566.1      765\n",
              "XP_024113020.1      765\n",
              "AFA52005.1          764\n",
              "XP_004053220.3      764\n",
              "AFA52003.1          764\n",
              "XP_003845964.2      764\n",
              "AAB47721.1          764\n",
              "XP_037853004.1      764\n",
              "AFA52004.1          764\n",
              "AFA52007.1          764\n",
              "NP_006112.3         764\n",
              "AAG41947.1          764\n",
              "PNI33041.1          764\n",
              "AFA52002.1          764\n",
              "AKI71766.1          762\n",
              "NP_001104288.1      754\n",
              "NP_001092039.1      750\n",
              "AKI72179.1          746\n",
              "XP_011830060.1      738\n",
              "XP_003896357.1      738\n",
              "XP_041531508.1      731\n",
              "XP_049979777.1      731\n",
              "XP_048292811.1      731\n",
              "XP_038175877.1      731\n",
              "XP_005363739.1      731\n",
              "XP_003497018.1      728\n",
              "XP_035307851.1      728\n",
              "XP_007450704.1      726\n",
              "XP_049979776.1      725\n",
              "XP_007450706.1      725\n",
              "XP_007450705.1      725\n",
              "XP_009246042.2      721\n",
              "XP_002823331.2      721\n",
              "XP_008684476.1      717\n",
              "ERE89181.1          711\n",
              "KAB0376067.1        710\n",
              "KAB0344636.1        710\n",
              "XP_043318553.1      709\n",
              "XP_046952093.1      707\n",
              "XP_043318554.1      707\n",
              "VFV36904.1          707\n",
              "XP_003982813.1      707\n",
              "XP_025785714.1      707\n",
              "XP_003921263.1      705\n",
              "XP_007195818.1      704\n",
              "XP_036718747.1      699\n",
              "XP_047417547.1      699\n",
              "XP_006157541.1      698\n",
              "XP_029785028.1      691\n",
              "XP_003733643.1      688\n",
              "XP_026958275.1      687\n",
              "XP_004265635.1      687\n",
              "XP_019783093.1      687\n",
              "XP_036038246.1      686\n",
              "XP_030713115.1      686\n",
              "XP_042128705.1      686\n",
              "XP_028725629.1      686\n",
              "XP_006991260.1      686\n",
              "TEA24756.1          685\n",
              "XP_012306220.1      684\n",
              "XP_049737514.1      682\n",
              "XP_025234413.1      682\n",
              "XP_003405511.1      682\n",
              "XP_011785995.1      681\n",
              "XP_010372265.1      681\n",
              "XP_011785996.1      681\n",
              "XP_025234412.1      681\n",
              "XP_033077867.1      681\n",
              "XP_011785997.1      681\n",
              "XP_028634999.1      681\n",
              "XP_003252555.1      680\n",
              "BAG36698.1          678\n",
              "XP_032108832.1      677\n",
              "XP_045022283.1      674\n",
              "XP_006054012.1      674\n",
              "XP_004007775.1      673\n",
              "NP_776945.1         673\n",
              "XP_017380605.1      673\n",
              "XP_019815249.1      673\n",
              "KAI4570259.1        672\n",
              "KAI4584709.1        672\n",
              "XP_020769440.1      672\n",
              "NP_001182597.1      672\n",
              "XP_041583813.1      671\n",
              "XP_040116376.1      671\n",
              "XP_005678993.1      671\n",
              "XP_033050550.1      670\n",
              "XP_032148987.1      670\n",
              "XP_008820781.1      670\n",
              "XP_032148988.1      670\n",
              "XP_017357224.1      670\n",
              "XP_037584898.1      670\n",
              "XP_032148984.1      670\n",
              "XP_017710454.1      670\n",
              "XP_012305692.1      670\n",
              "XP_024622854.1      669\n",
              "XP_032498638.1      669\n",
              "XP_003939248.1      669\n",
              "XP_011729152.1      668\n",
              "XP_032498639.1      668\n",
              "XP_007980329.1      668\n",
              "KAG5211924.1        667\n",
              "XP_031236962.1      666\n",
              "XP_018886394.1      665\n",
              "EHH17509.1          665\n",
              "XP_010387823.1      665\n",
              "XP_002807152.2      665\n",
              "EHH52338.1          665\n",
              "XP_021506511.1      663\n",
              "NP_000080.2         661\n",
              "KAI2546688.1        661\n",
              "XP_001168894.1      661\n",
              "EAW76795.1          661\n",
              "AAH42586.1          661\n",
              "XP_032616930.1      661\n",
              "XP_003809763.2      661\n",
              "CAA98969.1          661\n",
              "AAB69977.1          660\n",
              "NP_001003187.1      658\n",
              "XP_025327115.1      658\n",
              "XP_011810471.1      658\n",
              "XP_011810470.1      658\n",
              "XP_023082504.2      658\n",
              "XP_040820670.1      656\n",
              "C0HJP6.1            656\n",
              "XP_025859557.1      656\n",
              "XP_008569871.1      652\n",
              "XP_024105396.1      650\n",
              "VTJ77294.1          641\n",
              "XP_051007948.1      641\n",
              "XP_026247293.1      640\n",
              "XP_015343716.1      640\n",
              "XP_027811242.1      640\n",
              "XP_005331878.1      640\n",
              "KAF7462415.1        640\n",
              "XP_004582363.1      639\n",
              "XP_021045970.1      638\n",
              "KAI4586844.1        638\n",
              "KAI4576835.1        638\n",
              "ELW49126.1          636\n",
              "XP_032029554.1      634\n",
              "XP_003252138.1      634\n",
              "XP_032029555.1      634\n",
              "XP_007105361.1      630\n",
              "XP_020019043.1      628\n",
              "XP_029414203.1      628\n",
              "AAB93981.1          626\n",
              "AAA36153.1          615\n",
              "XP_004394135.1      614\n",
              "XP_025728689.1      614\n",
              "XP_027428797.1      613\n",
              "XP_050638861.1      608\n",
              "XP_050638860.1      608\n",
              "XP_045350759.1      607\n",
              "XP_019298883.1      607\n",
              "C0HJP8.1            604\n",
              "XP_033077866.1      603\n",
              "XP_017716918.1      603\n",
              "XP_011785993.1      603\n",
              "XP_011785992.1      603\n",
              "XP_011785994.1      603\n",
              "XP_010372266.1      603\n",
              "XP_036285294.1      600\n",
              "XP_015413820.1      599\n",
              "XP_005876138.1      599\n",
              "XP_036184054.1      599\n",
              "EFB18514.1          599\n",
              "XP_006085025.1      598\n",
              "XP_023606493.1      598\n",
              "XP_002711050.2      598\n",
              "EPQ13864.1          598\n",
              "XP_008147226.1      596\n",
              "KAF6319303.1        595\n",
              "P85154.3            594\n",
              "CAD7690419.1        593\n",
              "XP_049749558.1      593\n",
              "XP_010585560.1      593\n",
              "VCW49962.1          590\n",
              "XP_014748466.1      586\n",
              "CAA68709.1          583\n",
              "XP_036285296.1      583\n",
              "XP_028012762.1      581\n",
              "XP_023066808.2      578\n",
              "XP_023066809.2      578\n",
              "C0HJN6.1            576\n",
              "XP_004429138.2      574\n",
              "XP_012505708.1      566\n",
              "XP_003782725.1      565\n",
              "XP_045040128.1      563\n",
              "XP_024431324.1      563\n",
              "ELK19008.1          562\n",
              "XP_028685338.1      560\n",
              "XP_050604779.1      559\n",
              "XP_028929007.1      555\n",
              "XP_045720640.1      553\n",
              "XP_034876481.1      553\n",
              "XP_032258023.1      552\n",
              "XP_035979250.1      552\n",
              "XP_006730730.1      548\n",
              "XP_020019045.1      544\n",
              "XP_007937996.1      542\n",
              "NP_005546.2         541\n",
              "KAI4066093.1        541\n",
              "EAW96627.1          541\n",
              "C0HJN4.1            541\n",
              "XP_045409884.1      539\n",
              "EAW96626.1          535\n",
              "XP_038608307.1      533\n",
              "XP_021545228.1      532\n",
              "XP_017525321.2      530\n",
              "XP_036771076.1      530\n",
              "XP_017357223.1      529\n",
              "XP_017357222.1      529\n",
              "XP_032148981.1      529\n",
              "KAF5928480.1        528\n",
              "XP_004429134.1      528\n",
              "XP_050604781.1      525\n",
              "NWV22044.1          524\n",
              "MXQ85928.1          521\n",
              "KAF4010207.1        518\n",
              "XP_025258838.1      516\n",
              "XP_025258841.1      516\n",
              "XP_025258836.1      516\n",
              "XP_025258837.1      516\n",
              "XP_011857520.1      513\n",
              "XP_011857508.1      513\n",
              "XP_001098486.2      513\n",
              "XP_011857498.1      513\n",
              "XP_003906481.2      513\n",
              "XP_008001534.1      513\n",
              "XP_005570994.2      513\n",
              "XP_005570994.1      513\n",
              "XP_011857514.1      513\n",
              "ERE83599.1          510\n",
              "XP_008954697.1      510\n",
              "NP_005545.1         510\n",
              "XP_004053223.3      510\n",
              "BAG61372.1          510\n",
              "AAH69269.1          510\n",
              "AAP36273.1          510\n",
              "AKI72623.1          510\n",
              "AKI70419.1          510\n",
              "ALQ33447.1          509\n",
              "AKI72619.1          509\n",
              "BAF82934.1          509\n",
              "XP_021569231.1      508\n",
              "KAI5237728.1        508\n",
              "KAI2565842.1        506\n",
              "AAH24292.1          506\n",
              "NP_000415.2         506\n",
              "KYO17269.1          506\n",
              "XP_006275766.2      505\n",
              "XP_002752551.1      504\n",
              "EAW96635.1          504\n",
              "XP_023066816.1      504\n",
              "XP_023066815.1      504\n",
              "XP_026310465.1      504\n",
              "EAW96630.1          504\n",
              "XP_010372278.2      503\n",
              "AKI72618.1          503\n",
              "XP_018893529.1      500\n",
              "NP_001266996.1      500\n",
              "XP_024113012.1      500\n",
              "XP_024113013.1      500\n",
              "PNI33053.1          500\n",
              "XP_004053218.1      500\n",
              "XP_005571002.1      499\n",
              "XP_030856703.1      499\n",
              "XP_005571002.2      499\n",
              "NXP47142.1          499\n",
              "NXN75400.1          499\n",
              "XP_011857375.1      499\n",
              "XP_031307639.1      499\n",
              "AAC41769.1          499\n",
              "XP_031307638.1      499\n",
              "XP_006189294.1      499\n",
              "XP_002823309.2      499\n",
              "XP_032335790.1      499\n",
              "XP_011785979.1      498\n",
              "XP_009646521.2      498\n",
              "XP_011785980.1      498\n",
              "XP_011785978.1      498\n",
              "XP_011785976.1      498\n",
              "XP_005571000.1      498\n",
              "XP_005571000.2      498\n",
              "XP_011785975.1      498\n",
              "XP_011785982.1      498\n",
              "XP_050604774.1      498\n",
              "XP_008001528.2      498\n",
              "XP_011727188.1      498\n",
              "XP_015007267.2      498\n",
              "XP_011785981.1      498\n",
              "KFQ25477.1          498\n",
              "AAC41768.1          498\n",
              "XP_008001529.2      498\n",
              "XP_010189784.1      498\n",
              "KFP10617.1          498\n",
              "XP_045376332.1      498\n",
              "XP_033077873.1      497\n",
              "XP_037851964.1      497\n",
              "EHH20757.1          497\n",
              "XP_025232694.1      497\n",
              "XP_010372273.2      496\n",
              "XP_033077871.1      496\n",
              "XP_009179048.3      495\n",
              "XP_036310739.1      495\n",
              "XP_011903568.1      495\n",
              "XP_025258857.1      495\n",
              "XP_025258856.1      495\n",
              "XP_025258851.1      495\n",
              "XP_025258852.1      495\n",
              "XP_009920628.1      495\n",
              "XP_010569819.1      495\n",
              "KFQ06829.1          495\n",
              "NWZ59689.1          495\n",
              "XP_025258854.1      495\n",
              "XP_025258859.1      495\n",
              "XP_025258850.1      495\n",
              "EHH20756.1          495\n",
              "XP_009554503.1      495\n",
              "KFO72666.1          495\n",
              "XP_011727191.1      495\n",
              "EHH66300.1          494\n",
              "XP_006210056.1      494\n",
              "NXD88147.1          494\n",
              "XP_031534140.1      494\n",
              "XP_021071953.1      494\n",
              "XP_051497339.1      494\n",
              "XP_009999013.1      494\n",
              "NP_775109.2         494\n",
              "KFU91935.1          494\n",
              "KAI4066094.1        493\n",
              "XP_035112583.1      493\n",
              "NXX34546.1          493\n",
              "KAF3826215.1        493\n",
              "XP_011857399.1      493\n",
              "XP_011857405.1      493\n",
              "EHH20755.1          491\n",
              "XP_002823313.2      491\n",
              "XP_024113011.1      491\n",
              "XP_003252127.1      490\n",
              "XP_011903569.1      490\n",
              "XP_015007266.2      489\n",
              "XP_015274225.1      489\n",
              "XP_008102284.2      489\n",
              "XP_011857388.1      489\n",
              "XP_011903562.1      489\n",
              "Name: Proteins, dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 49
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "###   Direct Matching for Collagen-Motif Peptides"
      ],
      "metadata": {
        "id": "5q1F42rByuv1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "peptides_to_examine = [\n",
        "    'IITHPNFNGNTLDNDIMLIK',\n",
        "    'LGEHNIDVLEGNEQFINAAK',\n",
        "    'YSQGNVSAVGVTYDGHTALTR',\n",
        "    'TPPAGVFYQGWSATPIANGSLGHDIHHPR',\n",
        "    'GPPGPMGPPGIAGPPGESGR',\n",
        "    'SDLEMQYETLQEELMALKK'\n",
        "]\n",
        "\n",
        "\n",
        "print(\"Peptides from sequence_position_dict:\")\n",
        "for peptide in peptides_to_examine:\n",
        "    if peptide in sequence_position_dict:\n",
        "        details = sequence_position_dict[peptide]\n",
        "        # Check if 'gene' and 'position' keys exist before accessing\n",
        "        gene = details.get('gene', 'Gene information not available')\n",
        "        position = details.get('position', 'Position information not available')\n",
        "        print(f\"Peptide: {peptide}, Gene: {gene}, Position: {position}\")\n",
        "    else:\n",
        "        print(f\"Peptide: {peptide} not found in sequence_position_dict.\")\n"
      ],
      "metadata": {
        "id": "2I8htize_hTv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "L5E8amGF_vz8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the directory for saving the output file\n",
        "output_directory = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/sample1/'\n",
        "os.makedirs(output_directory, exist_ok=True)\n",
        "\n",
        "# Get the current timestamp\n",
        "now = datetime.now().strftime(\"%Y-%m-%d_%H-%M-%S\")\n",
        "\n",
        "# Create the filename with study names and timestamp\n",
        "filename = f\"{output_directory}{'_'.join(STUDY_NAMES)}_{now}_sequence_positions.json\"\n",
        "\n",
        "# Assuming 'combined_df' is to be saved, convert it to a dictionary\n",
        "sequence_position_dict = combined_df.to_dict()\n",
        "with open(filename, 'w') as json_file:\n",
        "    json.dump(sequence_position_dict, json_file, indent=4)\n",
        "\n",
        "print(f'Dictionary saved to JSON file: {filename}')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5KMlTsTGAIFs",
        "outputId": "a13bcc3a-ed6d-488e-bcae-c5a44de3536d"
      },
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Dictionary saved to JSON file: /content/drive/MyDrive/Colab_Notebooks/NovorCloud/sample1/CE&Tuuli_CATE_2024-03-17_13-24-29_sequence_positions.json\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# #Peek at the output of the Dictionaries =============================== report\n",
        "# random_entries = random.sample(list(sequence_position_dict.items()), 100)\n",
        "# for key, value in random_entries:\n",
        "#     print(f\"Key: {key}, Value: {value}\")\n",
        "\n",
        "# print(sequence_position_dict)\n",
        "# #  matched_peptides in sequence_position_dict  dictionary\n",
        "# for peptide, details in islice(sequence_position_dict.items(), 10):\n",
        "#     print(f\"Peptide: {peptide}, Gene: {details['gene']}, Position: {details['position']}\")\n"
      ],
      "metadata": {
        "id": "mTnMikukFU2M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Step 4: Identify Unmatched Peptides\n",
        "\n",
        "After attempting direct matching, identify peptides that remain unmatched:"
      ],
      "metadata": {
        "id": "XMJpkD6fs5ov"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "combined_df"
      ],
      "metadata": {
        "id": "YzeTBMdfbqdn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "combined_df"
      ],
      "metadata": {
        "id": "bbnyyFE4q0Kk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eCpbz9TZ0dLV"
      },
      "outputs": [],
      "source": [
        "# #Reporting\n",
        "# combined_df.head()\n",
        "# #Check the df again!\n",
        "# combined_df.head()\n",
        "# combined_df.shape\n",
        "# combined_df.info()\n",
        "# #combined_df.describe()\n",
        "# combined_df['Sequence'].nunique()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yDWInnY-Rfg4"
      },
      "outputs": [],
      "source": [
        "# #To view the matches\n",
        "# print(\"Collagen Matches:\")\n",
        "# for peptide, details in sequence_position_dict.items():\n",
        "#     print(f\"Peptide: {peptide}, Gene: {details['gene']}, Position: {details['position']}\")\n",
        "# print(\"\\nNCP Matches:\")\n",
        "# for peptide, details in NCP_peptide_positions_dict.items():\n",
        "#     print(f\"Peptide: {peptide}, Gene: {details['gene']}, Position: {details['position']}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3FNQCIQo1_wT"
      },
      "outputs": [],
      "source": [
        "# # Report on the most frequent proteins - this is very SLOW!!!\n",
        "# report_df = extract_protein_frequencies_and_associations(combined_df, 'Proteins')\n",
        "\n",
        "# # Display or export `report_df` as needed\n",
        "# print(report_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "j0vgjVsK37iV"
      },
      "outputs": [],
      "source": [
        "# #Writing this report.csv\n",
        "# # Construct the file path first\n",
        "# csv_file_path = f'{BASE_PATH}/{STUDY_NAME}_prot_freq.csv'\n",
        "\n",
        "# # Export the DataFrame to CSV using the correct method name\n",
        "# report_df.to_csv(csv_file_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "w253hUG-K6VC"
      },
      "outputs": [],
      "source": [
        "# ##Check the df again!\n",
        "# #combined_df.head()\n",
        "# #combined_df.shape\n",
        "# #combined_df.info()\n",
        "# ##combined_df.describe()\n",
        "# #combined_df['Sequence'].nunique()\n",
        "# #combined_df['Miss.Clv.Sites'].nunique()\n",
        "# combined_df.head()\n",
        "# list(combined_df.columns)\n",
        "# #print(combined_df[['GNC&spp', 'Sequence','Proteins', 'Positions', 'start_position']].head(40))\n",
        "# spp_GNC_counts = combined_df['GNC&spp'].value_counts()\n",
        "\n",
        "# # Print the counts of different entries\n",
        "# print(spp_GNC_counts)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Generate smaller df - iltered_combined_df- for the Expand step\n",
        "# Replace 'None' and empty strings with np.nan for uniformity\n",
        "combined_df['GNC&spp'] = combined_df['GNC&spp'].replace({None: np.nan, '': np.nan})\n",
        "\n",
        "# Filter rows where 'GNC&spp' is not NaN\n",
        "filtered_combined_df = combined_df.dropna(subset=['GNC&spp'])\n",
        "\n",
        "# Now, filtered_combined_df contains only the rows where 'GNC&spp' is not None, NaN, or blank\n",
        "# You can perform your operations on filtered_combined_df\n",
        "\n",
        "# Print the head of the filtered DataFrame to verify the filtering\n",
        "print(filtered_combined_df.head())\n"
      ],
      "metadata": {
        "id": "hp2-BVEQFbg3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#COUNT GNC\n",
        "unique_names_GNC = filtered_combined_df['GNC&spp'].unique()\n",
        "name_counts_GNC = filtered_combined_df['GNC&spp'].value_counts()\n",
        "print(\"Unique GNC:\", unique_names_GNC)\n",
        "print(\"Counts of each GNC:\", name_counts_GNC)"
      ],
      "metadata": {
        "id": "g4zFIJqMkIaO",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 211
        },
        "outputId": "c2d36cbf-6bf8-4b72-8dca-98c3c3b35efc"
      },
      "execution_count": 60,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'filtered_combined_df' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-60-219bac11c768>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#COUNT GNC\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0munique_names_GNC\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfiltered_combined_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GNC&spp'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mname_counts_GNC\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfiltered_combined_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'GNC&spp'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Unique GNC:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0munique_names_GNC\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Counts of each GNC:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname_counts_GNC\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'filtered_combined_df' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nOUOVdeFe1XS"
      },
      "source": [
        "## Expand df"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# # Assuming 'combined_df' is original dataframe with columns [''GNC&spp'', 'start_position', 'peptide_length', 'sequence']\n",
        "# df_sorted = combined_df.sort_values(by=['GNC&spp','start_position', 'peptide_length','Sequence'], ascending=[True, False, False,True])"
      ],
      "metadata": {
        "id": "ACJlWUkrleA-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 61,
      "metadata": {
        "id": "1RHve46MY0BD"
      },
      "outputs": [],
      "source": [
        "# Expand combined_df to detail each peptide's amino acid positions and associated PTMs\n",
        "#exp_df = expand_df_to(filtered_combined_df) only contains collagen\n",
        "exp_df = expand_df_to(combined_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 70,
      "metadata": {
        "id": "Pn9kqvWuKPQS",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7ad14594-ab3b-4c99-adbe-88217bb53437"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "     id sample_name    GNC      spp aa  position ptm confidence  spectraId  \\\n",
            "0  None        Cz16  TRYUP  Sus_sco  I        69           None       6381   \n",
            "1  None        Cz16  TRYUP  Sus_sco  I        70           None       6381   \n",
            "2  None        Cz16  TRYUP  Sus_sco  T        71           None       6381   \n",
            "3  None        Cz16  TRYUP  Sus_sco  H        72           None       6381   \n",
            "4  None        Cz16  TRYUP  Sus_sco  P        73           None       6381   \n",
            "\n",
            "            mz  z      pepMass     err         score  scanNum  \\\n",
            "0  2300.162702  2  2300.159002  0.0037  1.528900e-10     6381   \n",
            "1  2300.162702  2  2300.159002  0.0037  1.528900e-10     6381   \n",
            "2  2300.162702  2  2300.159002  0.0037  1.528900e-10     6381   \n",
            "3  2300.162702  2  2300.159002  0.0037  1.528900e-10     6381   \n",
            "4  2300.162702  2  2300.159002  0.0037  1.528900e-10     6381   \n",
            "\n",
            "                                   peptide_gene_name    RT   ppm  \n",
            "0  5XW1_A/P00761.1/1AKS_A/1AVW_A/1FMG_A/1AN1_E/3M...  None  None  \n",
            "1  5XW1_A/P00761.1/1AKS_A/1AVW_A/1FMG_A/1AN1_E/3M...  None  None  \n",
            "2  5XW1_A/P00761.1/1AKS_A/1AVW_A/1FMG_A/1AN1_E/3M...  None  None  \n",
            "3  5XW1_A/P00761.1/1AKS_A/1AVW_A/1FMG_A/1AN1_E/3M...  None  None  \n",
            "4  5XW1_A/P00761.1/1AKS_A/1AVW_A/1FMG_A/1AN1_E/3M...  None  None  \n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(1075011, 18)"
            ]
          },
          "metadata": {},
          "execution_count": 70
        }
      ],
      "source": [
        "# Display the first few rows of the expanded DataFrame to verify the result\n",
        "print(exp_df.head())\n",
        "list(exp_df.columns)\n",
        "exp_df.shape"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "unique_names_GNC = exp_df['GNC'].unique()\n",
        "name_counts_GNC = exp_df['GNC'].value_counts()\n",
        "print(\"Unique sample names:\", unique_names_GNC)\n",
        "print(\"Counts of each sample name:\", name_counts_GNC)"
      ],
      "metadata": {
        "id": "J_2FITDLfNrs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 63,
      "metadata": {
        "id": "xuoVc8SuyR0G"
      },
      "outputs": [],
      "source": [
        "# Rename entries in 'sample_name' column using the rename dictionary\n",
        "exp_df['sample_name'] = exp_df['sample_name'].replace(rename_dict)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Rename PTMs based on conditions\n",
        "exp_df['ptm'] = exp_df.apply(\n",
        "    lambda x: 'Hydroxylation' if x['aa'] == 'P' and x['ptm'] == 'Oxidation' else\n",
        "             'DeamidationN' if x['aa'] == 'N' and x['ptm'] == 'Deamidated' else\n",
        "             'DeamidationQ' if x['aa'] == 'Q' and x['ptm'] == 'Deamidated' else\n",
        "             x['ptm'],  # Leave other PTMs unchanged\n",
        "    axis=1\n",
        ")"
      ],
      "metadata": {
        "id": "ZSsfDYdRj6XB"
      },
      "execution_count": 62,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 64,
      "metadata": {
        "id": "pBCk62y8u_0Q",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1e39b5cf-bb4b-439b-b6bf-9fdc17ddd910"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Unique sample names: ['Cz16' 'Cz12' 'M2' 'Cz13' 'Cz3' 'Cz4' 'Cz17' 'Cz9' 'Cz14' 'Cz19' 'Cz22'\n",
            " 'P3' 'Cz1' 'Cz18' 'M1' 'Cz8' 'Cz21' 'Cz6' 'Cz15' 'Cz5' 'P4' 'M4' 'Cz7'\n",
            " 'M3' 'Cz10' 'Cz2' 'P2' 'Cz11' 'P1' 'Cz20' '09_ICB' 'MACHINE_BSA' '16_OL'\n",
            " '12_ICib' '10_ICB' 'Cz17_CFrag' '06_ICB' '03_OBO' 'Cz11_CFrag' '13_ICib'\n",
            " '14_OL' 'Cz5b_CFrag' '02_OBi' '05_ICB' 'INJECTION_BLANK']\n",
            "Counts of each sample name: 10_ICB             64262\n",
            "Cz6                53366\n",
            "Cz22               51508\n",
            "Cz9                50398\n",
            "06_ICB             49847\n",
            "Cz17               49504\n",
            "09_ICB             49199\n",
            "16_OL              48987\n",
            "Cz1                37640\n",
            "Cz14               37350\n",
            "Cz11_CFrag         37123\n",
            "14_OL              36535\n",
            "12_ICib            35896\n",
            "Cz5b_CFrag         35818\n",
            "Cz19               35307\n",
            "13_ICib            33950\n",
            "Cz11               29732\n",
            "Cz18               28926\n",
            "P3                 27535\n",
            "Cz17_CFrag         26356\n",
            "Cz5                21706\n",
            "M3                 21558\n",
            "P4                 20418\n",
            "Cz3                16434\n",
            "M2                 15745\n",
            "Cz8                14291\n",
            "Cz4                13874\n",
            "P2                 13741\n",
            "Cz16               13689\n",
            "MACHINE_BSA        12106\n",
            "Cz2                11061\n",
            "03_OBO             10879\n",
            "P1                 10726\n",
            "M1                 10413\n",
            "Cz15               10117\n",
            "Cz10                8921\n",
            "Cz12                7792\n",
            "M4                  6543\n",
            "Cz13                6534\n",
            "02_OBi              3308\n",
            "Cz7                 2528\n",
            "05_ICB              1523\n",
            "Cz21                1465\n",
            "Cz20                 340\n",
            "INJECTION_BLANK       60\n",
            "Name: sample_name, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "unique_names = exp_df['sample_name'].unique()\n",
        "name_counts = exp_df['sample_name'].value_counts()\n",
        "print(\"Unique sample names:\", unique_names)\n",
        "print(\"Counts of each sample name:\", name_counts)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the value counts of ptm entries\n",
        "ptm_counts = exp_df['ptm'].value_counts()\n",
        "\n",
        "# Print the ptm entries and their counts\n",
        "print(ptm_counts.head(20))\n",
        "\n",
        "# Calculate ptm_counts without aggregating sample_name\n",
        "ptm_counts = exp_df.groupby(['sample_name','GNC', 'position', 'aa', 'ptm']).size().reset_index(name='ptm_count')\n",
        "\n",
        "# Calculate total_counts without aggregating sample_name\n",
        "total_counts = exp_df.groupby(['sample_name','GNC', 'position', 'aa']).size().reset_index(name='total_count')\n",
        "\n",
        "# Merge ptm_counts with the original exp_df to preserve sample_name and score\n",
        "df_merged = pd.merge(exp_df, ptm_counts, on=['sample_name','GNC', 'position', 'aa', 'ptm'], how='left')\n",
        "df_merged = pd.merge(df_merged, total_counts, on=['sample_name','GNC', 'position', 'aa'], how='left')\n",
        "\n",
        "# Calculate the percentage of aa in each position having a specific ptm\n",
        "df_merged['ptm_percentage'] = (df_merged['ptm_count'] / df_merged['total_count']) * 100\n",
        "\n",
        "# Filter to include only rows where ptm is not None or empty (i.e., ptm is present)\n",
        "result_with_ptms = df_merged[df_merged['ptm'].notna() & (df_merged['ptm'] != '')]\n",
        "\n",
        "# Display the filtered result with ptm details included\n",
        "print(result_with_ptms[['GNC', 'sample_name', 'position', 'aa', 'ptm', 'ptm_percentage', 'sample_name', 'score']].head())\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "L6iZjbr9kOM_",
        "outputId": "a1c8b1e4-10ce-46ff-b7c5-de05bd18ebc3"
      },
      "execution_count": 65,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                     984914\n",
            "Hydroxylation         70272\n",
            "DeamidationN           5958\n",
            "Oxidation              5947\n",
            "DeamidationQ           5196\n",
            "Carbamidomethyl         349\n",
            "Dioxidation             229\n",
            "Trioxidation            220\n",
            "Methyl                  121\n",
            "Methylphosphonate        90\n",
            "Pro->pyro-Glu            89\n",
            "Carboxymethyl            81\n",
            "Ammonia-loss             79\n",
            "Formyl                   79\n",
            "Pro->Asn                 65\n",
            "Dehydrated               59\n",
            "Ethanedithiol            58\n",
            "Carboxy                  54\n",
            "Carbonyl                 51\n",
            "Asn->Met                 39\n",
            "Name: ptm, dtype: int64\n",
            "      GNC sample_name  position aa           ptm  ptm_percentage sample_name  \\\n",
            "7   TRYUP        Cz16        76  N  DeamidationN       18.627451        Cz16   \n",
            "16  TRYUP        Cz16        85  M     Oxidation       86.915888        Cz16   \n",
            "36  TRYUP        Cz16        85  M     Oxidation       86.915888        Cz16   \n",
            "47  TRYUP        Cz16        76  N  DeamidationN       18.627451        Cz16   \n",
            "56  TRYUP        Cz16        85  M     Oxidation       86.915888        Cz16   \n",
            "\n",
            "           score  \n",
            "7   1.528900e-10  \n",
            "16  1.528900e-10  \n",
            "36  2.943950e-10  \n",
            "47  5.677730e-10  \n",
            "56  5.677730e-10  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#COUNT GNC\n",
        "unique_names_GNC = exp_df['GNC'].unique()\n",
        "name_counts_GNC = exp_df['GNC'].value_counts()\n",
        "print(\"Unique GNC:\", unique_names_GNC)\n",
        "print(\"Counts of each GNC:\", name_counts_GNC)"
      ],
      "metadata": {
        "id": "jPtMbqxNdy3M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Explore PTMs"
      ],
      "metadata": {
        "id": "OsufALpCJSLe"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Reshaping exp_df"
      ],
      "metadata": {
        "id": "CP15HEokj2ik"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def create_grouped_df(exp_df):\n",
        "    \"\"\"\n",
        "    Create a grouped dataframe from the input dataframe `exp_df`.\n",
        "\n",
        "    Args:\n",
        "        exp_df (pandas.DataFrame): The input dataframe containing protein identification data.\n",
        "\n",
        "    Returns:\n",
        "        pandas.DataFrame: The grouped dataframe containing the requested analysis.\n",
        "    \"\"\"\n",
        "    # Group the data by GNC, sample_name, and position\n",
        "    grouped = exp_df.groupby(['GNC', 'sample_name', 'position'])\n",
        "\n",
        "    # Create a list to store the grouped analysis\n",
        "    grouped_analysis = []\n",
        "\n",
        "    # Iterate over each group\n",
        "    for group_keys, group_data in grouped:\n",
        "        # Unpack the group keys\n",
        "        gnc, sample_name, position = group_keys\n",
        "\n",
        "        # Get the top 3 amino acids and their counts\n",
        "        aa_counts = group_data['aa'].value_counts().head(3)\n",
        "        top_aas = aa_counts.index.tolist()\n",
        "        aa_counts = aa_counts.to_dict()\n",
        "\n",
        "        # Initialize a dictionary to store the analysis for this group\n",
        "        group_analysis = {\n",
        "            'GNC': gnc,\n",
        "            'sample_name': sample_name,\n",
        "            'position': position\n",
        "        }\n",
        "\n",
        "        # Analyze the top 3 amino acids\n",
        "        for i, aa in enumerate(top_aas, start=1):\n",
        "            aa_key = f'aa{i}'\n",
        "            group_analysis[aa_key] = aa  # Add the amino acid name\n",
        "            n_aa_key = f'n_{aa_key}'\n",
        "            score_mean_key = f'{aa_key}_score_mean'\n",
        "            score_max_key = f'{aa_key}_score_max'\n",
        "            score_sd_key = f'{aa_key}_score_SD'\n",
        "            score_keys = [f'score{j}_{aa_key}' for j in range(1, 4)]\n",
        "            spectra_keys = [f'spectraId{j}_{aa_key}' for j in range(1, 4)]\n",
        "\n",
        "            # Count of the amino acid\n",
        "            group_analysis[n_aa_key] = aa_counts[aa]\n",
        "\n",
        "            # Mean, max, and standard deviation of the confidence scores\n",
        "            aa_scores = group_data.loc[group_data['aa'] == aa, 'confidence']\n",
        "            group_analysis[score_mean_key] = aa_scores.mean()\n",
        "            group_analysis[score_max_key] = aa_scores.max()\n",
        "            group_analysis[score_sd_key] = aa_scores.std()\n",
        "\n",
        "            # Top 3 scores and spectra IDs\n",
        "            aa_scores = group_data.loc[group_data['aa'] == aa, ['score', 'spectraId']].nlargest(3, 'score')\n",
        "            for j, (score_key, spectra_key) in enumerate(zip(score_keys, spectra_keys)):\n",
        "                if j < len(aa_scores):\n",
        "                    group_analysis[score_key] = aa_scores.iloc[j]['score']\n",
        "                    group_analysis[spectra_key] = aa_scores.iloc[j]['spectraId']\n",
        "                else:\n",
        "                    group_analysis[score_key] = None\n",
        "                    group_analysis[spectra_key] = None\n",
        "\n",
        "            # Analyze the PTMs for this amino acid\n",
        "            aa_ptms = group_data.loc[group_data['aa'] == aa, 'ptm'].value_counts()\n",
        "            top_ptms = aa_ptms.index.tolist()[:2]\n",
        "            aa_ptms = aa_ptms.to_dict()\n",
        "\n",
        "            for j, ptm in enumerate(top_ptms, start=1):\n",
        "                ptm_key = f'ptm{j}_{aa_key}'\n",
        "                ptm_perc_key = f'%{ptm_key}'\n",
        "                ptm_score_mean_key = f'{ptm_key}_score_mean'\n",
        "                ptm_score_max_key = f'{ptm_key}_score_max'\n",
        "                ptm_score_sd_key = f'{ptm_key}_score_SD'\n",
        "                ptm_score_keys = [f'score{k}_{ptm_key}' for k in range(1, 4)]\n",
        "                ptm_spectra_keys = [f'spectraId{k}_{ptm_key}' for k in range(1, 4)]\n",
        "\n",
        "\n",
        "                # PTM name and percentage\n",
        "                group_analysis[ptm_key] = ptm\n",
        "                group_analysis[ptm_perc_key] = aa_ptms.get(ptm, 0) / (group_data[(group_data['aa'] == aa) & (group_data['position'] == position)]['aa'].count()) * 100\n",
        "\n",
        "\n",
        "                # Mean, max, and standard deviation of the confidence scores for the PTM\n",
        "                ptm_scores = group_data.loc[(group_data['aa'] == aa) & (group_data['ptm'] == ptm), 'confidence']\n",
        "                group_analysis[ptm_score_mean_key] = ptm_scores.mean()\n",
        "                group_analysis[ptm_score_max_key] = ptm_scores.max()\n",
        "                group_analysis[ptm_score_sd_key] = ptm_scores.std()\n",
        "\n",
        "                # Top 3 scores and spectra IDs for the PTM\n",
        "                ptm_scores = group_data.loc[(group_data['aa'] == aa) & (group_data['ptm'] == ptm), ['score', 'spectraId']].nlargest(3, 'score')\n",
        "                for k, (score_key, spectra_key) in enumerate(zip(ptm_score_keys, ptm_spectra_keys)):\n",
        "                    if k < len(ptm_scores):\n",
        "                        group_analysis[score_key] = ptm_scores.iloc[k]['score']\n",
        "                        group_analysis[spectra_key] = ptm_scores.iloc[k]['spectraId']\n",
        "                    else:\n",
        "                        group_analysis[score_key] = None\n",
        "                        group_analysis[spectra_key] = None\n",
        "\n",
        "        # Append the analysis for this group to the list\n",
        "        grouped_analysis.append(group_analysis)\n",
        "\n",
        "    # Create the grouped dataframe from the list of dictionaries\n",
        "    grouped_df = pd.DataFrame(grouped_analysis)\n",
        "\n",
        "    return grouped_df"
      ],
      "metadata": {
        "id": "nxsDp5wo9cSH"
      },
      "execution_count": 71,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#-----------------------the SLOW one-----------------\n",
        "#\n",
        "group_df = create_grouped_df(exp_df)"
      ],
      "metadata": {
        "id": "m7vdjlIA27rN"
      },
      "execution_count": 72,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# with list constructor\n",
        "col_list = list(group_df.columns)\n",
        "# with tolist method\n",
        "col_list = group_df.columns.tolist()\n",
        "print(col_list)"
      ],
      "metadata": {
        "id": "u0uMkNlPBxwh",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1b2ceec6-1f5f-46ec-9a57-34ea68292d64"
      },
      "execution_count": 73,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['GNC', 'sample_name', 'position', 'aa1', 'n_aa1', 'aa1_score_mean', 'aa1_score_max', 'aa1_score_SD', 'score1_aa1', 'spectraId1_aa1', 'score2_aa1', 'spectraId2_aa1', 'score3_aa1', 'spectraId3_aa1', 'ptm1_aa1', '%ptm1_aa1', 'ptm1_aa1_score_mean', 'ptm1_aa1_score_max', 'ptm1_aa1_score_SD', 'score1_ptm1_aa1', 'spectraId1_ptm1_aa1', 'score2_ptm1_aa1', 'spectraId2_ptm1_aa1', 'score3_ptm1_aa1', 'spectraId3_ptm1_aa1', 'ptm2_aa1', '%ptm2_aa1', 'ptm2_aa1_score_mean', 'ptm2_aa1_score_max', 'ptm2_aa1_score_SD', 'score1_ptm2_aa1', 'spectraId1_ptm2_aa1', 'score2_ptm2_aa1', 'spectraId2_ptm2_aa1', 'score3_ptm2_aa1', 'spectraId3_ptm2_aa1', 'aa2', 'n_aa2', 'aa2_score_mean', 'aa2_score_max', 'aa2_score_SD', 'score1_aa2', 'spectraId1_aa2', 'score2_aa2', 'spectraId2_aa2', 'score3_aa2', 'spectraId3_aa2', 'ptm1_aa2', '%ptm1_aa2', 'ptm1_aa2_score_mean', 'ptm1_aa2_score_max', 'ptm1_aa2_score_SD', 'score1_ptm1_aa2', 'spectraId1_ptm1_aa2', 'score2_ptm1_aa2', 'spectraId2_ptm1_aa2', 'score3_ptm1_aa2', 'spectraId3_ptm1_aa2', 'aa3', 'n_aa3', 'aa3_score_mean', 'aa3_score_max', 'aa3_score_SD', 'score1_aa3', 'spectraId1_aa3', 'score2_aa3', 'spectraId2_aa3', 'score3_aa3', 'spectraId3_aa3', 'ptm1_aa3', '%ptm1_aa3', 'ptm1_aa3_score_mean', 'ptm1_aa3_score_max', 'ptm1_aa3_score_SD', 'score1_ptm1_aa3', 'spectraId1_ptm1_aa3', 'score2_ptm1_aa3', 'spectraId2_ptm1_aa3', 'score3_ptm1_aa3', 'spectraId3_ptm1_aa3', 'ptm2_aa3', '%ptm2_aa3', 'ptm2_aa3_score_mean', 'ptm2_aa3_score_max', 'ptm2_aa3_score_SD', 'score1_ptm2_aa3', 'spectraId1_ptm2_aa3', 'score2_ptm2_aa3', 'spectraId2_ptm2_aa3', 'score3_ptm2_aa3', 'spectraId3_ptm2_aa3', 'ptm2_aa2', '%ptm2_aa2', 'ptm2_aa2_score_mean', 'ptm2_aa2_score_max', 'ptm2_aa2_score_SD', 'score1_ptm2_aa2', 'spectraId1_ptm2_aa2', 'score2_ptm2_aa2', 'spectraId2_ptm2_aa2', 'score3_ptm2_aa2', 'spectraId3_ptm2_aa2']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "group_df"
      ],
      "metadata": {
        "id": "Sl64tUYYBIDM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(group_df)"
      ],
      "metadata": {
        "id": "WEK0GVQx3FzN",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "daf73e3e-9f24-4fa6-ec97-be6243fd6606"
      },
      "execution_count": 74,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "         GNC sample_name  position aa1  n_aa1  aa1_score_mean  aa1_score_max  \\\n",
            "0        ALB      06_ICB        65   L      7             NaN            NaN   \n",
            "1        ALB      06_ICB        66   V      7             NaN            NaN   \n",
            "2        ALB      06_ICB        67   N      7             NaN            NaN   \n",
            "3        ALB      06_ICB        68   E      7             NaN            NaN   \n",
            "4        ALB      06_ICB        69   L      7             NaN            NaN   \n",
            "...      ...         ...       ...  ..    ...             ...            ...   \n",
            "80535  TRYUP          P4       102   V     22             NaN            NaN   \n",
            "80536  TRYUP          P4       103   S     22             NaN            NaN   \n",
            "80537  TRYUP          P4       104   I     21             NaN            NaN   \n",
            "80538  TRYUP          P4       105   P     22             NaN            NaN   \n",
            "80539  TRYUP          P4       106   R     22             NaN            NaN   \n",
            "\n",
            "       aa1_score_SD  score1_aa1  spectraId1_aa1  ...  %ptm2_aa2  \\\n",
            "0               NaN    0.095769          3816.0  ...        NaN   \n",
            "1               NaN    0.095769          3816.0  ...        NaN   \n",
            "2               NaN    0.095769          3816.0  ...        NaN   \n",
            "3               NaN    0.095769          3816.0  ...        NaN   \n",
            "4               NaN    0.095769          3816.0  ...        NaN   \n",
            "...             ...         ...             ...  ...        ...   \n",
            "80535           NaN    0.238656          2839.0  ...        NaN   \n",
            "80536           NaN    0.238656          2839.0  ...        NaN   \n",
            "80537           NaN    0.238656          2839.0  ...        NaN   \n",
            "80538           NaN    0.238656          2839.0  ...        NaN   \n",
            "80539           NaN    0.238656          2839.0  ...        NaN   \n",
            "\n",
            "       ptm2_aa2_score_mean  ptm2_aa2_score_max  ptm2_aa2_score_SD  \\\n",
            "0                      NaN                 NaN                NaN   \n",
            "1                      NaN                 NaN                NaN   \n",
            "2                      NaN                 NaN                NaN   \n",
            "3                      NaN                 NaN                NaN   \n",
            "4                      NaN                 NaN                NaN   \n",
            "...                    ...                 ...                ...   \n",
            "80535                  NaN                 NaN                NaN   \n",
            "80536                  NaN                 NaN                NaN   \n",
            "80537                  NaN                 NaN                NaN   \n",
            "80538                  NaN                 NaN                NaN   \n",
            "80539                  NaN                 NaN                NaN   \n",
            "\n",
            "      score1_ptm2_aa2  spectraId1_ptm2_aa2  score2_ptm2_aa2  \\\n",
            "0                 NaN                  NaN              NaN   \n",
            "1                 NaN                  NaN              NaN   \n",
            "2                 NaN                  NaN              NaN   \n",
            "3                 NaN                  NaN              NaN   \n",
            "4                 NaN                  NaN              NaN   \n",
            "...               ...                  ...              ...   \n",
            "80535             NaN                  NaN              NaN   \n",
            "80536             NaN                  NaN              NaN   \n",
            "80537             NaN                  NaN              NaN   \n",
            "80538             NaN                  NaN              NaN   \n",
            "80539             NaN                  NaN              NaN   \n",
            "\n",
            "       spectraId2_ptm2_aa2  score3_ptm2_aa2  spectraId3_ptm2_aa2  \n",
            "0                      NaN              NaN                  NaN  \n",
            "1                      NaN              NaN                  NaN  \n",
            "2                      NaN              NaN                  NaN  \n",
            "3                      NaN              NaN                  NaN  \n",
            "4                      NaN              NaN                  NaN  \n",
            "...                    ...              ...                  ...  \n",
            "80535                  NaN              NaN                  NaN  \n",
            "80536                  NaN              NaN                  NaN  \n",
            "80537                  NaN              NaN                  NaN  \n",
            "80538                  NaN              NaN                  NaN  \n",
            "80539                  NaN              NaN                  NaN  \n",
            "\n",
            "[80540 rows x 102 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "PFIND_FOLDER1 = f'pFind_{STUDY_NAME1}'  # Folder path for study 1\n",
        "directory_path = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/{STUDY_NAME1}/{PFIND_FOLDER1}/result/'\n",
        "\n",
        "# Define the directory for saving the output file\n",
        "#directory_path = f'/content/drive/MyDrive/Colab_Notebooks/NovorCloud/sample1/'\n",
        "os.makedirs(directory_path , exist_ok=True)\n",
        "group_df.to_csv(f'{directory_path}pFind_grouped2.csv', index=False)"
      ],
      "metadata": {
        "id": "AD7ItgteAcOh"
      },
      "execution_count": 68,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8pJQpogBKkeY"
      },
      "outputs": [],
      "source": [
        "#Writing exp.csv\n",
        "# Construct the file path first\n",
        "csv_file_path = f'{BASE_PATH}/{STUDY_NAME}exp.csv'\n",
        "\n",
        "# Export the DataFrame to CSV using the correct method name\n",
        "exp_df.to_csv(csv_file_path)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#grouped_CGPT_df = analyze_protein_data(test_df)"
      ],
      "metadata": {
        "id": "-buZytO-4KpY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#print(grouped_CGPT_df)"
      ],
      "metadata": {
        "id": "YqBgRtU35u3Q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#----------------------------------Summary Table of Data\n",
        "\n",
        "# Calculate counts for 'GNC&spp' across the entire DataFrame\n",
        "counts = group_df['GNC'].value_counts()\n",
        "\n",
        "# Filter 'GNC&spp' entries that meet the threshold criterion (e.g., count >= 100)\n",
        "filtered_entries = counts[counts >= 100].index.tolist()\n",
        "\n",
        "# Filter the DataFrame to only include those 'GNC&spp' entries\n",
        "filtered_df = group_df[group_df['GNC'].isin(filtered_entries)]\n",
        "\n",
        "# Prepare data for PrettyTable\n",
        "# Initialize PrettyTable with 'File_Name' plus each 'GNC&spp' that meets the threshold as columns\n",
        "columns = ['sample_name'] + filtered_entries\n",
        "pt = PrettyTable()\n",
        "pt.field_names = columns\n",
        "\n",
        "# Populate PrettyTable rows with counts of 'GNC&spp' for each 'File_Name'\n",
        "for file_name in filtered_df['sample_name'].unique():\n",
        "    row_data = [file_name]\n",
        "    for entry in filtered_entries:\n",
        "        count = filtered_df[(filtered_df['sample_name'] == file_name) & (filtered_df['GNC'] == entry)].shape[0]\n",
        "        row_data.append(count)\n",
        "    pt.add_row(row_data)\n",
        "\n",
        "# Set table title\n",
        "pt.title = 'Summary Output'\n",
        "\n",
        "# Print the PrettyTable\n",
        "print(pt)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zkyuwwSaQ_Gl",
        "outputId": "ac574ebd-7d61-4c07-b36f-a6e68112ad49"
      },
      "execution_count": 69,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "+-----------------------------------------------------------------------------------------------------------------------------------------+\n",
            "|                                                              Summary Output                                                             |\n",
            "+-----------------+--------+--------+--------+------+-----+------+--------+------+-------+-----+---------+-----+--------+--------+--------+\n",
            "|   sample_name   | COL1A1 | COL1A2 | COL3A1 | K2C1 | KS9 | KS10 | COLunk | K22E | TRYUP | PIV | COL12A1 | ALB | COL6A2 | FibMod | COL5A1 |\n",
            "+-----------------+--------+--------+--------+------+-----+------+--------+------+-------+-----+---------+-----+--------+--------+--------+\n",
            "|      06_ICB     |  638   |  620   |  152   |  66  |  87 |  18  |  946   |  12  |   40  |  0  |   441   | 100 |  106   |  111   |   0    |\n",
            "|      09_ICB     |  671   |  645   |   98   |  52  | 152 |  27  |  698   |  35  |   40  |  0  |   560   | 101 |   94   |  131   |   12   |\n",
            "|      10_ICB     |  842   |  807   |  510   | 134  |  93 |  60  |  291   |  24  |   50  |  0  |    25   |  47 |   44   |   0    |   12   |\n",
            "|     12_ICib     |  899   |  729   |  581   |  97  | 173 | 109  |   0    |  96  |   50  |  0  |    0    |  10 |   10   |   0    |   0    |\n",
            "|     13_ICib     |  733   |  713   |  460   | 120  | 157 | 124  |   0    |  47  |   50  |  0  |    0    |  19 |   10   |   0    |   0    |\n",
            "|      14_OL      |  807   |  820   |  209   | 426  | 498 | 346  |  439   | 275  |   76  |  0  |   218   |  57 |  101   |   48   |   12   |\n",
            "|      16_OL      |  815   |  783   |  507   | 199  | 174 | 233  |   0    | 229  |   58  |  0  |    0    |  26 |   0    |   0    |   11   |\n",
            "|       Cz14      |  654   |  601   |  331   |  40  |  0  |  41  |   9    |  43  |   58  |  0  |    0    |  13 |   21   |   0    |   12   |\n",
            "|       Cz18      |  680   |  593   |  344   | 122  | 254 | 204  |   0    | 133  |   50  |  0  |    0    |  10 |   0    |   0    |   0    |\n",
            "|       Cz22      |  755   |  721   |  126   |  24  |  20 |  19  |  352   |  24  |   40  |  0  |   208   |  34 |   45   |   0    |   12   |\n",
            "|    Cz5b_CFrag   |  699   |  660   |  282   |  12  |  0  |  0   |  263   |  0   |   0   |  0  |   188   |  10 |   30   |   30   |   12   |\n",
            "|        M2       |  155   |  186   |   0    |  85  |  98 | 121  |   0    |  31  |   58  | 361 |    0    | 106 |   0    |   0    |   0    |\n",
            "|        M4       |  232   |  356   |   84   |  63  |  7  |  82  |  131   |  20  |   58  | 238 |    0    |  40 |   17   |   0    |   0    |\n",
            "|   MACHINE_BSA   |  446   |  416   |   42   |  0   |  0  |  0   |   0    |  0   |   58  |  0  |    0    | 375 |   0    |   0    |   0    |\n",
            "|    Cz11_CFrag   |  697   |  658   |  426   |  12  |  0  |  0   |   78   |  0   |   0   |  0  |    82   |  0  |   19   |   17   |   12   |\n",
            "|      03_OBO     |  172   |  199   |   0    | 233  | 129 | 220  |   0    | 157  |   70  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|      05_ICB     |   16   |   47   |   0    |  78  |  47 |  66  |   0    |  40  |   64  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz1       |  713   |  598   |  328   | 167  | 202 | 188  |   9    | 102  |   58  |  0  |    0    |  0  |   20   |   0    |   12   |\n",
            "|       Cz10      |  454   |  479   |  151   |  43  |  25 |  31  |   0    |  35  |   72  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz11      |  685   |  576   |  269   | 165  | 124 | 120  |   0    |  61  |   58  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz12      |   45   |   90   |   0    | 146  | 178 | 169  |   0    |  91  |   72  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz13      |  528   |  407   |  152   |  12  |  0  |  25  |   0    |  35  |   58  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz15      |  284   |  327   |  106   |  59  | 149 |  99  |   0    |  62  |   50  |  0  |    0    |  0  |   10   |   0    |   0    |\n",
            "|       Cz16      |  215   |  258   |   9    | 234  | 238 | 282  |   0    | 116  |   91  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz17      |  751   |  724   |  350   | 248  | 339 | 259  |   9    | 202  |   58  |  0  |    0    |  0  |   10   |   0    |   22   |\n",
            "|    Cz17_CFrag   |  724   |  598   |  339   |  35  |  11 |  0   |   11   |  12  |   40  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz19      |  641   |  569   |  293   | 285  | 256 | 340  |   0    | 308  |   73  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz2       |  516   |  421   |  224   |  36  |  9  |  28  |   0    |  0   |   58  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz3       |  547   |  401   |  233   |  64  |  36 |  25  |   0    |  12  |   58  |  0  |    0    |  0  |   0    |   0    |   9    |\n",
            "|       Cz4       |  309   |  326   |   78   | 193  | 101 | 146  |   0    |  82  |   72  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz5       |  676   |  560   |  344   |  0   |  0  |  24  |   0    |  19  |   40  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz6       |  736   |  685   |  383   | 141  |  87 | 122  |   0    |  0   |   58  |  0  |    0    |  0  |   0    |   0    |   23   |\n",
            "|       Cz7       |  122   |  110   |   0    |  0   |  25 |  0   |   0    |  0   |   58  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz8       |  242   |  266   |   36   |  12  |  27 |  9   |   0    |  0   |   61  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz9       |  775   |  727   |  378   | 134  |  48 | 181  |   0    |  37  |   58  |  0  |    0    |  0  |   0    |   0    |   22   |\n",
            "|        M1       |  194   |  206   |   20   | 207  | 280 | 156  |   22   |  86  |   76  | 327 |    0    |  0  |   0    |   0    |   0    |\n",
            "|        M3       |  729   |  562   |  330   |  71  |  0  |  27  |   9    |  0   |   40  | 197 |    0    |  0  |   0    |   0    |   0    |\n",
            "|        P1       |  396   |  392   |  159   | 126  |  39 |  58  |   30   |  71  |   58  | 276 |    0    |  0  |   0    |   0    |   0    |\n",
            "|        P2       |  381   |  377   |  110   | 281  | 226 | 246  |   9    | 144  |   58  | 304 |    0    |  0  |   0    |   0    |   0    |\n",
            "|        P3       |  592   |  544   |  237   | 308  | 334 | 269  |   0    | 217  |   58  | 265 |    0    |  0  |   0    |   0    |   0    |\n",
            "|        P4       |  441   |  413   |   58   | 161  | 204 | 153  |   0    |  69  |   58  | 308 |    0    |  0  |   0    |   0    |   0    |\n",
            "|      02_OBi     |   0    |   0    |   0    | 252  | 113 | 213  |   0    | 209  |   94  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz21      |   0    |   0    |   0    |  24  |  23 |  29  |   0    |  14  |   58  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "|       Cz20      |   0    |   0    |   0    |  0   |  0  |  0   |   0    |  0   |   40  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "| INJECTION_BLANK |   0    |   0    |   0    |  0   |  0  |  0   |   0    |  0   |   20  |  0  |    0    |  0  |   0    |   0    |   0    |\n",
            "+-----------------+--------+--------+--------+------+-----+------+--------+------+-------+-----+---------+-----+--------+--------+--------+\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# GNC to examine (replace with the GNC you want to analyze)\n",
        "gnc_of_interest = 'COL3A1'\n",
        "\n",
        "# Filter the DataFrame to only include the GNC of interest\n",
        "gnc_df = group_df[group_df['GNC'] == gnc_of_interest]\n",
        "\n",
        "# Print the GNC DataFrame\n",
        "print(gnc_df)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SKbPVe67J2cJ",
        "outputId": "f4c67578-7008-49ff-e5be-b47b399ff295"
      },
      "execution_count": 76,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "          GNC sample_name  position aa1  n_aa1  aa1_score_mean  aa1_score_max  \\\n",
            "44447  COL3A1      06_ICB       224   G     46             NaN            NaN   \n",
            "44448  COL3A1      06_ICB       225   E     46             NaN            NaN   \n",
            "44449  COL3A1      06_ICB       226   M     46             NaN            NaN   \n",
            "44450  COL3A1      06_ICB       227   G     46             NaN            NaN   \n",
            "44451  COL3A1      06_ICB       228   P     46             NaN            NaN   \n",
            "...       ...         ...       ...  ..    ...             ...            ...   \n",
            "53181  COL3A1          P4       793   M      9             NaN            NaN   \n",
            "53182  COL3A1          P4       794   P      9             NaN            NaN   \n",
            "53183  COL3A1          P4       795   G      9             NaN            NaN   \n",
            "53184  COL3A1          P4       796   S      6             NaN            NaN   \n",
            "53185  COL3A1          P4       797   R      9             NaN            NaN   \n",
            "\n",
            "       aa1_score_SD  score1_aa1  spectraId1_aa1  ...  %ptm2_aa2  \\\n",
            "44447           NaN    0.216692          3763.0  ...        NaN   \n",
            "44448           NaN    0.216692          3763.0  ...        NaN   \n",
            "44449           NaN    0.216692          3763.0  ...        NaN   \n",
            "44450           NaN    0.216692          3763.0  ...        NaN   \n",
            "44451           NaN    0.216692          3763.0  ...        NaN   \n",
            "...             ...         ...             ...  ...        ...   \n",
            "53181           NaN    0.280936          1232.0  ...        NaN   \n",
            "53182           NaN    0.280936          1232.0  ...        NaN   \n",
            "53183           NaN    0.280936          1232.0  ...        NaN   \n",
            "53184           NaN    0.280936          1232.0  ...        NaN   \n",
            "53185           NaN    0.280936          1232.0  ...        NaN   \n",
            "\n",
            "       ptm2_aa2_score_mean  ptm2_aa2_score_max  ptm2_aa2_score_SD  \\\n",
            "44447                  NaN                 NaN                NaN   \n",
            "44448                  NaN                 NaN                NaN   \n",
            "44449                  NaN                 NaN                NaN   \n",
            "44450                  NaN                 NaN                NaN   \n",
            "44451                  NaN                 NaN                NaN   \n",
            "...                    ...                 ...                ...   \n",
            "53181                  NaN                 NaN                NaN   \n",
            "53182                  NaN                 NaN                NaN   \n",
            "53183                  NaN                 NaN                NaN   \n",
            "53184                  NaN                 NaN                NaN   \n",
            "53185                  NaN                 NaN                NaN   \n",
            "\n",
            "      score1_ptm2_aa2  spectraId1_ptm2_aa2  score2_ptm2_aa2  \\\n",
            "44447             NaN                  NaN              NaN   \n",
            "44448             NaN                  NaN              NaN   \n",
            "44449             NaN                  NaN              NaN   \n",
            "44450             NaN                  NaN              NaN   \n",
            "44451             NaN                  NaN              NaN   \n",
            "...               ...                  ...              ...   \n",
            "53181             NaN                  NaN              NaN   \n",
            "53182             NaN                  NaN              NaN   \n",
            "53183             NaN                  NaN              NaN   \n",
            "53184             NaN                  NaN              NaN   \n",
            "53185             NaN                  NaN              NaN   \n",
            "\n",
            "       spectraId2_ptm2_aa2  score3_ptm2_aa2  spectraId3_ptm2_aa2  \n",
            "44447                  NaN              NaN                  NaN  \n",
            "44448                  NaN              NaN                  NaN  \n",
            "44449                  NaN              NaN                  NaN  \n",
            "44450                  NaN              NaN                  NaN  \n",
            "44451                  NaN              NaN                  NaN  \n",
            "...                    ...              ...                  ...  \n",
            "53181                  NaN              NaN                  NaN  \n",
            "53182                  NaN              NaN                  NaN  \n",
            "53183                  NaN              NaN                  NaN  \n",
            "53184                  NaN              NaN                  NaN  \n",
            "53185                  NaN              NaN                  NaN  \n",
            "\n",
            "[8739 rows x 102 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "column_headers = gnc_df.columns.tolist()\n",
        "print(column_headers)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5yrdNsrAtY7X",
        "outputId": "4bf29b28-1d3e-4f81-b886-9c8e895fa460"
      },
      "execution_count": 97,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['GNC', 'sample_name', 'position', 'aa1', 'n_aa1', 'aa1_score_mean', 'aa1_score_max', 'aa1_score_SD', 'score1_aa1', 'spectraId1_aa1', 'score2_aa1', 'spectraId2_aa1', 'score3_aa1', 'spectraId3_aa1', 'ptm1_aa1', '%ptm1_aa1', 'ptm1_aa1_score_mean', 'ptm1_aa1_score_max', 'ptm1_aa1_score_SD', 'score1_ptm1_aa1', 'spectraId1_ptm1_aa1', 'score2_ptm1_aa1', 'spectraId2_ptm1_aa1', 'score3_ptm1_aa1', 'spectraId3_ptm1_aa1', 'ptm2_aa1', '%ptm2_aa1', 'ptm2_aa1_score_mean', 'ptm2_aa1_score_max', 'ptm2_aa1_score_SD', 'score1_ptm2_aa1', 'spectraId1_ptm2_aa1', 'score2_ptm2_aa1', 'spectraId2_ptm2_aa1', 'score3_ptm2_aa1', 'spectraId3_ptm2_aa1', 'aa2', 'n_aa2', 'aa2_score_mean', 'aa2_score_max', 'aa2_score_SD', 'score1_aa2', 'spectraId1_aa2', 'score2_aa2', 'spectraId2_aa2', 'score3_aa2', 'spectraId3_aa2', 'ptm1_aa2', '%ptm1_aa2', 'ptm1_aa2_score_mean', 'ptm1_aa2_score_max', 'ptm1_aa2_score_SD', 'score1_ptm1_aa2', 'spectraId1_ptm1_aa2', 'score2_ptm1_aa2', 'spectraId2_ptm1_aa2', 'score3_ptm1_aa2', 'spectraId3_ptm1_aa2', 'aa3', 'n_aa3', 'aa3_score_mean', 'aa3_score_max', 'aa3_score_SD', 'score1_aa3', 'spectraId1_aa3', 'score2_aa3', 'spectraId2_aa3', 'score3_aa3', 'spectraId3_aa3', 'ptm1_aa3', '%ptm1_aa3', 'ptm1_aa3_score_mean', 'ptm1_aa3_score_max', 'ptm1_aa3_score_SD', 'score1_ptm1_aa3', 'spectraId1_ptm1_aa3', 'score2_ptm1_aa3', 'spectraId2_ptm1_aa3', 'score3_ptm1_aa3', 'spectraId3_ptm1_aa3', 'ptm2_aa3', '%ptm2_aa3', 'ptm2_aa3_score_mean', 'ptm2_aa3_score_max', 'ptm2_aa3_score_SD', 'score1_ptm2_aa3', 'spectraId1_ptm2_aa3', 'score2_ptm2_aa3', 'spectraId2_ptm2_aa3', 'score3_ptm2_aa3', 'spectraId3_ptm2_aa3', 'ptm2_aa2', '%ptm2_aa2', 'ptm2_aa2_score_mean', 'ptm2_aa2_score_max', 'ptm2_aa2_score_SD', 'score1_ptm2_aa2', 'spectraId1_ptm2_aa2', 'score2_ptm2_aa2', 'spectraId2_ptm2_aa2', 'score3_ptm2_aa2', 'spectraId3_ptm2_aa2']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def analyze_protein_variations(df, gnc):\n",
        "    # Filter for the specific GNC\n",
        "    df = df[df['GNC'] == gnc]\n",
        "\n",
        "    # Initialize result containers\n",
        "    missing_aa_positions = []\n",
        "    single_aa_positions = []\n",
        "    multiple_aa_positions = []\n",
        "\n",
        "    # Group by 'position' to analyze each position\n",
        "    grouped = df.groupby('position')\n",
        "\n",
        "    for position, group in grouped:\n",
        "        # Identify missing amino acids\n",
        "        if group[['aa1', 'aa2', 'aa3']].isna().all().all():\n",
        "            missing_aa_positions.append(position)\n",
        "            continue\n",
        "\n",
        "        # Identify conserved amino acids (single AA across all samples)\n",
        "        unique_aas = pd.concat([group['aa1'], group['aa2'], group['aa3']]).dropna().unique()\n",
        "        if len(unique_aas) == 1:\n",
        "            aa_count = group['n_aa1'].sum()  # Assuming 'n_aa1' counts occurrences of 'aa1'\n",
        "            single_aa_positions.append((position, unique_aas[0], aa_count))\n",
        "            continue\n",
        "\n",
        "        # Analyze positions with multiple amino acids\n",
        "        for aa_col in ['aa1', 'aa2', 'aa3']:\n",
        "            for aa in group[aa_col].dropna().unique():\n",
        "                aa_count = group[group[aa_col] == aa]['n_aa1'].sum()  # Assuming 'n_aa1' corresponds to 'aa1'\n",
        "                top_score = group[group[aa_col] == aa][f'{aa_col}_score_max'].min()\n",
        "                multiple_aa_positions.append((position, aa, aa_count, top_score))\n",
        "\n",
        "    # Convert to DataFrames for easier handling\n",
        "    single_aa_df = pd.DataFrame(single_aa_positions, columns=['Position', 'Amino Acid', 'Count'])\n",
        "    multiple_aa_df = pd.DataFrame(multiple_aa_positions, columns=['Position', 'Amino Acid', 'Count', 'Top Score'])\n",
        "\n",
        "    return missing_aa_positions, single_aa_df, multiple_aa_df\n",
        "\n",
        "# Assuming 'df' is your DataFrame and 'COL1A2' is the GNC of interest\n",
        "missing_aa_positions, single_aa_df, multiple_aa_df = analyze_protein_variations(gnc_df, 'COL3A1')\n",
        "\n",
        "# Printing tables\n",
        "# Missing Amino Acid Positions\n",
        "print(\"Missing Amino Acid Positions:\")\n",
        "if missing_aa_positions:\n",
        "    table = PrettyTable(['Position'])\n",
        "    for position in missing_aa_positions:\n",
        "        table.add_row([position])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No missing amino acid positions found.\")\n",
        "\n",
        "# Single Amino Acid Positions\n",
        "print(\"\\nSingle Amino Acid Positions:\")\n",
        "if not single_aa_df.empty:\n",
        "    table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "    for row in single_aa_df.itertuples(index=False):\n",
        "        table.add_row([row.position, row.aa1, row.countt])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No positions with single amino acids found.\")\n",
        "\n",
        "\n",
        "# # Print single amino acid positions\n",
        "# print(\"\\nSingle Amino Acid Positions:\")\n",
        "# if not protein_variations['Single AA Positions'].empty:\n",
        "#     table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "#     for row in protein_variations['Sin# Single Amino Acid Positions\n",
        "print(\"\\nSingle Amino Acid Positions:\")\n",
        "if not single_aa_df.empty:\n",
        "    table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "    for row in single_aa_df.itertuples(index=False):\n",
        "        table.add_row([row.position, row.aa1, row.countt])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No positions with single amino acids found.\")\n",
        "gle AA Positions'].itertuples():\n",
        "#         table.add_row([row.position, row.aa1, row.count])\n",
        "#     print(table)\n",
        "# else:\n",
        "#     print(\"No positions with single amino acids found.\")\n",
        "\n",
        "# Multiple Amino Acid Positions\n",
        "print(\"\\nMultiple Amino Acid Positions:\")\n",
        "if not multiple_aa_df.empty:\n",
        "    table = PrettyTable(['Position', 'Amino Acid', 'Count', 'Top Score'])\n",
        "    for row in multiple_aa_df.itertuples(index=False):\n",
        "        table.add_row([row.Position, row.Amino_Acid, row.Count, row.Top_Score])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No positions with multiple amino acids found.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 280
        },
        "id": "-kEa70U14wZD",
        "outputId": "a81e6125-3b87-431d-ed61-2e10c52e5995"
      },
      "execution_count": 114,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Missing Amino Acid Positions:\n",
            "No missing amino acid positions found.\n",
            "\n",
            "Single Amino Acid Positions:\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'Pandas' object has no attribute 'position'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-114-cf881b24a65a>\u001b[0m in \u001b[0;36m<cell line: 58>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPrettyTable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Position'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Amino Acid'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Count'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msingle_aa_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitertuples\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m         \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_row\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maa1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcountt\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'Pandas' object has no attribute 'position'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from prettytable import PrettyTable\n",
        "\n",
        "def analyze_protein_variations(gnc_df):\n",
        "    # Ensure 'position' is not set as an index to avoid ambiguity\n",
        "    gnc_df.reset_index(drop=True, inplace=True)\n",
        "\n",
        "    # Group data by position\n",
        "    grouped_by_position = gnc_df.groupby('position')\n",
        "\n",
        "    # Analyze positions with missing amino acids\n",
        "    # Filtering directly, so no ambiguity should arise here\n",
        "    missing_aa_positions = gnc_df[gnc_df['aa1'].isna()].groupby('position').size().index.tolist()\n",
        "\n",
        "    # Analyze positions with a single amino acid\n",
        "    # This approach avoids ambiguity by working with grouped objects directly\n",
        "    single_aa_summary = gnc_df.dropna(subset=['aa1']).groupby('position').filter(lambda x: x['aa1'].nunique() == 1)\n",
        "    single_aa_summary = single_aa_summary.groupby(['position', 'aa1']).size().reset_index(name='count')\n",
        "\n",
        "    # Analyze positions with multiple amino acids\n",
        "    multiple_aa_positions = gnc_df.dropna(subset=['aa1']).groupby('position').filter(lambda x: x['aa1'].nunique() > 1)\n",
        "    multiple_aa_analysis = multiple_aa_positions.groupby(['position', 'aa1']).size().reset_index(name='count')\n",
        "\n",
        "    # Combine results\n",
        "    summary = {\n",
        "        'Missing AA Positions': missing_aa_positions,\n",
        "        'Single AA Positions': single_aa_summary,\n",
        "        'Multiple AA Positions': multiple_aa_analysis\n",
        "    }\n",
        "\n",
        "    return summary\n",
        "\n",
        "\n",
        "# Assuming 'gnc_df' is your DataFrame filtered for a specific GNC of interest\n",
        "protein_variations = analyze_protein_variations(gnc_df)\n",
        "\n",
        "# Print missing amino acid positions\n",
        "print(\"Missing Amino Acid Positions:\")\n",
        "if protein_variations['Missing AA Positions']:\n",
        "    table = PrettyTable(['Position'])\n",
        "    table.add_rows([[position] for position in protein_variations['Missing AA Positions']])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No missing amino acid positions found.\")\n",
        "\n",
        "# Print single amino acid positions\n",
        "print(\"\\nSingle Amino Acid Positions:\")\n",
        "if not protein_variations['Single AA Positions'].empty:\n",
        "    table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "    for row in protein_variations['Single AA Positions'].itertuples():\n",
        "        table.add_row([row.position, row.aa1, row.count])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No positions with single amino acids found.\")\n",
        "\n",
        "# Print multiple amino acid positions\n",
        "print(\"\\nMultiple Amino Acid Positions:\")\n",
        "if not protein_variations['Multiple AA Positions'].empty:\n",
        "    table = PrettyTable(['Position', 'Amino Acid', 'Count', 'Top Score'])\n",
        "    for row in protein_variations['Multiple AA Positions'].itertuples():\n",
        "        table.add_row([row.position, row.aa1, row.count, row.top_score])\n",
        "    print(table)\n",
        "else:\n",
        "    print(\"No positions with multiple amino acids found.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "sLxcXobH1rXg",
        "outputId": "aa18bfc3-bbf2-44e4-a368-5ea2237d9a01"
      },
      "execution_count": 104,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Missing Amino Acid Positions:\n",
            "No missing amino acid positions found.\n",
            "\n",
            "Single Amino Acid Positions:\n",
            "+----------+------------+-------+\n",
            "| Position | Amino Acid | Count |\n",
            "+----------+------------+-------+\n",
            "|    42    |     G      |   1   |\n",
            "|    43    |     P      |   1   |\n",
            "|    44    |     P      |   1   |\n",
            "|    45    |     G      |   1   |\n",
            "|    46    |     P      |   1   |\n",
            "|    47    |     P      |   1   |\n",
            "|    48    |     G      |   1   |\n",
            "|    49    |     A      |   1   |\n",
            "|    50    |     I      |   1   |\n",
            "|    51    |     G      |   1   |\n",
            "|    52    |     P      |   1   |\n",
            "|    53    |     S      |   1   |\n",
            "|    54    |     G      |   1   |\n",
            "|    55    |     P      |   1   |\n",
            "|    56    |     A      |   1   |\n",
            "|    57    |     G      |   1   |\n",
            "|    58    |     K      |   1   |\n",
            "|    72    |     G      |   1   |\n",
            "|    73    |     G      |   12  |\n",
            "|    74    |     F      |   12  |\n",
            "|    75    |     P      |   12  |\n",
            "|    76    |     G      |   12  |\n",
            "|    77    |     P      |   12  |\n",
            "|    78    |     P      |   12  |\n",
            "|    79    |     G      |   12  |\n",
            "|    80    |     M      |   12  |\n",
            "|    81    |     K      |   12  |\n",
            "|   111    |     G      |   1   |\n",
            "|   112    |     L      |   1   |\n",
            "|   113    |     K      |   1   |\n",
            "|   114    |     G      |   1   |\n",
            "|   115    |     E      |   1   |\n",
            "|   116    |     N      |   1   |\n",
            "|   117    |     G      |   1   |\n",
            "|   118    |     L      |   1   |\n",
            "|   119    |     P      |   1   |\n",
            "|   120    |     G      |   2   |\n",
            "|   121    |     E      |   2   |\n",
            "|   122    |     N      |   2   |\n",
            "|   123    |     G      |   2   |\n",
            "|   124    |     A      |   2   |\n",
            "|   125    |     P      |   2   |\n",
            "|   126    |     G      |   2   |\n",
            "|   127    |     P      |   2   |\n",
            "|   128    |     M      |   2   |\n",
            "|   129    |     G      |   2   |\n",
            "|   130    |     P      |   2   |\n",
            "|   131    |     R      |   2   |\n",
            "|   138    |     G      |   10  |\n",
            "|   146    |     A      |   23  |\n",
            "|   156    |     R      |   1   |\n",
            "|   180    |     G      |   25  |\n",
            "|   181    |     E      |   25  |\n",
            "|   182    |     V      |   25  |\n",
            "|   183    |     G      |   25  |\n",
            "|   184    |     P      |   25  |\n",
            "|   185    |     A      |   25  |\n",
            "|   186    |     G      |   25  |\n",
            "|   187    |     S      |   25  |\n",
            "|   188    |     P      |   25  |\n",
            "|   189    |     G      |   25  |\n",
            "|   190    |     S      |   25  |\n",
            "|   191    |     S      |   25  |\n",
            "|   192    |     G      |   25  |\n",
            "|   194    |     P      |   25  |\n",
            "|   195    |     G      |   25  |\n",
            "|   196    |     Q      |   25  |\n",
            "|   197    |     R      |   25  |\n",
            "|   222    |     G      |   1   |\n",
            "|   223    |     G      |   2   |\n",
            "|   238    |     L      |   29  |\n",
            "|   242    |     R      |   20  |\n",
            "|   243    |     G      |   3   |\n",
            "|   245    |     P      |   17  |\n",
            "|   258    |     R      |   16  |\n",
            "|   276    |     G      |   2   |\n",
            "|   280    |     G      |   18  |\n",
            "|   281    |     E      |   18  |\n",
            "|   282    |     A      |   18  |\n",
            "|   283    |     G      |   18  |\n",
            "|   284    |     S      |   18  |\n",
            "|   285    |     P      |   18  |\n",
            "|   286    |     G      |   18  |\n",
            "|   287    |     I      |   18  |\n",
            "|   288    |     A      |   18  |\n",
            "|   289    |     G      |   18  |\n",
            "|   290    |     P      |   18  |\n",
            "|   291    |     K      |   18  |\n",
            "|   300    |     G      |   3   |\n",
            "|   301    |     E      |   3   |\n",
            "|   302    |     P      |   3   |\n",
            "|   303    |     G      |   3   |\n",
            "|   304    |     A      |   3   |\n",
            "|   305    |     N      |   3   |\n",
            "|   306    |     G      |   3   |\n",
            "|   308    |     P      |   3   |\n",
            "|   309    |     G      |   3   |\n",
            "|   311    |     A      |   3   |\n",
            "|   312    |     G      |   3   |\n",
            "|   313    |     E      |   3   |\n",
            "|   314    |     R      |   3   |\n",
            "|   315    |     G      |   1   |\n",
            "|   335    |     P      |   13  |\n",
            "|   361    |     G      |   26  |\n",
            "|   366    |     R      |   24  |\n",
            "|   367    |     G      |   8   |\n",
            "|   369    |     P      |   8   |\n",
            "|   370    |     G      |   8   |\n",
            "|   371    |     S      |   8   |\n",
            "|   372    |     P      |   8   |\n",
            "|   373    |     G      |   8   |\n",
            "|   374    |     G      |   8   |\n",
            "|   375    |     P      |   8   |\n",
            "|   376    |     G      |   8   |\n",
            "|   377    |     S      |   8   |\n",
            "|   378    |     D      |   8   |\n",
            "|   379    |     G      |   8   |\n",
            "|   383    |     P      |   5   |\n",
            "|   395    |     P      |   16  |\n",
            "|   403    |     Q      |   20  |\n",
            "|   404    |     P      |   20  |\n",
            "|   405    |     G      |   20  |\n",
            "|   406    |     V      |   20  |\n",
            "|   407    |     M      |   20  |\n",
            "|   408    |     G      |   20  |\n",
            "|   409    |     F      |   20  |\n",
            "|   410    |     P      |   20  |\n",
            "|   411    |     G      |   20  |\n",
            "|   412    |     P      |   20  |\n",
            "|   413    |     K      |   20  |\n",
            "|   426    |     G      |   5   |\n",
            "|   427    |     G      |   7   |\n",
            "|   428    |     G      |   7   |\n",
            "|   429    |     P      |   7   |\n",
            "|   430    |     G      |   7   |\n",
            "|   431    |     G      |   7   |\n",
            "|   432    |     P      |   7   |\n",
            "|   433    |     G      |   7   |\n",
            "|   434    |     P      |   7   |\n",
            "|   435    |     Q      |   7   |\n",
            "|   436    |     G      |   7   |\n",
            "|   437    |     P      |   7   |\n",
            "|   438    |     A      |   7   |\n",
            "|   439    |     G      |   7   |\n",
            "|   441    |     G      |   3   |\n",
            "|   442    |     E      |   3   |\n",
            "|   443    |     T      |   3   |\n",
            "|   444    |     G      |   3   |\n",
            "|   445    |     P      |   3   |\n",
            "|   446    |     Q      |   3   |\n",
            "|   447    |     G      |   3   |\n",
            "|   448    |     P      |   3   |\n",
            "|   449    |     P      |   3   |\n",
            "|   450    |     G      |   3   |\n",
            "|   451    |     P      |   3   |\n",
            "|   452    |     T      |   3   |\n",
            "|   453    |     G      |   3   |\n",
            "|   454    |     P      |   3   |\n",
            "|   456    |     G      |   3   |\n",
            "|   457    |     D      |   3   |\n",
            "|   458    |     K      |   3   |\n",
            "|   492    |     G      |   3   |\n",
            "|   502    |     G      |   18  |\n",
            "|   515    |     P      |   28  |\n",
            "|   518    |     A      |   28  |\n",
            "|   519    |     G      |   28  |\n",
            "|   521    |     P      |   28  |\n",
            "|   525    |     R      |   25  |\n",
            "|   526    |     G      |   6   |\n",
            "|   530    |     P      |   6   |\n",
            "|   535    |     G      |   6   |\n",
            "|   537    |     K      |   1   |\n",
            "|   538    |     G      |   1   |\n",
            "|   539    |     A      |   1   |\n",
            "|   540    |     A      |   1   |\n",
            "|   541    |     G      |   1   |\n",
            "|   542    |     P      |   1   |\n",
            "|   543    |     P      |   1   |\n",
            "|   571    |     G      |   5   |\n",
            "|   572    |     D      |   5   |\n",
            "|   573    |     K      |   5   |\n",
            "|   574    |     G      |   6   |\n",
            "|   575    |     E      |   6   |\n",
            "|   576    |     P      |   6   |\n",
            "|   577    |     G      |   6   |\n",
            "|   578    |     S      |   6   |\n",
            "|   579    |     S      |   6   |\n",
            "|   580    |     G      |   7   |\n",
            "|   581    |     V      |   7   |\n",
            "|   582    |     D      |   7   |\n",
            "|   583    |     G      |   7   |\n",
            "|   584    |     A      |   7   |\n",
            "|   585    |     P      |   7   |\n",
            "|   586    |     G      |   7   |\n",
            "|   587    |     K      |   7   |\n",
            "|   588    |     D      |   7   |\n",
            "|   589    |     G      |   7   |\n",
            "|   590    |     P      |   7   |\n",
            "|   599    |     P      |   18  |\n",
            "|   624    |     R      |   18  |\n",
            "|   664    |     G      |   5   |\n",
            "|   665    |     E      |   5   |\n",
            "|   666    |     G      |   5   |\n",
            "|   667    |     G      |   5   |\n",
            "|   668    |     P      |   5   |\n",
            "|   669    |     P      |   5   |\n",
            "|   670    |     G      |   5   |\n",
            "|   671    |     A      |   5   |\n",
            "|   672    |     A      |   5   |\n",
            "|   673    |     G      |   5   |\n",
            "|   674    |     P      |   5   |\n",
            "|   675    |     A      |   5   |\n",
            "|   676    |     G      |   5   |\n",
            "|   677    |     G      |   5   |\n",
            "|   678    |     S      |   5   |\n",
            "|   679    |     G      |   5   |\n",
            "|   680    |     P      |   5   |\n",
            "|   681    |     A      |   5   |\n",
            "|   682    |     G      |   5   |\n",
            "|   683    |     P      |   5   |\n",
            "|   684    |     P      |   5   |\n",
            "|   685    |     G      |   5   |\n",
            "|   686    |     P      |   5   |\n",
            "|   687    |     Q      |   5   |\n",
            "|   688    |     G      |   5   |\n",
            "|   689    |     V      |   5   |\n",
            "|   697    |     G      |   28  |\n",
            "|   708    |     R      |   24  |\n",
            "|   762    |     G      |   6   |\n",
            "|   770    |     P      |   25  |\n",
            "|   787    |     L      |   29  |\n",
            "|   788    |     A      |   29  |\n",
            "|   789    |     G      |   29  |\n",
            "|   790    |     P      |   29  |\n",
            "|   791    |     P      |   29  |\n",
            "|   792    |     G      |   29  |\n",
            "|   793    |     M      |   29  |\n",
            "|   794    |     P      |   29  |\n",
            "|   795    |     G      |   29  |\n",
            "|   797    |     R      |   29  |\n",
            "|   799    |     G      |   1   |\n",
            "|   800    |     S      |   1   |\n",
            "|   801    |     P      |   1   |\n",
            "|   802    |     G      |   1   |\n",
            "|   803    |     P      |   1   |\n",
            "|   804    |     Q      |   1   |\n",
            "|   805    |     G      |   1   |\n",
            "|   806    |     I      |   1   |\n",
            "|   807    |     K      |   1   |\n",
            "|   822    |     G      |   21  |\n",
            "|   823    |     P      |   21  |\n",
            "|   824    |     P      |   21  |\n",
            "|   825    |     G      |   21  |\n",
            "|   826    |     P      |   21  |\n",
            "|   827    |     Q      |   21  |\n",
            "|   828    |     G      |   21  |\n",
            "|   829    |     L      |   21  |\n",
            "|   830    |     P      |   21  |\n",
            "|   831    |     G      |   21  |\n",
            "|   832    |     L      |   21  |\n",
            "|   833    |     A      |   21  |\n",
            "|   834    |     G      |   21  |\n",
            "|   835    |     T      |   21  |\n",
            "|   837    |     G      |   21  |\n",
            "|   838    |     E      |   21  |\n",
            "|   839    |     P      |   21  |\n",
            "|   840    |     G      |   21  |\n",
            "|   841    |     R      |   21  |\n",
            "|   842    |     D      |   8   |\n",
            "|   843    |     G      |   8   |\n",
            "|   844    |     N      |   8   |\n",
            "|   845    |     P      |   8   |\n",
            "|   846    |     G      |   8   |\n",
            "|   847    |     S      |   8   |\n",
            "|   848    |     D      |   8   |\n",
            "|   849    |     G      |   8   |\n",
            "|   850    |     L      |   8   |\n",
            "|   851    |     P      |   8   |\n",
            "|   852    |     G      |   8   |\n",
            "|   853    |     R      |   8   |\n",
            "|   854    |     R      |   2   |\n",
            "|   855    |     D      |   2   |\n",
            "|   856    |     G      |   2   |\n",
            "|   857    |     A      |   2   |\n",
            "|   858    |     P      |   2   |\n",
            "|   859    |     G      |   2   |\n",
            "|   860    |     A      |   2   |\n",
            "|   861    |     G      |   3   |\n",
            "|   862    |     D      |   3   |\n",
            "|   863    |     R      |   3   |\n",
            "|   864    |     G      |   6   |\n",
            "|   865    |     E      |   6   |\n",
            "|   866    |     N      |   6   |\n",
            "|   867    |     G      |   6   |\n",
            "|   868    |     S      |   6   |\n",
            "|   869    |     P      |   6   |\n",
            "|   870    |     G      |   6   |\n",
            "|   871    |     A      |   6   |\n",
            "|   872    |     P      |   6   |\n",
            "|   873    |     G      |   6   |\n",
            "|   875    |     P      |   6   |\n",
            "|   876    |     G      |   6   |\n",
            "|   877    |     H      |   6   |\n",
            "|   921    |     R      |   4   |\n",
            "|   940    |     G      |   6   |\n",
            "|   941    |     F      |   6   |\n",
            "|   942    |     P      |   6   |\n",
            "|   943    |     G      |   6   |\n",
            "|   944    |     N      |   6   |\n",
            "|   945    |     P      |   6   |\n",
            "|   946    |     G      |   6   |\n",
            "|   947    |     A      |   6   |\n",
            "|   977    |     P      |   21  |\n",
            "|   992    |     P      |   17  |\n",
            "|   996    |     R      |   15  |\n",
            "+----------+------------+-------+\n",
            "\n",
            "Multiple Amino Acid Positions:\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'Pandas' object has no attribute 'top_score'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-104-228a4237c609>\u001b[0m in \u001b[0;36m<cell line: 58>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPrettyTable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Position'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Amino Acid'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Count'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Top Score'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprotein_variations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Multiple AA Positions'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitertuples\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m         \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_row\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maa1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtop_score\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'Pandas' object has no attribute 'top_score'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "def analyze_protein_variations(gnc_df):\n",
        "    # Group data by position\n",
        "    grouped_by_position = gnc_df.groupby('position')\n",
        "\n",
        "    # Initialize containers for the analysis results\n",
        "    missing_aa_positions = []\n",
        "    single_aa_summary = []\n",
        "    multiple_aa_analysis = []\n",
        "\n",
        "    # Analyze each position\n",
        "    for position, group in grouped_by_position:\n",
        "        # Identify missing amino acids by checking if all values in 'aa1' are NaN\n",
        "        if group['aa1'].isna().all():\n",
        "            missing_aa_positions.append(position)\n",
        "            continue  # Move to the next position if the current one has no data\n",
        "\n",
        "        # Analyze positions with a single amino acid\n",
        "        if group['aa1'].nunique() == 1:\n",
        "            aa = group['aa1'].dropna().unique()[0]\n",
        "            count = group['n_aa1'].sum()  # Summing up occurrences across all samples\n",
        "            single_aa_summary.append({'position': position, 'amino_acid': aa, 'count': count})\n",
        "        else:\n",
        "            # Analyze positions with multiple amino acids\n",
        "            aa_counts = group['aa1'].value_counts().reset_index()\n",
        "            aa_counts.columns = ['amino_acid', 'count']\n",
        "            for _, row in aa_counts.iterrows():\n",
        "                multiple_aa_analysis.append({'position': position, 'amino_acid': row['amino_acid'], 'count': row['count']})\n",
        "\n",
        "    # Convert analysis results into DataFrame for easy handling\n",
        "    missing_aa_df = pd.DataFrame(missing_aa_positions, columns=['Position'])\n",
        "    single_aa_df = pd.DataFrame(single_aa_summary)\n",
        "    multiple_aa_df = pd.DataFrame(multiple_aa_analysis)\n",
        "\n",
        "    # Compile the summary into a dictionary\n",
        "    summary = {\n",
        "        'Missing AA Positions': missing_aa_df,\n",
        "        'Single AA Positions': single_aa_df,\n",
        "        'Multiple AA Positions': multiple_aa_df\n",
        "    }\n",
        "\n",
        "    return summary\n",
        "\n",
        "# Assuming 'gnc_df' is your DataFrame filtered for a specific GNC\n",
        "# Example usage:\n",
        "# protein_variations = analyze_protein_variations(gnc_df)\n",
        "# The 'protein_variations' dictionary now contains three keys with DataFrames as their values.\n"
      ],
      "metadata": {
        "id": "He11tMcj0x4l"
      },
      "execution_count": 101,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Print missing amino acid positions\n",
        "print(\"Missing Amino Acid Positions:\")\n",
        "if protein_variations['Missing AA Positions']:\n",
        "  table = PrettyTable(['Position'])\n",
        "  table.add_rows([[position] for position in protein_variations['Missing AA Positions']])\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No missing amino acid positions found.\")\n",
        "\n",
        "# Print single amino acid positions\n",
        "print(\"\\nSingle Amino Acid Positions:\")\n",
        "if protein_variations['Single AA Positions'].any():\n",
        "  table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "  table.add_rows(protein_variations['Single AA Positions'].to_records(index=False))\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No positions with single amino acids found.\")\n",
        "\n",
        "# Print multiple amino acid positions\n",
        "print(\"\\nMultiple Amino Acid Positions:\")\n",
        "if protein_variations['Multiple AA Positions'].any():\n",
        "  table = PrettyTable(['Position', 'Amino Acid', 'Count', 'Top Score'])\n",
        "  table.add_rows(protein_variations['Multiple AA Positions'].to_records(index=False))\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No positions with multiple amino acids found.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 228
        },
        "id": "S4Pw_vp2fLcH",
        "outputId": "4ba85d6d-54a0-4bed-c45a-ce08e30fd0b9"
      },
      "execution_count": 102,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Missing Amino Acid Positions:\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'protein_variations' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-102-53d435f58506>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Print missing amino acid positions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Missing Amino Acid Positions:\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mif\u001b[0m \u001b[0mprotein_variations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Missing AA Positions'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m   \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPrettyTable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Position'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_rows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mposition\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprotein_variations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Missing AA Positions'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'protein_variations' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def analyze_protein_data(df, gnc):\n",
        "    # Filter the DataFrame for the specified GNC\n",
        "    df_gnc = group_df[df['GNC'] == gnc]\n",
        "\n",
        "    # Prepare the output DataFrame structure\n",
        "    columns = ['position', 'sample_name', 'amino_acid', 'count', 'score1', 'PTM_type', 'PTM_score']\n",
        "    analysis_results = pd.DataFrame(columns=columns)\n",
        "\n",
        "    # Iterate through each unique position\n",
        "    for position in df_gnc['position'].unique():\n",
        "        df_position = df_gnc[df_gnc['position'] == position]\n",
        "\n",
        "        # Iterate through each sample_name\n",
        "        for sample in df_position['sample_name'].unique():\n",
        "            df_sample = df_position[df_position['sample_name'] == sample]\n",
        "\n",
        "            # Analyze amino acids and PTMs for this sample and position\n",
        "            for aa_col in ['aa1', 'aa2', 'aa3']:\n",
        "                aa_data = df_sample[[aa_col, f'n_{aa_col}', f'score1_{aa_col}']].dropna()\n",
        "\n",
        "                for _, row in aa_data.iterrows():\n",
        "                    amino_acid = row[aa_col]\n",
        "                    count = row[f'n_{aa_col}']\n",
        "                    score1 = row[f'score1_{aa_col}']\n",
        "\n",
        "                    # Add PTM analysis here if applicable\n",
        "                    ptm_types = []  # Placeholder for actual PTM analysis\n",
        "                    ptm_scores = []  # Placeholder for actual PTM analysis\n",
        "\n",
        "                    # Append to results DataFrame\n",
        "                    result_row = pd.DataFrame([[position, sample, amino_acid, count, score1, ptm_types, ptm_scores]], columns=columns)\n",
        "                    analysis_results = pd.concat([analysis_results, result_row], ignore_index=True)\n",
        "\n",
        "    return analysis_results\n",
        "\n",
        "# Assuming 'df' is your DataFrame and 'COL3A1' is the GNC of interest\n",
        "results_df = analyze_protein_data(group_df, 'COL3A1')\n"
      ],
      "metadata": {
        "id": "st5j29puvwhJ"
      },
      "execution_count": 100,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def analyze_protein_variations(gnc_df):\n",
        "    # Group data by position\n",
        "    grouped_by_position = gnc_df.groupby('position')\n",
        "\n",
        "    # Analyze positions with missing amino acids\n",
        "    missing_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].isna().any())['position'].tolist()\n",
        "\n",
        "    # Analyze positions with a single amino acid\n",
        "    single_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() == 1)\n",
        "    single_aa_summary = single_aa_positions[['position', 'aa1']].value_counts().reset_index(name='count')\n",
        "\n",
        "    # Analyze positions with multiple amino acids\n",
        "    multiple_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() > 1)\n",
        "    multiple_aa_analysis = []\n",
        "    if not multiple_aa_positions.empty:  # Check if there are positions with multiple amino acids\n",
        "        for (position, idx), group_data in multiple_aa_positions.groupby(level=0):\n",
        "            # Check for any non-missing amino acid column (handle potential MultiIndex)\n",
        "            amino_acid_cols = ['aa1', 'aa2', 'aa3']  # Assuming amino acid columns\n",
        "            for aa_col in amino_acid_cols:\n",
        "                if not group_data[aa_col].isna().all():\n",
        "                    aa = group_data[aa_col].dropna().iloc[0]  # Get the first non-missing amino acid\n",
        "                    count = group_data[aa_col].value_counts().get(aa, 0)  # Count occurrences\n",
        "\n",
        "                    multiple_aa_analysis.append({'position': position, 'aa1': aa, 'count': count})\n",
        "                    break  # Exit the inner loop after finding a non-missing amino acid column\n",
        "\n",
        "    # Combine results\n",
        "    summary = {\n",
        "        'Missing AA Positions': missing_aa_positions,\n",
        "        'Single AA Positions': single_aa_summary,\n",
        "        'Multiple AA Positions': multiple_aa_analysis\n",
        "    }\n",
        "\n",
        "    summary = {\n",
        "        'Missing AA Positions': missing_aa_positions,\n",
        "        'Single AA Positions': single_aa_summary,\n",
        "        'Multiple AA Positions': multiple_aa_analysis\n",
        "    }\n",
        "    return summary"
      ],
      "metadata": {
        "id": "2OABvwNFfDVT"
      },
      "execution_count": 95,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "protein_variations = analyze_protein_variations(gnc_df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 263
        },
        "id": "om2vs7K1gO9G",
        "outputId": "29d72627-2b68-4f6c-b827-bfaebfc8e593"
      },
      "execution_count": 96,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "cannot unpack non-iterable int object",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-96-db7f9df0937d>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprotein_variations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0manalyze_protein_variations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnc_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-95-e357282611c3>\u001b[0m in \u001b[0;36manalyze_protein_variations\u001b[0;34m(gnc_df)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0mmultiple_aa_analysis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mmultiple_aa_positions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Check if there are positions with multiple amino acids\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_data\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmultiple_aa_positions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m             \u001b[0;31m# Check for any non-missing amino acid column (handle potential MultiIndex)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0mamino_acid_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'aa1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa3'\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# Assuming amino acid columns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: cannot unpack non-iterable int object"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def analyze_protein_variations(gnc_df):\n",
        "    # Group data by position\n",
        "    grouped_by_position = gnc_df.groupby('position')\n",
        "\n",
        "    # Analyze positions with missing amino acids\n",
        "    missing_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].isna().all())['position'].tolist()\n",
        "\n",
        "    # Analyze positions with a single amino acid\n",
        "    single_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() == 1)\n",
        "    single_aa_summary = single_aa_positions[['position', 'aa1']].value_counts().reset_index(name='count')\n",
        "\n",
        "    # Analyze positions with multiple amino acids\n",
        "    multiple_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() > 1)\n",
        "    multiple_aa_analysis = []\n",
        "    if not multiple_aa_positions.empty:  # Check if there are positions with multiple amino acids\n",
        "        for (position, idx), group_data in multiple_aa_positions:\n",
        "            # Check for any non-missing amino acid column (handle potential MultiIndex)\n",
        "            amino_acid_cols = ['aa1', 'aa2', 'aa3']  # Assuming amino acid columns\n",
        "            for aa_col in amino_acid_cols:\n",
        "                if not group_data[aa_col].isna().all():\n",
        "                    aa = group_data[aa_col].dropna().iloc[0]  # Get the first non-missing amino acid\n",
        "                    count = group_data[aa_col].value_counts().get(aa, 0)  # Count occurrences\n",
        "\n",
        "                    # Access PTM scores dynamically based on the amino acid column\n",
        "                    ptm_scores = [group_data[f'ptm{i}_{aa_col}'].dropna().iloc[idx] for i in range(1, len(group_data.columns) // 2 + 1) if not pd.isna(group_data[f'ptm{i}_{aa_col}'].iloc[idx])]\n",
        "                    top_score = max(ptm_scores) if ptm_scores else None\n",
        "\n",
        "                    multiple_aa_analysis.append({'position': position, 'aa1': aa, 'count': count, 'top_score': top_score})\n",
        "                    break  # Exit the inner loop after finding a non-missing amino acid column\n",
        "\n",
        "    # Combine results\n",
        "    summary = {\n",
        "        'Missing AA Positions': missing_aa_positions,\n",
        "        'Single AA Positions': single_aa_summary,\n",
        "        'Multiple AA Positions': pd.DataFrame(multiple_aa_analysis)\n",
        "    }\n",
        "    return summary\n",
        "\n",
        "# Analyze variations in the protein\n",
        "protein_variations = analyze_protein_variations(gnc_df.copy())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 297
        },
        "id": "EXZADNJtL2I2",
        "outputId": "16b6fbdd-b32e-4aaf-e447-167665bfc5ac"
      },
      "execution_count": 91,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "too many values to unpack (expected 2)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-91-e57a575774cc>\u001b[0m in \u001b[0;36m<cell line: 40>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[0;31m# Analyze variations in the protein\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m \u001b[0mprotein_variations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0manalyze_protein_variations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnc_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-91-e57a575774cc>\u001b[0m in \u001b[0;36manalyze_protein_variations\u001b[0;34m(gnc_df)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0mmultiple_aa_analysis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mmultiple_aa_positions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Check if there are positions with multiple amino acids\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_data\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmultiple_aa_positions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m             \u001b[0;31m# Check for any non-missing amino acid column (handle potential MultiIndex)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0mamino_acid_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'aa1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa3'\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# Assuming amino acid columns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 2)"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# # Function to analyze variations in a protein\n",
        "# def analyze_protein_variations(gnc_df):\n",
        "#   # Group data by position\n",
        "#   grouped_by_position = gnc_df.groupby('position')\n",
        "\n",
        "#   # Analyze positions with missing amino acids\n",
        "#   missing_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].isna().any())['position'].tolist()\n",
        "\n",
        "#   # Analyze positions with a single amino acid\n",
        "#   single_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() == 1)\n",
        "#   single_aa_summary = single_aa_positions[['position', 'aa1']].value_counts().reset_index(name='count')\n",
        "\n",
        "#   # Analyze positions with multiple amino acids\n",
        "#   multiple_aa_positions = grouped_by_position.filter(lambda x: x['aa1'].nunique() > 1)\n",
        "#   multiple_aa_analysis = []\n",
        "#   for position, group_data in multiple_aa_positions:\n",
        "#     for aa_col in ['aa1', 'aa2', 'aa3']:  # Assuming you have 3 amino acid columns (aa1, aa2, aa3)\n",
        "#       if not pd.isna(group_data[aa_col].any()):  # Check if there's data in this column\n",
        "#         aa = group_data[aa_col].iloc[0]  # Assuming the first non-null value is the amino acid\n",
        "#         count = group_data[aa_col].value_counts().get(aa, 0)  # Count occurrences (default to 0 if not present)\n",
        "#         ptm_scores = [group_data[f'ptm{i}_score1'].iloc[idx] for i in range(1, len(group_data.columns) // 2 + 1) if not pd.isna(group_data[f'ptm{i}_{aa_col}'].iloc[idx])]\n",
        "#         top_score = max(ptm_scores) if ptm_scores else None\n",
        "#         multiple_aa_analysis.append({'position': position, 'aa1': aa, 'count': count, 'top_score': top_score})\n",
        "\n",
        "\n",
        "#   # Combine results\n",
        "#   summary = {\n",
        "#       'Missing AA Positions': missing_aa_positions,\n",
        "#       'Single AA Positions': single_aa_summary,\n",
        "#       'Multiple AA Positions': multiple_aa_df\n",
        "#   }\n",
        "#   return summary\n",
        "\n",
        "# # Analyze variations in the protein\n",
        "# protein_variations = analyze_protein_variations(gnc_df.copy())\n",
        "\n",
        "# Print the summary tables (assuming PrettyTable is available)\n",
        "from prettytable import PrettyTable\n",
        "\n",
        "# Print missing amino acid positions\n",
        "print(\"Missing Amino Acid Positions:\")\n",
        "if protein_variations['Missing AA Positions']:\n",
        "  table = PrettyTable(['Position'])\n",
        "  table.add_rows([[position] for position in protein_variations['Missing AA Positions']])\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No missing amino acid positions found.\")\n",
        "\n",
        "# Print single amino acid positions\n",
        "print(\"\\nSingle Amino Acid Positions:\")\n",
        "if protein_variations['Single AA Positions'].any():\n",
        "  table = PrettyTable(['Position', 'Amino Acid', 'Count'])\n",
        "  table.add_rows(protein_variations['Single AA Positions'].to_records(index=False))\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No positions with single amino acids found.\")\n",
        "\n",
        "# Print multiple amino acid positions\n",
        "print(\"\\nMultiple Amino Acid Positions:\")\n",
        "if protein_variations['Multiple AA Positions'].any():\n",
        "  table = PrettyTable(['Position', 'Amino Acid', 'Count', 'Top Score'])\n",
        "  table.add_rows(protein_variations['Multiple AA Positions'].to_records(index=False))\n",
        "  print(table)\n",
        "else:\n",
        "  print(\"No positions with multiple amino acids found.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "gAdcsfYlKO18",
        "outputId": "e862a677-b4cb-47f5-957f-bd9558638258"
      },
      "execution_count": 83,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "too many values to unpack (expected 2)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-83-426033edc7c3>\u001b[0m in \u001b[0;36m<cell line: 35>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;31m# Analyze variations in the protein\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m \u001b[0mprotein_variations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0manalyze_protein_variations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnc_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m \u001b[0;31m# Print the summary tables (assuming PrettyTable is available)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-83-426033edc7c3>\u001b[0m in \u001b[0;36manalyze_protein_variations\u001b[0;34m(gnc_df)\u001b[0m\n\u001b[1;32m     14\u001b[0m   \u001b[0mmultiple_aa_positions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgrouped_by_position\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'aa1'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnunique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m   \u001b[0mmultiple_aa_analysis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m   \u001b[0;32mfor\u001b[0m \u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_data\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmultiple_aa_positions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0maa_col\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'aa1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'aa3'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Assuming you have 3 amino acid columns (aa1, aa2, aa3)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgroup_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maa_col\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Check if there's data in this column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 2)"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "XiL5PhuyCGNq"
      },
      "outputs": [],
      "source": [
        "# # Filter the DataFrame to find rows with ptm \"Deamidated\"\n",
        "# filtered_ptm_df = df_filtered[df_filtered['ptm'] == \"Deamidated\"]\n",
        "\n",
        "# # Check if any rows were found\n",
        "# if not filtered_ptm_df.empty:\n",
        "#   # Print a limited number of rows (e.g., the first 3)\n",
        "#   print(filtered_ptm_df.head(3))\n",
        "# else:\n",
        "#   # If no rows are found, print a message\n",
        "#   print(\"No rows found with ptm='Deamidated'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "TQJ73qwnvsvi"
      },
      "outputs": [],
      "source": [
        "# unique_names = result_with_ptms['sample_name'].unique()\n",
        "# name_counts = result_with_ptms['sample_name'].value_counts()\n",
        "# print(\"Unique sample names:\", unique_names)\n",
        "# print(\"Counts of each sample name:\", name_counts)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}